<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="ipynb_website:version" content="0.9.7" />
<meta name="viewport" content="width=device-width, initial-scale=1" />

<link rel="stylesheet" type="text/css" href="../css/jt.css">

<link rel="stylesheet" type="text/css" href="../css/toc2.css">

<link href="../site_libs/jqueryui-1.11.4/jquery-ui.css">
<link rel="stylesheet" href="../site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<link rel="stylesheet" href="../site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jqueryui/1.9.1/jquery-ui.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>

<link rel="stylesheet"
      href="../site_libs/highlightjs/null.min.css"
      type="text/css" />

<script src="../site_libs/highlightjs/highlight.pack.js"></script>
<script>hljs.initHighlightingOnLoad();</script>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>

<script src="../js/doc_toc.js"></script>
<script src="../js/docs.js"></script>

<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.2/MathJax.js?config=TeX-MML-AM_CHTML"></script>
<script>
    MathJax.Hub.Config({
        extensions: ["tex2jax.js"],
        jax: ["input/TeX", "output/HTML-CSS"],
        tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true
        },
        "HTML-CSS": {
            preferredFont: "TeX",
            availableFonts: ["TeX"],
            styles: {
                scale: 110,
                ".MathJax_Display": {
                    "font-size": "110%",
                }
            }
        }
    });
</script>
<script>
function filterDataFrame(id) {
    var input = document.getElementById("search_" + id);
    var filter = input.value.toUpperCase();
    var table = document.getElementById("dataframe_" + id);
    var tr = table.getElementsByTagName("tr");
    // Loop through all table rows, and hide those who don't match the search query
    for (var i = 1; i < tr.length; i++) {
        for (var j = 0; j < tr[i].cells.length; ++j) {
            var matched = false;
            if (tr[i].cells[j].innerHTML.toUpperCase().indexOf(filter) != -1) {
                tr[i].style.display = "";
                matched = true
                break;
            }
            if (!matched)
                tr[i].style.display = "none";
        }
    }
}
function sortDataFrame(id, n, dtype) {
    var table = document.getElementById("dataframe_" + id);
    var tb = table.tBodies[0]; // use `<tbody>` to ignore `<thead>` and `<tfoot>` rows
    var tr = Array.prototype.slice.call(tb.rows, 0); // put rows into array
    if (dtype === 'numeric') {
        var fn = function(a, b) { 
            return parseFloat(a.cells[n].textContent) <= parseFloat(b.cells[n].textContent) ? -1 : 1;
        }
    } else {
        var fn = function(a, b) {
            var c = a.cells[n].textContent.trim().localeCompare(b.cells[n].textContent.trim()); 
            return c > 0 ? 1 : (c < 0 ? -1 : 0) }
    }
    var isSorted = function(array, fn) {
        if (array.length < 2)
            return 1;
        var direction = fn(array[0], array[1]); 
        for (var i = 1; i < array.length - 1; ++i) {
            var d = fn(array[i], array[i+1]);
            if (d == 0)
                continue;
            else if (direction == 0)
                direction = d;
            else if (direction != d)
                return 0;
            }
        return direction;
    }
    var sorted = isSorted(tr, fn);
    if (sorted == 1 || sorted == -1) {
        // if sorted already, reverse it
        for(var i = tr.length - 1; i >= 0; --i)
            tb.appendChild(tr[i]); // append each row in order
    } else {
        tr = tr.sort(fn);
        for(var i = 0; i < tr.length; ++i)
            tb.appendChild(tr[i]); // append each row in order
    }
}
</script>

<script>
$( document ).ready(function(){
            var cfg={'threshold':3,     // depth of toc (number of levels)
             'number_sections': false,
             'toc_cell': false,          // useless here
             'toc_window_display': true, // display the toc window
             "toc_section_display": "block", // display toc contents in the window
             'sideBar':true,       // sidebar or floating window
             'navigate_menu':false       // navigation menu (only in liveNotebook -- do not change)
            }
            var st={};                  // some variables used in the script
            st.rendering_toc_cell = false;
            st.config_loaded = false;
            st.extension_initialized=false;
            st.nbcontainer_marginleft = $('#notebook-container').css('margin-left')
            st.nbcontainer_marginright = $('#notebook-container').css('margin-right')
            st.nbcontainer_width = $('#notebook-container').css('width')
            st.oldTocHeight = undefined
            st.cell_toc = undefined;
            st.toc_index=0;
            // fire the main function with these parameters
            table_of_contents(cfg, st);
            var file=variant-annotationDict[$("h1:first").attr("id")];
            $("#toc-level0 a").css("color","#126dce");
            $('a[href="#'+$("h1:first").attr("id")+'"]').hide()
            var docs=variant-annotationArray;
            var docs_map=variant-annotationArrayMap;
            var pos=variant-annotationArray.indexOf(file);
            for (var a=pos;a>=0;a--){
                  $('<li><a href="'+docs[a]+'.html"><font color="#073642"><b>'+docs_map[docs[a]].replace(/_/g," ")+'</b></font></a></li>').insertBefore("#toc-level0 li:eq(0)");
            }
            $('a[href="'+file+'.html'+'"]').css("color","#126dce");
            for (var a=pos+1;a<docs.length;a++){
                  $(".toc #toc-level0").append('<li><a href="'+docs[a]+'.html"><font color="#073642"><b>'+docs_map[docs[a]].replace(/_/g," ")+'</b></font></a></li>');
            }
            // $("#toc-header").hide(); // comment out because it prevents search bar from displaying
    });
</script>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');
  // mark it active
  menuAnchor.parent().addClass('active');
  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>
<div class="container-fluid main-container">
<!-- tabsets -->
<script src="../site_libs/navigation-1.1/tabsets.js"></script>
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>



<title>Some computational genomics workflows</title>

<style type = "text/css">
body {
  font-family: "Droid Sans";
  padding-top: 66px;
  padding-bottom: 40px;
}
</style>
</head>

<body>
<div tabindex="-1" id="notebook" class="border-box-sizing">
<div class="container" id="notebook-container">

<!-- code folding -->

<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">Some computational genomics workflows</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
<li>
  <a href="../index.html">Overview</a>
</li>
        
      </ul>
        
<ul class="nav navbar-nav navbar-right">
<li>
   <a href="http://github.com/cumc/bioworkflows"> <span class="fa fa-github"></span> </a>
</li>
</ul>
        
      </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="Annotation-of-exome-variants-using-Annovar">Annotation of exome variants using Annovar<a class="anchor-link" href="#Annotation-of-exome-variants-using-Annovar">&#182;</a></h1>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Aim">Aim<a class="anchor-link" href="#Aim">&#182;</a></h2><p>Prepare the data for further association analyses using the LMM.ipynb on rare variants.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Description-of-the-pipeline">Description of the pipeline<a class="anchor-link" href="#Description-of-the-pipeline">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>This pipeline provides 3 different possibilities depending on the type of input data you are starting with:</p>
<h3 id="Scenario-1-:-you-have-multiple-bim-files-(e.g.-one-per-chromosome)-and-you-want-to-merge-them-into-one-file-for-later-annotation-with-annovar">Scenario 1 : you have multiple bim files (e.g. one per chromosome) and you want to merge them into one file for later annotation with annovar<a class="anchor-link" href="#Scenario-1-:-you-have-multiple-bim-files-(e.g.-one-per-chromosome)-and-you-want-to-merge-them-into-one-file-for-later-annotation-with-annovar">&#182;</a></h3><p>Run <code>bim_merge</code> to concatenate all the bim files and then run <code>annovar</code> to annotate all the variants at once</p>
<h3 id="Scenario-2:-you-either-want-to-work-with-common-or-rare-variants.">Scenario 2: you either want to work with common or rare variants.<a class="anchor-link" href="#Scenario-2:-you-either-want-to-work-with-common-or-rare-variants.">&#182;</a></h3><p>Run <code>get_snps</code> using the <code>--maf</code> or <code>max-maf</code> depending on the type of variants you would like to extract and then run <code>annovar</code></p>
<h3 id="Scenario-3:-you-already-have-a-specific-list-of-variants-you-would-like-to-annotate-stored-in-a-bim-file.">Scenario 3: you already have a specific list of variants you would like to annotate stored in a bim file.<a class="anchor-link" href="#Scenario-3:-you-already-have-a-specific-list-of-variants-you-would-like-to-annotate-stored-in-a-bim-file.">&#182;</a></h3><p>Run <code>annovar</code></p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Command-interface">Command interface<a class="anchor-link" href="#Command-interface">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[1]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="err">!</span><span class="n">sos</span> <span class="n">run</span> <span class="n">annovar</span><span class="o">.</span><span class="n">ipynb</span> <span class="o">-</span><span class="n">h</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

    <div class="prompt"></div>


<div class="output_subarea output_stream output_stdout output_text">
<pre><span class="ansi-red-intense-fg">ERROR</span>: <span class="ansi-red-intense-fg">Notebook JSON is invalid: %s</span>
usage: sos run annovar.ipynb [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &#34;sos run -h&#34; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  bim_merge
  get_snps
  annovar

Global Workflow Options:
  --cwd VAL (as path, required)
                        the output directory for generated files
  --numThreads 2 (as int)
                        Specific number of threads to use
  --job-size 1 (as int)
                        For cluster jobs, number commands to run per job
  --build hg38
                        Human genome build
  --bim-name VAL (as path, required)
                        Name for the merged bimfiles
  --name-prefix VAL (as str, required)
                        Prefix for the name based on common/rare variant
                        filtering
  --annovar-module &#39;\nmodule load ANNOVAR/2020Jun08-foss-2018b-Perl-5.28.0\necho &#34;Module annovar loaded&#34;\n{cmd}\n&#39;
                        Load annovar module from cluster
  --container-annovar &#39;gaow/gatk4-annovar&#39;
                        Software container option
  --container-lmm &#39;statisticalgenetics/lmm:2.0&#39;

Sections
  bim_merge:            Merge all the bimfiles into a single file to use later
                        with awk Only need to run this cell once
    Workflow Options:
      --bimfiles  paths

  get_snps_1:           Get a list of common SNPs above (--maf) or rare SNPs
                        below (--max-maf) certain MAF
    Workflow Options:
      --bfiles  paths

                        bed files plink format
      --maf-filter 0.0 (as float)
                        Filter based on minor allele frequency (use when
                        filtering common variants)
      --max-maf-filter 0.001 (as float)
                        Filter based on the maximum maf allowed (use when
                        filtering rare variants)
      --geno-filter 0.0 (as float)
                        Filter out variants with missing call rate higher that
                        this value
      --hwe-filter 0.0 (as float)
                        Filter according to Hardy Weiberg Equilibrium
      --mind-filter 0.0 (as float)
                        Fitler out samples with missing rate higher than this
                        value
  get_snps_2:           Merge all of the common_var.snplist into a single file
                        and all the rare_var.snplist into another single file
  get_snps_3:           Search for common or rare variants in bimfile and
                        generate annovar input file
  annovar_1:            Create annovar input file
  annovar_2:            Annotate vcf file using ANNOVAR
    Workflow Options:
      --humandb VAL (as path, required)
                        humandb path for ANNOVAR
      --ukbb VAL (as path, required)
      --x-ref  path(f&#34;{ukbb}/mart_export_2019_LOFtools3.txt&#34;)

                        add xreffile to option without -exonicsplicing
                        mart_export_2019_LOFtools3.txt #xreffile latest option
                        -&gt; Phenotype description,HGNC symbol,MIM morbid descript
                        ion,CGD_CONDITION,CGD_inh,CGD_man,CGD_comm,LOF_tools
      --protocol refGene refGeneWithVer knownGene ensGene phastConsElements30way encRegTfbsClustered gwasCatalog gnomad211_genome gnomad211_exome gme kaviar_20150923 abraom avsnp150 dbnsfp41a dbscsnv11 regsnpintron clinvar_20200316 gene4denovo201907 (as list)
                        Annovar protocol
      --operation g g g gx r r r f f f f f f f f f f f (as list)
                        Annovar operation
      --arg &#34;-splicing 12 -exonicsplicing&#34; &#34;-splicing 30&#34; &#34;-splicing 12 -exonicsplicing&#34; &#34;-splicing 12&#34;               (as list)
                        Annovar args
</pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Illustration-with-minimal-working-example">Illustration with minimal working example<a class="anchor-link" href="#Illustration-with-minimal-working-example">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p><strong>Scenario 3:</strong> On Yale's cluster, here modify humandb and ukbb paths to match the location of the databases needed by annovar to function</p>

<pre><code>sos run ~/project/bioworkflows/variant-annotation/annovar.ipynb annovar \
    --cwd output \
    --bim_name ukb23156_c22.merged.filtered.bim \
    --humandb /gpfs/ysm/datasets/db/annovar/humandb \
    --ukbb /gpfs/gibbs/pi/dewan/data/UKBiobank \
    --job_size 1 \
    --name_prefix mwe_chr22 \
    --container_annovar /gpfs/gibbs/pi/dewan/data/UKBiobank/annovar.sif</code></pre>
<p>On Columbia's cluster</p>

<pre><code>sos run ~/project/bioworkflows/variant-annotation/annovar.ipynb annovar \
    --cwd output \
    --bim_name /mnt/mfs/statgen/UKBiobank/data/exome_files/project_VCF/plink_files/ukb23156_c22.merged.filtered.bim \
    --humandb /mnt/mfs/statgen/isabelle/REF/humandb  \
    --ukbb /mnt/mfs/statgen/isabelle/REF/humandb \
    --job_size 1 \
    --name_prefix mwe_chr22 \
    --container_annovar /mnt/mfs/statgen/containers/gatk4-annovar.sif</code></pre>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># the output directory for generated files</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># Specific number of threads to use</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">2</span>
<span class="c1"># For cluster jobs, number commands to run per job</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">1</span>
<span class="c1"># Human genome build</span>
<span class="kn">parameter:</span> <span class="n">build</span> <span class="o">=</span> <span class="s1">&#39;hg38&#39;</span>
<span class="c1"># Name for the merged bimfiles</span>
<span class="kn">parameter:</span> <span class="n">bim_name</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># Prefix for the name based on common/rare variant filtering</span>
<span class="kn">parameter:</span> <span class="n">name_prefix</span> <span class="o">=</span> <span class="nb">str</span>
<span class="c1"># Load annovar module from cluster</span>
<span class="kn">parameter:</span> <span class="n">annovar_module</span> <span class="o">=</span> <span class="s1">&#39;&#39;&#39;</span>
<span class="s1">module load ANNOVAR/2020Jun08-foss-2018b-Perl-5.28.0</span>
<span class="s1">echo &quot;Module annovar loaded&quot;</span>
<span class="s1">{cmd}</span>
<span class="s1">&#39;&#39;&#39;</span>
<span class="c1"># Software container option</span>
<span class="kn">parameter:</span> <span class="n">container_annovar</span> <span class="o">=</span> <span class="s1">&#39;gaow/gatk4-annovar&#39;</span>
<span class="kn">parameter:</span> <span class="n">container_lmm</span> <span class="o">=</span> <span class="s1">&#39;statisticalgenetics/lmm:2.0&#39;</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Format-file-for-plink-.bim">Format file for plink .bim<a class="anchor-link" href="#Format-file-for-plink-.bim">&#182;</a></h3><p>A text file with no header line, and one line per variant with the following six fields:</p>
<ol>
<li>Chromosome code (either an integer, or 'X'/'Y'/'XY'/'MT'; '0' indicates unknown) or name</li>
<li>Variant identifier</li>
<li>Position in morgans or centimorgans (safe to use dummy value of '0')</li>
<li>Base-pair coordinate (1-based; limited to 231-2)</li>
<li>Allele 1 (corresponding to clear bits in .bed; usually minor)</li>
<li>Allele 2 (corresponding to set bits in .bed; usually major)</li>
</ol>
<p>In the bim file the second column e.g <code>1:930232:C:T</code> contains the alleles in ref/alt mode</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># Merge all the bimfiles into a single file to use later with awk</span>
<span class="c1"># Only need to run this cell once</span>
<span class="p">[</span><span class="n">bim_merge</span><span class="p">:</span> <span class="kp">provides</span> <span class="o">=</span> <span class="n">bim_name</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">bimfiles</span> <span class="o">=</span> <span class="n">paths</span>
<span class="kn">input:</span> <span class="n">bimfiles</span>
<span class="kn">output:</span> <span class="n">bim_name</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;10h&#39;</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;10G&#39;</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{step_name}_{_output:bn}&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.stdout&#39;</span> 
      <span class="n">cat</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">}</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">}</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># Get a list of common SNPs above (--maf) or rare SNPs below (--max-maf) certain MAF</span>
<span class="p">[</span><span class="n">get_snps_1</span><span class="p">]</span>
<span class="c1"># bed files plink format</span>
<span class="kn">parameter:</span> <span class="n">bfiles</span> <span class="o">=</span> <span class="n">paths</span>
<span class="c1"># Filter based on minor allele frequency (use when filtering common variants)</span>
<span class="kn">parameter:</span> <span class="n">maf_filter</span> <span class="o">=</span> <span class="mf">0.0</span>
<span class="c1"># Filter based on the maximum maf allowed (use when filtering rare variants)</span>
<span class="kn">parameter:</span> <span class="n">max_maf_filter</span> <span class="o">=</span> <span class="mf">0.001</span>
<span class="c1"># Filter out variants with missing call rate higher that this value</span>
<span class="kn">parameter:</span> <span class="n">geno_filter</span> <span class="o">=</span> <span class="mf">0.0</span>
<span class="c1"># Filter according to Hardy Weiberg Equilibrium</span>
<span class="kn">parameter:</span> <span class="n">hwe_filter</span> <span class="o">=</span> <span class="mf">0.0</span>
<span class="c1"># Fitler out samples with missing rate higher than this value</span>
<span class="kn">parameter:</span> <span class="n">mind_filter</span> <span class="o">=</span> <span class="mf">0.0</span>
<span class="kn">input:</span> <span class="n">bfiles</span><span class="p">,</span> <span class="n">group_by</span><span class="o">=</span><span class="mi">1</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{cwd}/cache/{_input:bn}.{name_prefix}.snplist&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;10h&#39;</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;30G&#39;</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{step_name}_{_output:bn}&#39;</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container_lmm</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.stdout&#39;</span> 
    <span class="n">plink2</span> \
      <span class="o">--</span><span class="n">bfile</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">n</span><span class="p">}</span>\
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--maf </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">maf_filter</span><span class="p">)</span> <span class="k">if</span> <span class="n">maf_filter</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--max-maf </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">max_maf_filter</span><span class="p">)</span> <span class="k">if</span> <span class="n">max_maf_filter</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--geno </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">geno_filter</span><span class="p">)</span> <span class="k">if</span> <span class="n">geno_filter</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--hwe </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">hwe_filter</span><span class="p">)</span> <span class="k">if</span> <span class="n">hwe_filter</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--mind </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">mind_filter</span><span class="p">)</span> <span class="k">if</span> <span class="n">mind_filter</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">write</span><span class="o">-</span><span class="n">snplist</span> <span class="o">--</span><span class="n">no</span><span class="o">-</span><span class="nb">id</span><span class="o">-</span><span class="n">header</span>\
      <span class="o">--</span><span class="n">freq</span> \
      <span class="o">--</span><span class="n">threads</span> <span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># Merge all of the common_var.snplist into a single file and all the rare_var.snplist into another single file</span>
<span class="p">[</span><span class="n">get_snps_2</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">group_by</span><span class="o">=</span><span class="s1">&#39;all&#39;</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{cwd}/cache/{name_prefix}.snplist&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;10h&#39;</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;10G&#39;</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{step_name}_{_output:bn}&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[0]:n}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.stdout&#39;</span> 
      <span class="n">cat</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">}</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">}</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># Search for common or rare variants in bimfile and generate annovar input file</span>
<span class="p">[</span><span class="n">get_snps_3</span><span class="p">]</span>
<span class="kn">depends:</span> <span class="n">bim_name</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{cwd}/{_input:bn}.avinput&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;10h&#39;</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;10G&#39;</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{step_name}_{_output:bn}&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.stdout&#39;</span> 
    <span class="n">awk</span> <span class="o">-</span><span class="n">F</span><span class="s2">&quot; &quot;</span> <span class="s1">&#39;FNR==NR {lines[$1]; next} $2 in lines &#39;</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">}</span> <span class="err">$</span><span class="p">{</span><span class="n">bim_name</span><span class="p">}</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">tmp</span>
    <span class="n">awk</span> <span class="s1">&#39;{if ($2 ~ /D/) {print $1, $4, $4 + (length ($6) - length ($5)), $6, $5 } else {print $1, $4, $4, $6, $5 }}&#39;</span>  <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">tmp</span> <span class="o">&gt;</span>  <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">}</span>
    <span class="c1"># remove temporary files</span>
    <span class="n">rm</span> <span class="o">-</span><span class="n">f</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">tmp</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Annovar-details">Annovar details<a class="anchor-link" href="#Annovar-details">&#182;</a></h2><p>For a list of available <a href="https://hgdownload.soe.ucsc.edu/goldenPath/hg38/database/">databases</a></p>
<p>On Farnam's Yale HPC there is a folder for shared databases
<code>/gpfs/ysm/datasets/db/annovar/humandb</code></p>
<p>and a folder for the x_ref database <code>/gpfs/gibbs/pi/dewan/data/UKBiobank/mart_export_2019_LOFtools3.txt</code></p>
<p>On Yale's cluster there folder for shared databases for build hg19 is under Isabelle's folder
<code>/mnt/mfs/statgen/isabelle/REF/humandb</code></p>
<p>and the x_ref database is under that same folder <code>/mnt/mfs/statgen/isabelle/REF/humandb</code></p>
<h3 id="Important-note">Important note<a class="anchor-link" href="#Important-note">&#182;</a></h3><p>Please make sure you are using the correct build for your annotations UKBB exome data for 200K individuals need hg38 build</p>
<h3 id="Format-file-for-annovar-input">Format file for annovar input<a class="anchor-link" href="#Format-file-for-annovar-input">&#182;</a></h3><p>On each line, the first five space- or tab- delimited columns represent</p>
<ol>
<li>chromosome </li>
<li>start position </li>
<li>end position </li>
<li>the reference nucleotides</li>
<li>the observed nucleotides</li>
</ol>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># Create annovar input file</span>
<span class="p">[</span><span class="n">annovar_1</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">bim_name</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{cwd}/{_input:bn}.{build}.avinput&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;10h&#39;</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;10G&#39;</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:bn}&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.err&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.out&#39;</span> 
    <span class="n">awk</span> <span class="s1">&#39;{if ($2 ~ /D/) {print $1, $4, $4 + (length ($6) - length ($5)), $6, $5, $2} else {print $1, $4, $4, $6, $5, $2}}&#39;</span>  <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">}</span> <span class="o">&gt;</span>  <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">}</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># Annotate vcf file using ANNOVAR</span>
<span class="p">[</span><span class="n">annovar_2</span><span class="p">]</span>
<span class="c1"># humandb path for ANNOVAR</span>
<span class="kn">parameter:</span> <span class="n">humandb</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">parameter:</span> <span class="n">ukbb</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1">#add xreffile to option without -exonicsplicing</span>
<span class="c1">#mart_export_2019_LOFtools3.txt #xreffile latest option -&gt; Phenotype description,HGNC symbol,MIM morbid description,CGD_CONDITION,CGD_inh,CGD_man,CGD_comm,LOF_tools</span>
<span class="kn">parameter:</span> <span class="n">x_ref</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;{ukbb}/mart_export_2019_LOFtools3.txt&quot;</span><span class="p">)</span>
<span class="c1"># Annovar protocol</span>
<span class="kn">parameter:</span> <span class="n">protocol</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;refGene&#39;</span><span class="p">,</span> <span class="s1">&#39;refGeneWithVer&#39;</span><span class="p">,</span> <span class="s1">&#39;knownGene&#39;</span><span class="p">,</span> <span class="s1">&#39;ensGene&#39;</span><span class="p">,</span> <span class="s1">&#39;phastConsElements30way&#39;</span><span class="p">,</span> <span class="s1">&#39;encRegTfbsClustered&#39;</span><span class="p">,</span> <span class="s1">&#39;gwasCatalog&#39;</span><span class="p">,</span> <span class="s1">&#39;gnomad211_genome&#39;</span><span class="p">,</span> <span class="s1">&#39;gnomad211_exome&#39;</span><span class="p">,</span> <span class="s1">&#39;gme&#39;</span><span class="p">,</span> <span class="s1">&#39;kaviar_20150923&#39;</span><span class="p">,</span> <span class="s1">&#39;abraom&#39;</span><span class="p">,</span> <span class="s1">&#39;avsnp150&#39;</span><span class="p">,</span> <span class="s1">&#39;dbnsfp41a&#39;</span><span class="p">,</span> <span class="s1">&#39;dbscsnv11&#39;</span><span class="p">,</span> <span class="s1">&#39;regsnpintron&#39;</span><span class="p">,</span> <span class="s1">&#39;clinvar_20200316&#39;</span><span class="p">,</span> <span class="s1">&#39;gene4denovo201907&#39;</span><span class="p">]</span>
<span class="c1"># Annovar operation</span>
<span class="kn">parameter:</span> <span class="n">operation</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;g&#39;</span><span class="p">,</span> <span class="s1">&#39;g&#39;</span><span class="p">,</span> <span class="s1">&#39;g&#39;</span><span class="p">,</span> <span class="s1">&#39;gx&#39;</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="s1">&#39;f&#39;</span><span class="p">,</span> <span class="s1">&#39;f&#39;</span><span class="p">,</span> <span class="s1">&#39;f&#39;</span><span class="p">,</span> <span class="s1">&#39;f&#39;</span><span class="p">,</span> <span class="s1">&#39;f&#39;</span><span class="p">,</span> <span class="s1">&#39;f&#39;</span><span class="p">,</span> <span class="s1">&#39;f&#39;</span><span class="p">,</span> <span class="s1">&#39;f&#39;</span><span class="p">,</span> <span class="s1">&#39;f&#39;</span><span class="p">,</span> <span class="s1">&#39;f&#39;</span><span class="p">,</span> <span class="s1">&#39;f&#39;</span><span class="p">]</span>
<span class="c1"># Annovar args</span>
<span class="kn">parameter:</span> <span class="n">arg</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;&quot;-splicing 12 -exonicsplicing&quot;&#39;</span><span class="p">,</span> <span class="s1">&#39;&quot;-splicing 30&quot;&#39;</span><span class="p">,</span> <span class="s1">&#39;&quot;-splicing 12 -exonicsplicing&quot;&#39;</span><span class="p">,</span> <span class="s1">&#39;&quot;-splicing 12&quot;&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">]</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{cwd}/{_input:bn}.{build}_multianno.csv&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;60h&#39;</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;48G&#39;</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:bn}&#39;</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container_annovar</span><span class="p">,</span> <span class="n">volumes</span><span class="o">=</span><span class="p">[</span><span class="n">f</span><span class="s1">&#39;{humandb:a}:{humandb:a}&#39;</span><span class="p">,</span> <span class="n">f</span><span class="s1">&#39;{x_ref:ad}:{x_ref:ad}&#39;</span><span class="p">],</span> <span class="n">expand</span><span class="o">=</span><span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span><span class="o">=</span><span class="n">f</span><span class="s1">&#39;{_output:n}.err&#39;</span><span class="p">,</span> <span class="n">stdout</span><span class="o">=</span><span class="n">f</span><span class="s1">&#39;{_output:n}.out&#39;</span>
    <span class="c1">#do not add -intronhgvs as option -&gt; writes cDNA variants as HGVS but creates issues (+2 splice site reported only)</span>
    <span class="c1">#-nastring . can only be . for VCF files</span>
    <span class="c1">#regsnpintron might cause shifted lines (be carefull using)</span>
    <span class="n">table_annovar</span><span class="o">.</span><span class="n">pl</span> \
        <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">}</span> \
        <span class="err">$</span><span class="p">{</span><span class="n">humandb</span><span class="p">}</span> \
        <span class="o">-</span><span class="n">buildver</span> <span class="err">$</span><span class="p">{</span><span class="n">build</span><span class="p">}</span> \
        <span class="o">-</span><span class="n">out</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">nn</span><span class="p">}</span>\
        <span class="o">-</span><span class="n">remove</span> \
        <span class="o">-</span><span class="n">polish</span> \
        <span class="o">-</span><span class="n">nastring</span> <span class="o">.</span> \
        <span class="o">-</span><span class="n">protocol</span> <span class="err">$</span><span class="p">{</span><span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">protocol</span><span class="p">)}</span> \
        <span class="o">-</span><span class="n">operation</span> <span class="err">$</span><span class="p">{</span><span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">operation</span><span class="p">)}</span> \
        <span class="o">-</span><span class="n">arg</span> <span class="err">$</span><span class="p">{</span><span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">arg</span><span class="p">)}</span> \
        <span class="o">-</span><span class="n">csvout</span> \
        <span class="o">-</span><span class="n">xreffile</span> <span class="err">$</span><span class="p">{</span><span class="n">x_ref</span><span class="p">}</span>
</pre></div>

    </div>
</div>
</div>

</div>
<hr>
&copy 2020-current Gao Wang and others, Department of Neurology at Columbia University
<p><small>Exported from <a href="http://github.com/cumc/bioworkflows/blob/9ff8d6758674b2dd00fa44908723d589a81e8d8a/variant-annotation/annovar.ipynb"><code>variant-annotation/annovar.ipynb</code></a> committed by dmc2245 on Thu Apr 29 20:28:29 2021 <a href="http://github.com/cumc/bioworkflows/commit/9ff8d6758674b2dd00fa44908723d589a81e8d8a">revision 5, 9ff8d67</a> <a href="https://stephenslab.github.io/ipynb-website/notes.html#Note-about-commit-ids"><span class="fa fa-question-circle"></span></a></small></p>
</div>
</div>
</body>
</html>
