{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "# Multivariate prediction workflow\n",
    "\n",
    "This notebook applies mtlasso on data analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Input\n",
    "\n",
    "RDS format of a list of objects, in which case you can specify the names of objects corresponding to the quantities `X`, `Y`, etc.\n",
    "\n",
    "**FIXME: need to document the input data structure**\n",
    "**Also for prior files should they be stored similarly with data for each fold as `fold_??` tables**?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Output\n",
    "\n",
    "For each analysis unit we output:\n",
    "\n",
    "1. Analysis results in RDS format\n",
    "2. Default visualization plots\n",
    "\n",
    "**FIXME: at this point we dont have output figure yet**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Analysis examples\n",
    "\n",
    "```\n",
    "sos run /project/mstephens/fmorgante/bioworkflows/multivariate-prediction/mtlasso.ipynb mtlasso \\\n",
    "    --analysis-units ../data/gtex-v8-manifest-2ormore-tissues-nopath-nosuffix-test.txt \\\n",
    "    --data-dir ../data/cis_eqtl_analysis_ready  \\\n",
    "    --data-suffix GTEx_V8.rds \\\n",
    "    --name fold_1 \\\n",
    "    --wd ../output/gtex_mr_mash_analysis \\\n",
    "    --sample-partition ../data/gtex-v8-ids-folds.txt \\\n",
    "    --fold 1 \\\n",
    "    --mtlasso-script /project/mstephens/fmorgante/mr_mash_test/code/fit_mtlasso_missing_Y.py \\\n",
    "    --conda-env py38dsc \\\n",
    "    -c midway2.yml -q midway2\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "kernel": "SoS",
    "tags": []
   },
   "outputs": [],
   "source": [
    "[global]\n",
    "# Path to data directory\n",
    "parameter: data_dir = path\n",
    "# data file suffix\n",
    "parameter: data_suffix = str\n",
    "# Path to work directory where output locates\n",
    "parameter: wd = path(\"./output\")\n",
    "# An identifier for your run of analysis\n",
    "parameter: name = str"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[mtlasso]\n",
    "# single column file each line is the data filename\n",
    "parameter: analysis_units = path\n",
    "# Path to summary statistics directory\n",
    "parameter: sample_partition = path\n",
    "parameter: conda_env = str\n",
    "parameter: mtlasso_script = path\n",
    "parameter: fold = 1\n",
    "parameter: imiss = 0.05\n",
    "parameter: maf = 0.05\n",
    "parameter: var_cutoff = 0.05\n",
    "parameter: n_nonmiss_Y = 100\n",
    "parameter: standardize = \"TRUE\"\n",
    "parameter: max_iter = 5000\n",
    "parameter: verbose = \"FALSE\"\n",
    "parameter: nfolds = 5\n",
    "parameter: B_init = \"NULL\"\n",
    "parameter: grid_limits = \"NULL\"\n",
    "parameter: grid_length = 10\n",
    "# Only analyze `cis` variants -- cis = N means using N variants around the center column of X matrix  \n",
    "parameter: cis = 'NULL'\n",
    "parameter: seed = 999\n",
    "regions = [x.strip() for x in open(analysis_units).readlines() if x.strip() and not x.strip().startswith('#')]\n",
    "genes = [f\"{data_dir:a}/{x}.{data_suffix}\" for x in regions if path(f\"{data_dir:a}/{x}.{data_suffix}\").exists()]\n",
    "input: genes, group_by = 1\n",
    "output: f'{wd:a}/mtlasso/fold_{fold}/{_input:bn}_{name}_mtlasso.rds'\n",
    "task: trunk_workers = 4, trunk_size = 36, walltime = '4h', mem = '10G', cores = 1, tags = f'{step_name}_{_output[0]:bn}'\n",
    "R: expand = '${ }', stdout = f\"{_output[0]:n}.stdout\", stderr = f\"{_output[0]:n}.stderr\"\n",
    "    options(stringsAsFactors = FALSE)\n",
    "    set.seed(${seed})\n",
    "    library(reticulate)\n",
    "\n",
    "    ###Set some parameter variables (These should be set in the SoS script)\n",
    "    fold <- ${fold}\n",
    "    missing_rate_cutoff <- ${imiss}\n",
    "    maf_cutoff <- ${maf}\n",
    "    var_cutoff <- ${var_cutoff}\n",
    "    n_nonmiss_Y <- ${n_nonmiss_Y}\n",
    "    standardize <- ${standardize}\n",
    "    verbose <- ${verbose}\n",
    "    nfolds <- ${nfolds}\n",
    "    B_init <- ${B_init}\n",
    "    grid_limits <- ${grid_limits}\n",
    "    grid_length <- ${grid_length}\n",
    "\n",
    "\n",
    "  ###\n",
    "    # Utility functions\n",
    "    ###\n",
    "\n",
    "    ###Functions to compute MAF, missing genotype rate, impute missing, and filter X accordingly \n",
    "    compute_maf <- function(geno){\n",
    "      f <- mean(geno,na.rm = TRUE)/2\n",
    "      return(min(f, 1-f))\n",
    "    }\n",
    "\n",
    "    compute_missing <- function(geno){\n",
    "      miss <- sum(is.na(geno))/length(geno)\n",
    "      return(miss)\n",
    "    }\n",
    "\n",
    "    compute_non_missing_y <- function(y){\n",
    "      nonmiss <- sum(!is.na(y))\n",
    "      return(nonmiss)\n",
    "    }\n",
    "\n",
    "    mean_impute <- function(geno){\n",
    "      f <- apply(geno, 2, function(x) mean(x,na.rm = TRUE))\n",
    "      for (i in 1:length(f)) geno[,i][which(is.na(geno[,i]))] <- f[i]\n",
    "      return(geno)\n",
    "    }\n",
    "\n",
    "    filter_X <- function(X, missing_rate_thresh, maf_thresh, var_thresh) {\n",
    "      rm_col <- which(apply(X, 2, compute_missing) > missing_rate_thresh)\n",
    "      if (length(rm_col)) X <- X[, -rm_col]\n",
    "      rm_col <- which(apply(X, 2, compute_maf) < maf_thresh)\n",
    "      if (length(rm_col)) X <- X[, -rm_col]\n",
    "      X <- mean_impute(X)\n",
    "      rm_col <- which(matrixStats::colVars(X) < var_thresh)\n",
    "      if (length(rm_col)) X <- X[, -rm_col]\n",
    "      return(X)\n",
    "    }\n",
    "\n",
    "    filter_Y <- function(Y, n_nonmiss){\n",
    "      rm_col <- which(apply(Y, 2, compute_non_missing_y) < n_nonmiss)\n",
    "      if (length(rm_col)) Y <- Y[, -rm_col]\n",
    "      return(Y)\n",
    "    }\n",
    "  \n",
    "    ###Split the data in training and test\n",
    "    split_data <- function(X, Y, gtex_ids_folds, fold){\n",
    "      test_ids <- gtex_ids_folds[which(gtex_ids_folds$fold == fold), \"id\"]\n",
    "      Xtrain <- X[!(rownames(X) %in% test_ids), ]\n",
    "      Ytrain <- Y[!(rownames(Y) %in% test_ids), ]\n",
    "      Xtest <- X[rownames(X) %in% test_ids, ]\n",
    "      Ytest <- Y[rownames(Y) %in% test_ids, ]\n",
    "      \n",
    "      return(list(Xtrain=Xtrain, Ytrain=Ytrain, Xtest=Xtest, Ytest=Ytest))\n",
    "    }\n",
    "  \n",
    "    predict_general <- function(B, intercept, newx){\n",
    "      if(is.matrix(intercept))\n",
    "        intercept <- drop(intercept)\n",
    "      return(addtocols(newx %*% B, intercept))\n",
    "    }\n",
    "  \n",
    "    addtocols <- function (A, b){\n",
    "      t(t(A) + b)\n",
    "    }\n",
    "      \n",
    "    ###\n",
    "    # mtlasso code\n",
    "    ###\n",
    "  \n",
    "    ###Read in the data\n",
    "    dat <- readRDS(${_input:r})   \n",
    "    gtex_ids_folds <- read.table(${sample_partition:r}, header=TRUE, sep=\"\\t\")\n",
    "  \n",
    "    ###Extract sumstats and only for specified fold\n",
    "    fold_name <- paste0(\"fold_\", fold)\n",
    "\n",
    "    ###Extract and filter Y and X. NB the workflow will stop if filtered Y has < 2 tissues.\n",
    "    Y <- filter_Y(dat$y_res, n_nonmiss_Y)\n",
    "    if(is.matrix(Y)){\n",
    "      X <- filter_X(dat$X, missing_rate_cutoff, maf_cutoff, var_cutoff)\n",
    "  \n",
    "      ###Split the data in training and test sets\n",
    "      dat_split <- split_data(X, Y, gtex_ids_folds, fold)\n",
    "      Xtrain <- dat_split$Xtrain\n",
    "      Ytrain <- dat_split$Ytrain\n",
    "      Xtest <- dat_split$Xtest\n",
    "      Ytest <- dat_split$Ytest\n",
    "      rm(dat_split)\n",
    "      \n",
    "      time1 <- proc.time()\n",
    "\n",
    "      ###Fit mtlasso\n",
    "      use_condaenv(${conda_env:r})\n",
    "      source_python(${mtlasso_script:r})\n",
    "  \n",
    "      fit_mtlasso <- tryCatch({fit_sparse_multi_task_lasso_missing_Y(X=Xtrain, Y=Ytrain, standardize=standardize, \n",
    "                                                                     max_iter=max_iter, verbose=verbose, \n",
    "                                                                     nfolds=nfolds, grid_length=grid_length,\n",
    "                                                                     B_init=B_init, grid_limits=grid_limits)\n",
    "                              },\n",
    "                             error=function(e) {\n",
    "                                  message(\"Original mtlasso error message:\")\n",
    "                                  message(e)\n",
    "                                  return(NULL)\n",
    "                              },\n",
    "                             warning=function(w) {\n",
    "                                  message(\"Original mtlasso warning message:\")\n",
    "                                  message(w)\n",
    "                                  return(NULL)\n",
    "                              })\n",
    "\n",
    "    \n",
    "      if(!is.null(fit_mtlasso)){\n",
    "        time2 <- proc.time()\n",
    "        elapsed_time <- time2[\"elapsed\"] - time1[\"elapsed\"]\n",
    "        ###Make predictions\n",
    "        Yhat_test <- predict_general(B=fit_mtlasso$Bhat, intercept=fit_mtlasso$B0hat, newx=Xtest)\n",
    "  \n",
    "        ###Save results\n",
    "        resu <- list(Ytest=Ytest, Yhat_test=Yhat_test, elapsed_time=elapsed_time, model=fit_mtlasso)\n",
    "\n",
    "        saveRDS(resu, ${_output[0]:r})\n",
    "      } else {\n",
    "        saveRDS(NULL, ${_output[0]:r})\n",
    "      }\n",
    "    } else {\n",
    "      saveRDS(NULL, ${_output[0]:r})\n",
    "      message(\"Filtered Y has fewer than 2 tissues. Gene will not be analyzed.\")\n",
    "    }"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  },
  "sos": {
   "kernels": [
    [
     "SoS",
     "sos",
     "",
     ""
    ]
   ],
   "version": "0.22.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
