{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "# Principal component analysis\n",
    "\n",
    "## Aim\n",
    "\n",
    "The intention of this notebook is to generate the PCA analysis and plots.\n",
    "\n",
    "Steps to generate a PCA include removing related individuals, pruning variants in linkage disequilibrium (LD), and excluding outlier samples that can suggest poor genotyping quality or distant relatedness (also restrict to individuals of homogeneous ancestry).\n",
    "\n",
    "Pitfalls\n",
    "1. Some of the PCs may capture LD structure rather than population structure (decrease in power to detect associations in these regions of high LD)\n",
    "2. When projecting a new study dataset to the PCA space computed from a reference dataset: projected PCs are shrunk toward 0 in the new dataset \n",
    "3.  PC scores may capture outliers that are due to family structure, population structure or other reasons; it might be beneficial to detect and remove these individuals to maximize the population structure captured by PCA (in the case of removing a few outliers) or to restrict analyses to genetically homogeneous samples\n",
    "\n",
    "## Input\n",
    "\n",
    "1. bfile: the genotype array to calculate individual relationship\n",
    "2. genoFile in plink format (`.bed`, `.bim` and `.fam`) - could be different from bfile in the case of using for e.g. exome data/ imputed genotype data\n",
    "3. relatedness file (you can calculate relatedness using plink or king)\n",
    "\n",
    "## Output\n",
    "\n",
    "Form the kinship analysis\n",
    "1. Kinship table\n",
    "\n",
    "From the PCA analysis\n",
    "1. values\n",
    "2. vectors\n",
    "3. projection (PC's)\n",
    "4. loadings\n",
    "5. scale\n",
    "6. mahalanobis distances for outlier removal\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## General workflow\n",
    "\n",
    "1. Estimate relatedness of the individuals in the sample by using king (as implemented in Plink)\n",
    "2. Select specific SNPs and samples using Plink (QC maf>1%, geno=0.1, mind=0.1 and hwe=5e-08) \n",
    "3. SNPs thining by doing LD-pruning (window=50, shift=10, r2=0.1 are the defaults) and remove related individuals prior to PCA calculation\n",
    "4. First PCA run (using only unrelated individuals)\n",
    "5. Calculate mahalanobis distance and create outlier removal file (criteria 6SD from the mean)\n",
    "6. Re-calculate PC's without outliers\n",
    "7. Project related samples "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "# Command interface"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: sos run PCA.ipynb [workflow_name | -t targets] [options] [workflow_options]\n",
      "  workflow_name:        Single or combined workflows defined in this script\n",
      "  targets:              One or more targets to generate\n",
      "  options:              Single-hyphen sos parameters (see \"sos run -h\" for details)\n",
      "  workflow_options:     Double-hyphen workflow-specific parameters\n",
      "\n",
      "Workflows:\n",
      "  king\n",
      "  filter\n",
      "  flashpca\n",
      "\n",
      "Global Workflow Options:\n",
      "  --cwd VAL (as path, required)\n",
      "                        the output directory for generated files\n",
      "  --bfile VAL (as path, required)\n",
      "                        Path to genotype array file\n",
      "  --genoFile  paths\n",
      "\n",
      "                        Plink binary files\n",
      "  --phenoFile VAL (as path, required)\n",
      "                        The phenotypic file\n",
      "  --job-size 1 (as int)\n",
      "                        For cluster jobs, number commands to run per job\n",
      "  --numThreads 1 (as int)\n",
      "                        Number of threads\n",
      "  --eigensoft-module '\\nmodule load EIGENSOFT/7.2.1-foss-2018b\\necho \"Module Eigensoft v.7.2.1 loaded\"\\n{cmd}\\n'\n",
      "                        Merge data (FIXME:merge step) parameter: merge = True\n",
      "                        Load Eigensoft module from cluster\n",
      "  --container-lmm 'statisticalgenetics/lmm:1.8'\n",
      "                        Software container option\n",
      "\n",
      "Sections\n",
      "  king:                 Inference of relationships in the sample to remove\n",
      "                        closely related individuals\n",
      "    Workflow Options:\n",
      "      --kinship 0.0884 (as float)\n",
      "                        Filter based on kinship coefficient higher than a number\n",
      "                        (e.g first degree 0.25, second degree 0.125, third\n",
      "                        degree 0.0625)\n",
      "  filter_1:             Filter SNPs with MAF>1% for PCA analysis, select\n",
      "                        individuals and merge bed into one file\n",
      "    Workflow Options:\n",
      "      --keep-samples VAL (as path, required)\n",
      "                        The path to the file that contains the list of samples\n",
      "                        to keep (format FID, IID)\n",
      "      --maf-filter 0.01 (as float)\n",
      "                        MAF filter to use\n",
      "      --geno-filter 0.01 (as float)\n",
      "                        Maximum missingess per-variant\n",
      "      --mind-filter 0.02 (as float)\n",
      "                        Maximum missingness per-sample\n",
      "      --hwe-filter 5e-08 (as float)\n",
      "                        HWE filter\n",
      "  filter_2:             Merge all the .bed files into one bed file\n",
      "  filter_3:             LD prunning and remove related individuals (both ind of\n",
      "                        a pair)\n",
      "    Workflow Options:\n",
      "      --window 50 (as int)\n",
      "                        Window size\n",
      "      --shift 10 (as int)\n",
      "                        Shift window every 10 snps\n",
      "      --r2 0.1 (as float)\n",
      "  flashpca_1:           Run PCA analysis using flashpca\n",
      "    Workflow Options:\n",
      "      --k VAL (as int, required)\n",
      "                        Number of Principal Components to output. Default is 10\n",
      "      --trait-name VAL (as str, required)\n",
      "                        Name of the trait in the phenoFile (format FID, IID,\n",
      "                        father, mother,trait )\n",
      "      --stand binom2\n",
      "                        How to standardize X before PCA\n",
      "  flashpca_2:           Filter out outliers from bed file and filter related\n",
      "                        individuals to project them onto PCA\n",
      "  flashpca_3:           Re-do PCA without outliers\n",
      "    Workflow Options:\n",
      "      --k VAL (as int, required)\n",
      "                        Number of Principal Components to output. Default is 10\n",
      "      --trait-name VAL (as str, required)\n",
      "                        Name of the trait in the phenoFile (format FID, IID,\n",
      "                        father, mother,trait )\n",
      "      --stand binom2\n",
      "                        How to standardize X before PCA\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sos run PCA.ipynb -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## PCA analysis pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[global]\n",
    "# the output directory for generated files\n",
    "parameter: cwd = path\n",
    "# Path to genotype array file\n",
    "parameter: bfile =path\n",
    "# Plink binary files\n",
    "parameter: genoFile = paths\n",
    "# The phenotypic file\n",
    "parameter: phenoFile = path\n",
    "# For cluster jobs, number commands to run per job\n",
    "parameter: job_size = 1\n",
    "# Number of threads\n",
    "parameter: numThreads = 1\n",
    "# Number of Principal Components to output. Default is 10\n",
    "parameter: k = int\n",
    "# Name of the trait in the phenoFile (format FID, IID, father, mother,trait )\n",
    "parameter: trait_name = str\n",
    "# How to standardize X before PCA\n",
    "parameter: stand = \"binom2\"\n",
    "# Merge data (FIXME:merge step)\n",
    "#parameter: merge = True\n",
    "# Software container option\n",
    "parameter: container_lmm = 'statisticalgenetics/lmm:1.8'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# Inference of relationships in the sample to remove closely related individuals\n",
    "[king]\n",
    "# Filter based on kinship coefficient higher than a number (e.g first degree 0.25, second degree 0.125, third degree 0.0625)\n",
    "parameter: kinship = 0.0884\n",
    "# bfile corresponds to the genotypic array file\n",
    "input: bfile\n",
    "output: f'{cwd}/{_input:bn}.kin0', related_samples = f'{cwd}/{_input:bn}.related_id'\n",
    "task: trunk_workers = 1, walltime = '10h', mem = '30G', cores = numThreads, tags = f'{step_name}_{_output[0]:bn}'\n",
    "bash: container=container_lmm, expand= \"${ }\", stderr = f'{_output[0]}.stderr', stdout = f'{_output[0]}.stdout'\n",
    "    plink2 \\\n",
    "      --bfile ${_input:n} \\\n",
    "      --make-king-table \\\n",
    "      --threads ${numThreads} \\\n",
    "      --out ${_output[0]:n} \n",
    "      \n",
    "R:  container=container_lmm, expand= \"${ }\", stderr = f'{_output[1]}.stderr', stdout = f'{_output[1]}.stdout'\n",
    "    library(dplyr)\n",
    "    kin0 <- read.table(${_output[0]:r}, header=F)\n",
    "    colnames(kin0) <- c(\"FID1\",\"ID1\",\"FID2\",\"ID2\",\"NSNP\",\"HETHET\",\"IBS0\",\"KINSHIP\")\n",
    "    rel <- kin0 %>%\n",
    "        filter(KINSHIP >= ${kinship})\n",
    "    cat(\"There are\", nrow(rel),\"related individuals using a kinship threshold of ${kinship}\\n\")\n",
    "    IID <- sort(unique(unlist(rel[, c(\"ID1\", \"ID2\")])))\n",
    "    df <- data.frame(IID)\n",
    "    write.table(df,${_output[1]:r}, row.names=FALSE, col.names=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# Filter SNPs for PCA analysis and select individuals \n",
    "[filter_1]\n",
    "# The path to the file that contains the list of samples to keep (format FID, IID)\n",
    "parameter: keep_samples = path\n",
    "# MAF filter to use\n",
    "parameter: maf_filter = 0.01\n",
    "# Maximum missingess per-variant\n",
    "parameter: geno_filter = 0.01\n",
    "# Maximum missingness per-sample\n",
    "parameter: mind_filter = 0.02\n",
    "# HWE filter \n",
    "parameter: hwe_filter = 5e-08\n",
    "input: genoFile, group_by=1\n",
    "output: f'{cwd}/cache/{_input:bn}.filtered.bed'\n",
    "task: trunk_workers = 1, walltime = '10h', mem = '30G', cores = numThreads, tags = f'{step_name}_{_output:bn}'\n",
    "bash: container=container_lmm, expand= \"${ }\", stderr = f'{_output:n}.stderr', stdout = f'{_output:n}.stdout'\n",
    "    plink2 \\\n",
    "      --bfile ${_input:n} \\\n",
    "      ${('--maf %s' % maf_filter) if maf_filter > 0 else ''} ${('--geno %s' % geno_filter) if geno_filter > 0 else ''} ${('--hwe %s' % hwe_filter) if hwe_filter > 0 else ''} ${('--mind %s' % mind_filter) if mind_filter > 0 else ''} \\\n",
    "      --keep ${keep_samples}\\\n",
    "      --make-bed \\\n",
    "      --threads ${numThreads} \\\n",
    "      --out ${_output:n} "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# Merge all the .bed files into one bed file \n",
    "[filter_2: provides=[f'{cwd}/cache/{phenoFile:bn}.filtered.merged.bed']]\n",
    "input: group_by = 'all'\n",
    "output: bfile_merge = f'{cwd}/cache/{phenoFile:bn}.filtered.merged.bed'\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = '48h', mem = '60G', cores = numThreads, tags = f'{step_name}_{_output:bn}'\n",
    "bash: container=container_lmm, expand= \"${ }\", stderr = f'{_output:n}.stderr', stdout = f'{_output:n}.stdout'\n",
    "    echo -e ${' '.join([str(x)[:-4] for x in _input[1:]])} | sed 's/ /\\n/g' > ${_output:n}.merge_list\n",
    "    plink \\\n",
    "    --bfile ${_input[0]:n} \\\n",
    "    --merge-list ${_output:n}.merge_list \\\n",
    "    --make-bed \\\n",
    "    --out ${_output:n} \\\n",
    "    --threads ${numThreads} \\\n",
    "    --memory 48000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# LD prunning and remove related individuals (both ind of a pair)\n",
    "[filter_3]\n",
    "# Window size\n",
    "parameter: window = 50\n",
    "# Shift window every 10 snps\n",
    "parameter: shift = 10\n",
    "parameter: r2 = 0.1\n",
    "depends: f'{cwd}/cache/{phenoFile:bn}.filtered.merged.bed'\n",
    "input: named_output('related_samples')\n",
    "output: f'{cwd}/cache/{phenoFile:bn}.filtered.merged.prune.in', pruned_bed = f'{cwd}/{phenoFile:bn}.filtered.merged.prune.bed'\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = '48h', mem = '60G', cores = numThreads, tags = f'{step_name}_{_output[0]:bn}'\n",
    "bash: container=container_lmm, expand= \"${ }\", stderr = f'{_output[0]:n}.stderr', stdout = f'{_output[0]:n}.stdout'\n",
    "    plink2 \\\n",
    "    --bfile ${_depends:n} \\\n",
    "    --indep-pairwise ${window} ${shift} ${r2}  \\\n",
    "    --out ${_output[0]:nn} \\\n",
    "    --threads ${numThreads} \\\n",
    "    --memory 48000\n",
    "   \n",
    "    plink2 \\\n",
    "    --bfile ${_depends:n} \\\n",
    "    --extract ${_output[0]} \\\n",
    "    --remove ${_input} \\\n",
    "    --make-bed \\\n",
    "    --out ${_output[1]:n} "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# Run PCA analysis using flashpca\n",
    "[flashpca_1]\n",
    "input: f'{cwd}/{phenoFile:bn}.filtered.merged.prune.bed'\n",
    "output: f'{_input:n}.pca',\n",
    "        f'{_input:n}.pc1vpc2.png',\n",
    "        f'{_input:n}.pc3vpc4.png',\n",
    "        f'{_input:n}.pc5vpc6.png',\n",
    "        f'{_input:n}.pca.mahalanobis',\n",
    "        removed_outliers=f'{_input:bn}.pca.iid.no_outliers'\n",
    "task: trunk_workers = 1, walltime = '10h', mem = '30G', cores = numThreads, tags = f'{step_name}_{_output[0]:bn}'\n",
    "R: expand= \"${ }\", stderr = f'{_output[0]:n}.stderr', stdout = f'{_output[0]:n}.stdout'\n",
    "    # Load required libraries\n",
    "    library(dplyr)\n",
    "    library(ggplot2)\n",
    "    library(flashpcaR)\n",
    "    library(stats)\n",
    "    library(matrixStats)\n",
    "    # Read the PLINK binary files\n",
    "    fn <- ${_input:nr}\n",
    "    # Do the PCA computation\n",
    "    f <- flashpca(fn, ndim=${k}, stand=\"${stand}\", do_loadings=TRUE, check_geno=TRUE)\n",
    "    # Save the generated matrices to files\n",
    "    write.table(f$values,'${_output[0]:n}.values', sep=\" \", row.names=FALSE, col.names=FALSE) \n",
    "    write.table(f$vectors,'${_output[0]:n}.vectors', sep=\" \", row.names=TRUE, col.names=FALSE)\n",
    "    write.table(f$projection,'${_output[0]:n}.projection', sep=\" \", row.names=TRUE, col.names=FALSE)\n",
    "    write.table(f$loadings,'${_output[0]:n}.loadings', sep=\" \", row.names=FALSE, col.names=FALSE)\n",
    "    write.table(f$scale,'${_output[0]:n}.scale', sep=\" \", row.names=FALSE, col.names=FALSE)\n",
    "    # Use the projection file to generate plot\n",
    "    pca <- read.table('${_output[0]:n}.projection', sep=\" \")\n",
    "    colnames(pca) <- c(\"ID\",\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\",\"PC6\",\"PC7\",\"PC8\",\"PC9\",\"PC10\")\n",
    "    pca$IID <- sapply(strsplit(as.character(pca$ID),':'), \"[\", 1)\n",
    "    # Read fam file with phenotypes\n",
    "    pheno <- read.table(${phenoFile:r}, sep=\"\\t\", header=T )\n",
    "    pca_final <-merge(pheno, pca, by=\"IID\", all=FALSE)\n",
    "    \n",
    "    # Calculate mahalanobis distance\n",
    "    pc <- pca_final %>%\n",
    "         select(\"IID\", \"${trait_name}\",starts_with(\"PC\"))\n",
    "    pc_cov <- cov(pc[,3:12])\n",
    "    pc$mahal <- mahalanobis(pc[,3:12], center=FALSE, pc_cov)\n",
    "    pc$p <- pchisq(pc$mahal, df=9, lower.tail=FALSE)\n",
    "    # Set the cut-off to calculate outliers\n",
    "    manh_dis_sq_cutoff = quantile(pc$mahal,probs = 0.997) #6 sd from the mean chi-square double sided\n",
    "    cat(\"The cut-off for outlier removal is set to:\",manh_dis_sq_cutoff,\"and the number of individuals removed is:\", length(which(pc$mahal >= manh_dis_sq_cutoff)),\"\\n\")\n",
    "    \n",
    "    # Obtain the new sample\n",
    "    new_sample = pc[(pc$mahal <= manh_dis_sq_cutoff),1]\n",
    "    cat(\"The new sample size after outlier removal is:\",length(new_sample),\"\\n\")\n",
    "  \n",
    "    new_sample_df <- pc %>%\n",
    "    filter((IID%in%new_sample)) %>%\n",
    "    mutate(FID = IID) %>%\n",
    "    select(FID, IID, ${trait_name},starts_with(\"PC\"))\n",
    "  \n",
    "    new_sample_no_out <- pc %>%\n",
    "    filter(IID %in% new_sample) %>%\n",
    "    select(\"IID\")\n",
    "  \n",
    "    # Save file with mahalanobis distance and IID for PCA recalculation\n",
    "    write.table(new_sample_df,${_output[4]:r}, sep=\"\\t\", quote=FALSE, row.names=FALSE, col.names=TRUE)\n",
    "    write.table(new_sample_no_out,${_output[4]:r}, sep=\"\\t\", quote=FALSE, row.names=FALSE, col.names=FALSE)\n",
    "  \n",
    "    # Write the PC's to a file\n",
    "    write.table(pca_final,${_output[0]:r}, sep=\"\\t\", quote=FALSE, row.names=FALSE, col.names=TRUE)\n",
    "  \n",
    "    # Get the min and max values for x and y-axes \n",
    "    min_axis <- round(colMins(as.matrix(f$projection[sapply(f$projection, is.numeric)])),1)\n",
    "    max_axis <- round(colMaxs(as.matrix(f$projection[sapply(f$projection, is.numeric)])),1)\n",
    "  \n",
    "    png('${_output[1]}', width = 6, height = 4, unit='in', res=300)\n",
    "    ggplot(pca_final, aes(x=PC1, y=PC2)) + geom_point(aes(color=${trait_name}, shape=${trait_name}), size=2) + labs(title=\"PC1 vs PC2 exomed subset ${phenoFile:bn}\",x=\"PC1\", y=\"PC2\") +  scale_y_continuous(limits=c(min_axis, max_axis))  + scale_x_continuous(limits=c(min_axis, max_axis)) + theme_classic()\n",
    "    dev.off()\n",
    "  \n",
    "    png('${_output[2]}', width = 6, height = 4, unit='in', res=300)\n",
    "    ggplot(pca_final, aes(x=PC3, y=PC4)) + geom_point(aes(color=${trait_name}, shape=${trait_name}), size=2) + labs(title=\"PC3 vs PC4 exomed subset ${phenoFile:bn}\", x=\"PC3\", y=\"PC4\") +  scale_y_continuous(limits=c(min_axis, max_axis))  + scale_x_continuous(limits=c(min_axis, max_axis)) + theme_classic()\n",
    "    dev.off()\n",
    "    \n",
    "    png('${_output[3]}', width = 6, height = 4, unit='in', res=300)\n",
    "    ggplot(pca_final, aes(x=PC5, y=PC6)) + geom_point(aes(color=${trait_name}, shape=${trait_name}), size=2) + labs(title=\"P5 vs PC6 exomed subset ${phenoFile:bn}\",x=\"PC5\", y=\"PC6\") +   scale_y_continuous(limits=c(min_axis, max_axis))  + scale_x_continuous(limits=c(min_axis, max_axis)) + theme_classic()\n",
    "    dev.off()           "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# Filter out outliers from bed file and filter related individuals to project them onto PCA\n",
    "[flashpca_2]\n",
    "input: f'{cwd}/{phenoFile:bn}.filtered.merged.prune.bed', named_output('removed_outliers')\n",
    "output: f'{cwd}/{phenoFile:bn}.filtered.merged.prune.no_outliers.bed', f'{cwd}/{phenoFile:bn}.filtered.merged.prune.related_id.bed'\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = '48h', mem = '60G', cores = numThreads, tags = f'{step_name}_{_output[0]:bn}'\n",
    "bash: container=container_lmm, expand= \"${ }\", stderr = f'{_output[0]:n}.stderr', stdout = f'{_output[0]:n}.stdout'\n",
    "    \n",
    "    plink2 \\\n",
    "    --bfile ${_input[0]}\\\n",
    "    --keep ${_input[1]} \\\n",
    "    --make-bed \\\n",
    "    --out ${_output[0]:n} \n",
    "    \n",
    "    plink2 \\\n",
    "    --bfile ${_input[0]} \\\n",
    "    --keep ${cwd}/${bfile:bn}.related_id\\\n",
    "    --make-bed \\\n",
    "    --out ${_output[1]:n} "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# Re-do PCA without outliers\n",
    "[flashpca_3]\n",
    "input: f'{cwd}/{phenoFile:bn}.filtered.merged.prune.no_outliers.bed', f'{cwd}/{phenoFile:bn}.filtered.merged.prune.related_id.bed'\n",
    "output: f'{_input[0]:n}.pca',\n",
    "        f'{_input[0]:n}.pc1vpc2.no_out.png',\n",
    "        f'{_input[0]:n}.pc3vpc4.no_out.png',\n",
    "        f'{_input[0]:n}.pc5vpc6.no_out.png',\n",
    "        f'{_input[0]:n}.scree_cumPVE.png',\n",
    "        f'{_input[1]:n}.pca',\n",
    "        f'{_input[1]:n}.projected'\n",
    "task: trunk_workers = 1, walltime = '10h', mem = '30G', cores = numThreads, tags = f'{step_name}_{_output[0]:bn}'\n",
    "R: expand= \"${ }\", stderr = f'{_output[0]:n}.stderr', stdout = f'{_output[0]:n}.stdout'\n",
    "    # Load required libraries\n",
    "    library(dplyr)\n",
    "    library(ggplot2)\n",
    "    library(flashpcaR)\n",
    "    library(stats)\n",
    "    library(gridExtra)\n",
    "    # Read the PLINK binary files\n",
    "    fn <- ${_input[0]:nr}\n",
    "    # Do the PCA computation with outliers removed\n",
    "    f <- flashpca(fn, ndim=${k}, stand=\"${stand}\", do_loadings=TRUE, check_geno=TRUE)\n",
    "    # Save the generated matrices to files\n",
    "    write.table(f$values,'${_output[0]}.values', sep=\" \", row.names=FALSE, col.names=FALSE) \n",
    "    write.table(f$vectors,'${_output[0]}.vectors', sep=\" \", row.names=TRUE, col.names=FALSE)\n",
    "    write.table(f$projection,'${_output[0]}.projection', sep=\" \", row.names=TRUE, col.names=FALSE)\n",
    "    write.table(f$loadings,'${_output[0]}.loadings', sep=\" \", row.names=FALSE, col.names=FALSE)\n",
    "    write.table(f$scale,'${_output[0]}.scale', sep=\" \", row.names=FALSE, col.names=FALSE)\n",
    "    # Use the projection file to generate plot\n",
    "    pca <- read.table('${_output[0]}.projection', sep=\" \")\n",
    "    colnames(pca) <- c(\"ID\",\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\",\"PC6\",\"PC7\",\"PC8\",\"PC9\",\"PC10\")\n",
    "    pca$IID <- sapply(strsplit(as.character(pca$ID),':'), \"[\", 1)\n",
    "    # Read fam file with phenotypes\n",
    "    pheno <- read.table(${phenoFile:r}, sep=\"\\t\", header=T )\n",
    "    pca_final <-merge(pheno, pca, by=\"IID\", all=FALSE)\n",
    "    \n",
    "    # Write the PC's to a file\n",
    "    write.table(pca_final,${_output[0]:r}, sep=\"\\t\", quote=FALSE, row.names=FALSE, col.names=TRUE)\n",
    "  \n",
    "    # Get the min and max values for x and y-axes \n",
    "    min_axis <- round(colMins(as.matrix(f$projection[sapply(f$projection, is.numeric)])),1)\n",
    "    max_axis <- round(colMaxs(as.matrix(f$projection[sapply(f$projection, is.numeric)])),1)\n",
    "  \n",
    "    png('${_output[1]}', width = 6, height = 4, unit='in', res=300)\n",
    "    ggplot(pca_final, aes(x=PC1, y=PC2)) + geom_point(aes(color=${trait_name}, shape=${trait_name}), size=2) + labs(title=\"PC1 vs PC2 exomed subset ${phenoFile:bn}\",x=\"PC1\", y=\"PC2\") +  scale_y_continuous(limits=c(min_axis, max_axis))  + scale_x_continuous(limits=c(min_axis, max_axis)) + theme_classic()\n",
    "    dev.off()\n",
    "  \n",
    "    png('${_output[2]}', width = 6, height = 4, unit='in', res=300)\n",
    "    ggplot(pca_final, aes(x=PC3, y=PC4)) + geom_point(aes(color=${trait_name}, shape=${trait_name}), size=2) + labs(title=\"PC3 vs PC4 exomed subset ${phenoFile:bn}\", x=\"PC3\", y=\"PC4\") +  scale_y_continuous(limits=c(min_axis, max_axis))  + scale_x_continuous(limits=c(min_axis, max_axis)) + theme_classic()\n",
    "    dev.off()\n",
    "    \n",
    "    png('${_output[3]}', width = 6, height = 4, unit='in', res=300)\n",
    "    ggplot(pca_final, aes(x=PC5, y=PC6)) + geom_point(aes(color=${trait_name}, shape=${trait_name}), size=2) + labs(title=\"P5 vs PC6 exomed subset ${phenoFile:bn}\",x=\"PC5\", y=\"PC6\") + +  scale_y_continuous(limits=c(min_axis, max_axis))  + scale_x_continuous(limits=c(min_axis, max_axis)) + theme_classic()\n",
    "    dev.off() \n",
    "  \n",
    "    # Create scree plot\n",
    "    eigenvalues <- f$values\n",
    "    PVE <- eigenvalues[,1]\n",
    "    PVE <- round(PVE /sum(PVE), 2)\n",
    "    PVEplot <- qplot(c(1:10), PVE) + geom_line() + xlab(\"Principal Component\") + ylab(\"PVE\") + ggtitle(\"Scree Plot\") + ylim(0, 1) +scale_x_discrete(limits=factor(1:10))\n",
    "    PVE_cum <- cumsum(PVE) /sum(PVE)\n",
    "    cumPVEplot <- qplot(c(1:10), cumsum(PVE)) + geom_line() + xlab(\"Principal Component\") + ylab(\"PVE\") + ggtitle(\"Cumulative PVE Plot\") + ylim(0, 1) + scale_x_discrete(limits=factor(1:10))\n",
    "    png('${_output[4]}', width = 8, height = 4, unit='in', res=300)\n",
    "    grid.arrange(PVEplot, cumPVEplot, ncol = 2)\n",
    "    dev.off()\n",
    "  \n",
    "    # Poject related samples\n",
    "    frel <- ${_input[1]:nr}\n",
    "    fproject <- project(frel, loadings=f$loadings, orig_mean=f$center, orig_sd=f$scale)\n",
    "    # Get the min and max values for x and y-axes \n",
    "    min_axis <- round(colMins(as.matrix(fproject$projection[sapply(fproject$projection, is.numeric)])),1)\n",
    "    max_axis <- round(colMaxs(as.matrix(fproject$projection[sapply(fproject$projection, is.numeric)])),1)\n",
    "  \n",
    "    # Save the generated matrices to files\n",
    "    write.table(f$values,'${_output[5]}.values', sep=\" \", row.names=FALSE, col.names=FALSE) \n",
    "    write.table(f$vectors,'${_output[5]}.vectors', sep=\" \", row.names=TRUE, col.names=FALSE)\n",
    "    write.table(f$projection,'${_output[5]}.projection', sep=\" \", row.names=TRUE, col.names=FALSE)\n",
    "    write.table(f$loadings,'${_output[5]}.loadings', sep=\" \", row.names=FALSE, col.names=FALSE)\n",
    "    write.table(f$scale,'${_output[5]}.scale', sep=\" \", row.names=FALSE, col.names=FALSE)\n",
    "  \n",
    "    # Use the projection file to generate plot for the related individuals\n",
    "    pca <- read.table('${_output[5]}.projection', sep=\" \")\n",
    "    colnames(pca) <- c(\"ID\",\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\",\"PC6\",\"PC7\",\"PC8\",\"PC9\",\"PC10\")\n",
    "    pca$IID <- sapply(strsplit(as.character(pca$ID),':'), \"[\", 1)\n",
    "    # Read fam file with phenotypes\n",
    "    pheno <- read.table(${phenoFile:r}, sep=\"\\t\", header=TRUE)\n",
    "    pca_projection <-merge(pheno, pca, by=\"IID\", all=FALSE)\n",
    "  \n",
    "    png('${_output[6]}.PC1_PC2.png', width = 6, height = 4, unit='in', res=300)\n",
    "    ggplot(pca_projection, aes(x=PC1, y=PC2)) + geom_point(aes(color=${trait_name}, shape=${trait_name}), size=2) + labs(title=\"PC1 vs PC2 exomed subset ${phenoFile:bn}\",x=\"PC1\", y=\"PC2\") +  scale_y_continuous(limits=c(min_axis, max_axis))  + scale_x_continuous(limits=c(min_axis, max_axis)) + theme_classic()\n",
    "    dev.off()\n",
    "  \n",
    "    png('${_output[6]}.PC3_PC4.png', width = 6, height = 4, unit='in', res=300)\n",
    "    ggplot(pca_projection, aes(x=PC3, y=PC4)) + geom_point(aes(color=${trait_name}, shape=${trait_name}), size=2) + labs(title=\"PC3 vs PC4 exomed subset ${phenoFile:bn}\", x=\"PC3\", y=\"PC4\") +  scale_y_continuous(limits=c(min_axis, max_axis))  + scale_x_continuous(limits=c(min_axis, max_axis)) + theme_classic()\n",
    "    dev.off()\n",
    "    \n",
    "    png('${_output[6]}.PC5_PC6.png', width = 6, height = 4, unit='in', res=300)\n",
    "    ggplot(pca_projection, aes(x=PC5, y=PC6)) + geom_point(aes(color=${trait_name}, shape=${trait_name}), size=2) + labs(title=\"P5 vs PC6 exomed subset ${phenoFile:bn}\",x=\"PC5\", y=\"PC6\") +  scale_y_continuous(limits=c(min_axis, max_axis))  + scale_x_continuous(limits=c(min_axis, max_axis)) + theme_classic()\n",
    "    dev.off()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  },
  "sos": {
   "kernels": [
    [
     "Bash",
     "calysto_bash",
     "Bash",
     "#E6EEFF",
     ""
    ],
    [
     "SoS",
     "sos",
     "",
     "",
     "sos"
    ]
   ],
   "version": "0.21.21"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
