{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "# Principal component analysis\n",
    "\n",
    "The intention of this notebook is to perform the PCA analysis on genotype data and generate plots."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Overview\n",
    "\n",
    "\n",
    "Steps to generate a PCA include \n",
    "\n",
    "- removing related individuals\n",
    "- pruning variants in linkage disequilibrium (LD)\n",
    "- perform PCA analysis on genotype\n",
    "- excluding outlier samples in the PCA space for individuals of homogeneous self-reported ancestry. These outliers may suggest poor genotyping quality or distant relatedness.\n",
    "\n",
    "Pitfalls\n",
    "\n",
    "1. Some of the PCs may capture LD structure rather than population structure (decrease in power to detect associations in these regions of high LD)\n",
    "2. When projecting a new study dataset to the PCA space computed from a reference dataset: projected PCs are shrunk toward 0 in the new dataset\n",
    "3. PC scores may capture outliers that are due to family structure, population structure or other reasons; it might be beneficial to detect and remove these individuals to maximize the population structure captured by PCA (in the case of removing a few outliers) or to restrict analyses to genetically homogeneous samples\n",
    "\n",
    "## Input\n",
    "\n",
    "1. Genotype in PLINK format: could be different from bfile in the case of using for e.g. exome data / imputed genotype data\n",
    "2. Phentype files with population information and possibly disease or other labelling information, must have arbitrary column named 'FID' and 'IID'.\n",
    "\n",
    "The inputs should be splitted into sets of related and unrelated individuals. Additionally you may want to prepare data per population. See \"How to run this workflow\" section for more details.\n",
    "\n",
    "## Output\n",
    "\n",
    "1. PCA models (inside RDS file)\n",
    "2. PCA scores (inside RDS file)\n",
    "3. Mahalanobis distances and outliers to remove"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## General workflow\n",
    "\n",
    "1. Estimate relatedness of the individuals in the sample by PLINK 2 that implements the KING algorithm\n",
    "2. Select specific SNPs and samples using PLINK (QC maf>1%, geno missing rate (`geno`) 0.1, individual missing rate (`mind`) =0.1 and `hwe`=5e-08), and remove related individuals\n",
    "3. SNPs thining by doing LD-pruning (window=50, shift=10, r2=0.1 are the defaults)\n",
    "4. Run PCA using only unrelated individuals for all populations, and examine the resulting plot\n",
    "5. Project back related individuals, and generate a list of suggested samples to remove based on Mahalanobis distance test statistic per population. Default criteria is 0.997 percentile (two sided) but we recommend checking the output plot before and after removal and rethink about it.\n",
    "6. Remove outliers based on list previously generated\n",
    "\n",
    "Now the data should not have outliers. If you have subpopulations in the data, then additional steps should be applied for:\n",
    "\n",
    "7. Split data into different populations, each population data should have both related vs unrelated individual data-sets\n",
    "8. For each population, perform QC\n",
    "9. For each population, re-calculate per population PC's for unrelated individuals\n",
    "10. For each population, project related samples back to the PC space\n",
    "\n",
    "No need to remove outliers at this point."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Method\n",
    "\n",
    "Here is a quick recap of PCA analysis for those not immediately familiar with the method. PCA is a mathematical method to reduce dimensionality of the data while retaining most of the variation in the dataset. \n",
    "This is accomplished by identifying directions or Principal Components (PC's) that account for the maximum variation in the data. \n",
    "\n",
    "One common approach to PCA is based on the singular-value decomposition of the the data matrix $X$ (in our case the genotype matrix),\n",
    "$$X = U D V^T,$$\n",
    "where $U$ are the left eigenvectors, $D$ is the diagonal matrix of singular values, and $V$ are the right eigenvectors (also called loadings). \n",
    "\n",
    "PCA can also be done using the eigen-decomposition of $X X^T$\n",
    "$$X X^T = U S U^T,$$ where $S=D^2$ is the diagonal matrix of eigenvalues.\n",
    "$X$ is usually centred (mean-subtracted) or standardised (mean subtracted, then divided by standard deviation) before PCA.\n",
    "\n",
    "For PCA of SNP genotypes (at least in diploid organisms), the common standardisation is\n",
    "$$X_{ij}^{\\prime} = \\frac{X_{ij} - 2p_j}{\\sqrt{2 p_j (1 - p_j)}},$$\n",
    "where $X_{ij}$ is the genotype (minor allele dosage $\\{0, 1, 2\\}$) for the $i$th individual and the $j$th SNP, and $p_j$ is the minor allele frequency (MAF) for the $j$th SNP. In addition, the eigenvalues are scaled by the number of SNPs $m$ (equivalent to performing the eigen-decomposition of $XX^T/m$).\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## How to run this workflow"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "\n",
    "\n",
    "### Step 1: Estimate kinship in the sample\n",
    "Aim: identify and output closely related individuals prior to PCA analysis.\n",
    "\n",
    "```\n",
    "sos run GWAS_QC.ipynb king \\\n",
    "    --genoFile <all_samples.bed>\n",
    "```\n",
    "\n",
    "### Step 2 and 3: Sample selection and QC the genetic data \n",
    "\n",
    "1. QC based on MAF, sample and variant missigness and Hardy-Weinberg Equilibrium. You can provide a list of samples to keep, or to remove. For example:\n",
    "    - Only extract data for one population\n",
    "    - Only extract data for related individuals\n",
    "    - Only extract data for unrelated individuals\n",
    "   In current context we would like to extract data for unrelated individuals and proceed with the rest of the QC steps.\n",
    "2. LD pruning. Prune SNPs in linkage dissequilibrium to make sure the PCA actually captures population structure and not LD structure (which could reduce the power of detecting genetic associations in these LD-regions).\n",
    "\n",
    "Get unrelated individuals\n",
    "\n",
    "```\n",
    "sos run GWAS_QC.ipynb qc \\\n",
    "    --genoFile <all_samples.bed> \\\n",
    "    --remove-samples <king_output.related_id> \\\n",
    "    ...\n",
    "```\n",
    "\n",
    "Get related individuals but **use the same variants as before**. We therefore only run `qc:1` with a list of variants extracted, and other filtering parameters set to 0.\n",
    "\n",
    "```\n",
    "sos run GWAS_QC.ipynb qc:1 \\\n",
    "    --genoFile <all_samples.bed> \\\n",
    "    --keep-samples <king_output.related_id> \\\n",
    "    --keep-variants <previous_command_output.prune.in> \\\n",
    "    --maf-filter 0 --geno-filter 0 --mind-filter 0 --hwe-filter 0 \\\n",
    "    ...\n",
    "```\n",
    "\n",
    "### Step 4: PCA analysis for all samples\n",
    "\n",
    "```\n",
    "sos run PCA.ipynb flashpca \\\n",
    "    --genoFile <unrelated_samples.bed>\n",
    "    --phenoFile <phenotypes.txt>\n",
    "```\n",
    "\n",
    "### Step 5: Projection of related individuals and outlier detection \n",
    "\n",
    "```\n",
    "sos run PCA.ipynb project_samples \\\n",
    "    --pca-model <unrelated_samples.pca.rds> \\\n",
    "    --genoFile <related_samples.bed> \\\n",
    "    --phenoFile <phenotypes.txt>\n",
    "```\n",
    "\n",
    "### Step 6: Outlier removal\n",
    "\n",
    "```\n",
    "sos run GWAS_QC.ipynb qc \\\n",
    "    --remove-samples <mahalanobi_output.outliers> \\\n",
    "    --genoFile <all_samples.bed>\n",
    "```\n",
    "\n",
    "If you are analyzing a homogenous population then this is the end.\n",
    "\n",
    "### Step 7: Split data by population\n",
    "\n",
    "For each population, you can still use the GWAS_QC.ipynb and PCA.ipynb workflows, except that at this step you have to work manually on the `--remove-samples` and `--keep-samples` files.\n",
    "\n",
    "Please do this manually now to create `remove-samples` (of related and not in the current pop) and `keep-samples` (of related and in the current pop) for next steps.\n",
    "\n",
    "### Step 8: For each population perform QC\n",
    "\n",
    "Get unrelated individuals\n",
    "\n",
    "```\n",
    "sos run GWAS_QC.ipynb qc \\\n",
    "    --genoFile <all_samples.bed> \\\n",
    "    --remove-samples  <king_output.related_id AND samples not in this population> \\\n",
    "    ...\n",
    "```\n",
    "\n",
    "Get related individuals but **use the same variants as before**. We therefore only run `qc:1` with a list of variants extracted, and other filtering parameters set to 0.\n",
    "\n",
    "```\n",
    "sos run GWAS_QC.ipynb qc:1 \\\n",
    "    --genoFile <all_samples.bed> \\\n",
    "    --keep-samples <king_output.related_id AND samples within this population> \\\n",
    "    --keep-variants <previous_command_output.prune.in> \\\n",
    "    --maf-filter 0 --geno-filter 0 --mind-filter 0 --hwe-filter 0 \\\n",
    "    ...\n",
    "```\n",
    "\n",
    "### Step 9: For each population do PCA\n",
    "\n",
    "```\n",
    "sos run PCA.ipynb flashpca \\\n",
    "    --genoFile <unrelated_samples_this_pop.bed> \\\n",
    "    --phenoFile <phenotypes.txt> \\\n",
    "    ...\n",
    "```\n",
    "\n",
    "### Step 10: For each population projection of related individuals, no need for outlier detection\n",
    "\n",
    "Therefore we run `project_samples:1` only the first step.\n",
    "\n",
    "```\n",
    "sos run PCA.ipynb project_samples:1 \\\n",
    "    --pca-model <unrelated_samples_this_pop.pca.rds> \\\n",
    "    --genoFile <related_samples_this_pop.bed> \\\n",
    "    --phenoFile <phenotypes.txt> \\\n",
    "    ...\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## How to run MWE: \n",
    "\n",
    "running on the columbia cluster, using data of MWE_AD\n",
    "\n",
    "Step 1 Estimate kinship in the sample\n",
    "\n",
    "```\n",
    "sos run ~/GWAS_QC.ipynb king \\\n",
    "    --container_lmm /mnt/mfs/statgen/containers/lmm.sif \\\n",
    "    --cwd ~/output \\\n",
    "    --genoFile ~/MWE_AD/rename_chr22.bed \\\n",
    "    --name first \\\n",
    "    --kinship 0.13\n",
    "```\n",
    "\n",
    "Step 2 Get the genotypes (make bed)  for unrelated individuals\n",
    "\n",
    "```\n",
    "sos run ~/GWAS_QC.ipynb qc\\\n",
    "    --container_lmm /mnt/mfs/statgen/containers/lmm.sif \\\n",
    "    --cwd ~/output \\\n",
    "    --genoFile ~/MWE_AD/rename_chr22.bed \\\n",
    "    --remove_samples ~/output/rename_chr22.first.related_id\\\n",
    "    --maf_filter 0.5 \\\n",
    "    --geno_filter 0.2 \\\n",
    "    --mind_filter 0.1 \\\n",
    "    --hwe_filter 0.0 \\\n",
    "    --name unrelated \\\n",
    "    --window 50 \\\n",
    "    --shift 10 \\\n",
    "    --r2 0.5 \n",
    "```\n",
    "Step 3\n",
    "Get the genotype file (make bed) for related individuals\n",
    "\n",
    "```\n",
    "sos run  ~/GWAS_QC.ipynb qc:1 \\\n",
    "    --container_lmm /mnt/mfs/statgen/containers/lmm.sif \\\n",
    "    --cwd ~/output \\\n",
    "    --genoFile ~/MWE_AD/rename_chr22.bed \\\n",
    "    --keep-samples ~/output/rename_chr22.first.related_id\\\n",
    "    --keep-variants ~/output/cache/rename_chr22.unrelated.filtered.prune.in\\\n",
    "    --maf-filter 0 --geno-filter 0 --mind-filter 0 --hwe-filter 0\\\n",
    "    --name related\n",
    "```\n",
    "Step 4: Run PCA for unrelated individuals, and plot PCA plots (flash_pca_2 will call plot_pca to plot, no need to do flashpca+plot_pca)\n",
    "\n",
    "```\n",
    "sos run  ~/PCA.ipynb flashpca \\\n",
    "    --container_lmm /mnt/mfs/statgen/containers/lmm.sif \\\n",
    "    --cwd ~/output \\\n",
    "    --genoFile ~/output/cache/rename_chr22.unrelated.filtered.prune.bed\\\n",
    "    --phenoFile ~/MWE_AD/MWE_pheno.txt\\\n",
    "    --k 4 \\\n",
    "    --label_col RACE \\\n",
    "    --pop_col RACE \\\n",
    "    --plot_data ~/output/MWE_pheno.pca.rds\n",
    "```\n",
    "\n",
    "Step 5: \n",
    "Project back related individuals, detect & generate outlier for the entire dataset, and plot PCA plots without showing outliers\n",
    "\n",
    "```\n",
    "sos run  ~/PCA.ipynb project_samples+plot_pca \\\n",
    "    --container_lmm /mnt/mfs/statgen/containers/lmm.sif \\\n",
    "    --cwd ~/output \\\n",
    "    --genoFile ~/output/cache/rename_chr22.related.filtered.bed \\\n",
    "    --phenoFile ~/MWE_AD/MWE_pheno.txt\\\n",
    "    --label_col RACE \\\n",
    "    --plot_K 3\\\n",
    "    --pca_model ~/output/MWE_pheno.pca.rds \\\n",
    "    --pop_col RACE \\\n",
    "    --plot_data ~/output/MWE_pheno.pca.rds\n",
    "```\n",
    "\n",
    "Step 6: \n",
    "Outlier removal for the entire dataset\n",
    "```\n",
    "sos run ~/GWAS_QC.ipynb qc:1 \\\n",
    "    --container_lmm /mnt/mfs/statgen/containers/lmm.sif \\\n",
    "    --cwd ~/Downloads/output \\\n",
    "    --remove-samples ~/MWE_AD/MWE_pheno.projected.outliers \\\n",
    "    --genoFile ~/MWE_AD/rename_chr22.bed \\\n",
    "    --name no_outliers\n",
    "```\n",
    "\n",
    "Plot PCA with outliers circled in red\n",
    "```\n",
    "sos run ~/PCA.ipynb plot_pca \\\n",
    "    --container_lmm /mnt/mfs/statgen/containers/lmm.sif \\\n",
    "    --cwd ~/output \\\n",
    "    --outlier_file ~/MWE_AD/MWE_pheno.projected.outliers \\\n",
    "    --genoFile ~/MWE_AD/rename_chr22.bed \\\n",
    "    --phenoFile ~/MWE_AD/MWE_pheno.txt\\\n",
    "    --k 4 \\\n",
    "    --plot_K 3\\\n",
    "    --label_col RACE \\\n",
    "    --pop_col RACE \\\n",
    "    --plot_data ~/output/MWE_pheno.pca.rds\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Command interface"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[91mERROR\u001b[0m: \u001b[91mNotebook JSON is invalid: %s\u001b[0m\n",
      "usage: sos run PCA.ipynb [workflow_name | -t targets] [options] [workflow_options]\n",
      "  workflow_name:        Single or combined workflows defined in this script\n",
      "  targets:              One or more targets to generate\n",
      "  options:              Single-hyphen sos parameters (see \"sos run -h\" for details)\n",
      "  workflow_options:     Double-hyphen workflow-specific parameters\n",
      "\n",
      "Workflows:\n",
      "  flashpca\n",
      "  project_samples\n",
      "  plot_pca\n",
      "  detect_outliers\n",
      "\n",
      "Global Workflow Options:\n",
      "  --cwd VAL (as path, required)\n",
      "                        the output directory for generated files\n",
      "  --genoFile VAL (as path, required)\n",
      "                        Plink binary file\n",
      "  --phenoFile VAL (as path, required)\n",
      "                        The phenotypic file\n",
      "  --pop-col ''\n",
      "                        Name of the population column in the phenoFile\n",
      "  --pops  (as list)\n",
      "  --label-col VAL (as str, required)\n",
      "                        Name of the color label column in the phenoFile; can be\n",
      "                        the same as population column\n",
      "  --container-lmm 'statisticalgenetics/lmm:1.8'\n",
      "                        Software container option\n",
      "  --job-size 1 (as int)\n",
      "                        For cluster jobs, number commands to run per job\n",
      "  --walltime 5h\n",
      "                        Wall clock time expected\n",
      "  --mem 16G\n",
      "                        Memory expected\n",
      "  --numThreads 10 (as int)\n",
      "                        Number of threads\n",
      "\n",
      "Sections\n",
      "  flashpca_1:           Run PCA analysis using flashpca\n",
      "    Workflow Options:\n",
      "      --k 10 (as int)\n",
      "                        Number of Principal Components to output. Default is 10\n",
      "      --stand binom2\n",
      "                        How to standardize X before PCA\n",
      "  flashpca_2:\n",
      "    Workflow Options:\n",
      "      --outlier-file . (as path)\n",
      "  project_samples_1:    Project back to PCA model additional samples\n",
      "    Workflow Options:\n",
      "      --pca-model VAL (as path, required)\n",
      "                        Here genoFile is samples to be projected back\n",
      "  project_samples_2:\n",
      "    Workflow Options:\n",
      "      --prob 0.95 (as float)\n",
      "                        Set the probability to remove outliers eg 0.95 or 0.997\n",
      "      --[no-]robust (default to True)\n",
      "                        Robust Mahalanobis to outliers\n",
      "  plot_pca:             Plot PCA results. Can be used as: \"flashpca + plot_pca\"\n",
      "                        or as \"project_samples + plot_pca\"\n",
      "    Workflow Options:\n",
      "      --outlier-file . (as path)\n",
      "      --plot-data VAL (as path, required)\n",
      "  detect_outliers:      Calculate Mahalanobis distance per population and report\n",
      "                        outliers\n",
      "    Workflow Options:\n",
      "      --prob 0.95 (as float)\n",
      "                        Set the probability to remove outliers eg 0.95 or 0.997\n",
      "      --[no-]robust (default to True)\n",
      "                        Robust Mahalanobis to outliers\n",
      "      --pca-result VAL (as path, required)\n"
     ]
    }
   ],
   "source": [
    "sos run PCA.ipynb -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[global]\n",
    "# the output directory for generated files\n",
    "parameter: cwd = path\n",
    "# Plink binary file\n",
    "parameter: genoFile = path\n",
    "# The phenotypic file\n",
    "parameter: phenoFile = path\n",
    "# Name of the population column in the phenoFile\n",
    "parameter: pop_col = \"\"\n",
    "parameter: pops = []\n",
    "# Name of the color label column in the phenoFile; can be the same as population column\n",
    "parameter: label_col = str\n",
    "# Homogeneity of populations \n",
    "parameter: homogeneous = False\n",
    "# Software container option\n",
    "parameter: container_lmm = 'statisticalgenetics/lmm:1.8'\n",
    "# For cluster jobs, number commands to run per job\n",
    "parameter: job_size = 1\n",
    "# Wall clock time expected\n",
    "parameter: walltime = \"5h\"\n",
    "# Memory expected\n",
    "parameter: mem = \"16G\"\n",
    "# Number of threads\n",
    "parameter: numThreads = 10\n",
    "# Specify which step it is, to differentiate the RDS and pca PLOT outputs for different steps \n",
    "parameter: name= str\n",
    "suffix = '_'.join(pops)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## PCA analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# Run PCA analysis using flashpca \n",
    "[flashpca_1]\n",
    "# Number of Principal Components to output. Default is 10\n",
    "parameter: k = 10\n",
    "# How to standardize X before PCA\n",
    "parameter: stand = \"binom2\"\n",
    "input: genoFile, phenoFile\n",
    "output: f'{cwd}/{phenoFile:bn}.{(suffix+\".\") if suffix != \"\" else \"\"}pca.{name}.rds'\n",
    "task: trunk_workers = 1, walltime = walltime, mem = mem, cores = numThreads, tags = f'{step_name}_{_output[0]:bn}'\n",
    "R: container = container_lmm, expand = \"${ }\", stderr = f'{_output[0]}.stderr', stdout = f'{_output[0]}.stdout'\n",
    "    # Load required libraries\n",
    "    library(flashpcaR) \n",
    "    f <- flashpca(${_input[0]:nr}, ndim=${k}, stand=\"${stand}\", do_loadings=TRUE, check_geno=TRUE)\n",
    "    # Use the projection file to generate pca plot\n",
    "    pca <- as.data.frame(f$projection)\n",
    "    pca <- tibble::rownames_to_column(pca, \"ID\")\n",
    "    colnames(pca) <- c(\"ID\",paste0(\"PC\", 1:${k}))\n",
    "    pca$IID <- sapply(strsplit(as.character(pca$ID),':'), \"[\", 1)\n",
    "    # Read fam file with phenotypes\n",
    "    pheno <- read.table(${_input[1]:r}, sep=\"\\t\", header=T )\n",
    "    pca <-merge(pheno, pca, by=\"IID\", all=FALSE) \n",
    "    \n",
    "    #Calculate mean/median/cov per pop\n",
    "    if(${\"FALSE\" if homogeneous else \"TRUE\"}){\n",
    "        pop_group <- split(pca[ ,c(paste0(\"PC\", 1:${k}))], list(Group = pca$${pop_col}))\n",
    "        pc_cov <- lapply(pop_group, function(x) cov(x))\n",
    "        pc_mean <- lapply(pop_group, function(x) sapply(x, mean))\n",
    "        pc_median <- lapply(pop_group, function(x) sapply(x, median))\n",
    "    } else {\n",
    "        pc_cov <- cov(f$projection)\n",
    "        pc_mean <- apply(f$projection, 2, mean)\n",
    "        pc_median <- apply(f$projection, 2, median)\n",
    "    }\n",
    "    # Write the PC scores to a file\n",
    "    write.table(pca,\"${_output:n}.txt\", sep=\"\\t\", quote=FALSE, row.names=FALSE, col.names=TRUE)\n",
    "    # save results\n",
    "    saveRDS(list(pca_model = f, pc_scores = pca, pc_mean = pc_mean, pc_median = pc_median, pc_cov = pc_cov, k = ${k}, meta = \"${_input[1]:bn} ${suffix}\"), ${_output:r})\n",
    "\n",
    "[flashpca_2]\n",
    "parameter: outlier_file = path()                            \n",
    "output: f'{_input:nn}.{name}.pc.png',\n",
    "        f'{_input:nn}.{name}.scree.png'\n",
    "sos_run(\"plot_pca\", pca_model = _input, outlier_file = outlier_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Project related individuals back"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# Project back to PCA model additional samples\n",
    "[project_samples_1]\n",
    "# Here genoFile is samples to be projected back\n",
    "parameter: pca_model = path\n",
    "input: genoFile, phenoFile, pca_model\n",
    "output: f'{_input[1]:n}.{name}.projected.rds'\n",
    "task: trunk_workers = 1, walltime = walltime, mem = mem, cores = numThreads, tags = f'{step_name}_{_output[0]:bn}'\n",
    "R: container=container_lmm, expand= \"${ }\", stderr = f'{_output[0]:n}.stderr', stdout = f'{_output[0]:n}.stdout'\n",
    "    # Load required libraries\n",
    "    library(dplyr)\n",
    "    library(flashpcaR)\n",
    "    # Read the PLINK binary files\n",
    "    frel <- ${_input[0]:nr}\n",
    "    # Read loadings, center and scale from previous PCA\n",
    "    dat <- readRDS(${_input[2]:r})\n",
    "    f <- dat$pca_model\n",
    "    # Read the bim file to obtain reference alleles\n",
    "    bim <- read.table('${_input[0]:n}.bim')\n",
    "    ref <- as.character(bim[,5])\n",
    "    names(ref) <- bim[,2]\n",
    "    # Project related samples\n",
    "    fpro <- project(frel, loadings=f$loadings, orig_mean=f$center, orig_sd=f$scale, ref_allele=ref)\n",
    "    pca <- fpro$projection\n",
    "    k = dat$k\n",
    "    pca <- as.data.frame(fpro$projection)\n",
    "    pca <- tibble::rownames_to_column(pca, \"ID\")\n",
    "    colnames(pca) <- c(\"ID\",paste0(\"PC\", 1:k))\n",
    "    pca$IID <- sapply(strsplit(as.character(pca$ID),':'), \"[\", 1)\n",
    "    # Read fam file with phenotypes\n",
    "    pheno <- read.table(${_input[1]:r}, sep=\"\\t\", header=T )\n",
    "    pca <-merge(pheno, pca, by=\"IID\", all=FALSE) \n",
    "    dat$pc_scores = rbind(dat$pc_scores, pca)\n",
    "    \n",
    "    # Write the PC scores to a file\n",
    "    write.table(dat$pc_scores,\"${_output:n}.txt\", sep=\"\\t\", quote=FALSE, row.names=FALSE, col.names=TRUE)\n",
    "    # save results\n",
    "    saveRDS(dat, ${_output:r})\n",
    "  \n",
    "[project_samples_2]\n",
    "# Set the probability to remove outliers eg 0.95 or 0.997\n",
    "parameter: prob = 0.997\n",
    "# Robust Mahalanobis to outliers\n",
    "parameter: robust = True\n",
    "output: distance=f'{_input:n}.{name}.mahalanobis.rds',\n",
    "        removed_outliers=f'{_input:n}.{name}.outliers',\n",
    "        analysis_summary=f'{_input:n}.{name}.analysis_summary.md',\n",
    "        qqplot_mahalanobis=f'{_input:n}.{name}.mahalanobis_qq.png',\n",
    "        hist_mahalanobis=f'{_input:n}.{name}.mahalanobis_hist.png'\n",
    "sos_run(\"detect_outliers\", pca_result=_input, prob=prob, robust=robust)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Plot PCA results\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# Plot PCA results. Can be used:\n",
    "# independently as \"plot_pca\" or conbinded with other workflow as \"project_samples+plot_pca\"\n",
    "[plot_pca]\n",
    "parameter: outlier_file = path()\n",
    "parameter: plot_data = path\n",
    "parameter: plot_K = int\n",
    "input: plot_data\n",
    "output: f'{_input:nn}.{name}.pc.png',\n",
    "        f'{_input:nn}.{name}.scree.png'\n",
    "task: trunk_workers = 1, walltime = walltime, mem = mem, cores = 1, tags = f'{step_name}_{_output[0]:bn}'\n",
    "R: container = container_lmm,volumes = [f\"{outlier_file:ad}:{outlier_file:ad}\"], expand = \"${ }\", stderr = f'{_output[0]}.stderr', stdout = f'{_output[0]}.stdout'   \n",
    "    library(dplyr)\n",
    "    library(ggplot2)\n",
    "    library(gridExtra)\n",
    "    library(matrixStats)\n",
    "    dat = readRDS(${_input:r})\n",
    "    f = dat$pca_model\n",
    "    pca_final = dat$pc_scores\n",
    "    pca_final<-pca_final%>%\n",
    "        mutate(${label_col}=as.character(${label_col}))\n",
    "    k = dat$k\n",
    "    \n",
    "    ###\n",
    "    # Make the plots\n",
    "    ###\n",
    "    # Get the min and max values for x and y-axes \n",
    "    min_axis <- round(colMins(as.matrix(f$projection[sapply(f$projection, is.numeric)])),1)\n",
    "    max_axis <- round(colMaxs(as.matrix(f$projection[sapply(f$projection, is.numeric)])),1)\n",
    "    if (${\"TRUE\" if outlier_file.is_file() else \"FALSE\"}) {\n",
    "        outliers <- read.table(${outlier_file:r}, col.names=c(\"FID\", \"IID\"))\n",
    "        plot_pcs = function(pca_final, x, y, title=\"\") {\n",
    "          ggplot(pca_final, aes_string(x=x, y=y)) + geom_point(aes(color=${label_col}), size=1) + geom_point(data=filter(pca_final, IID %in% outliers$IID), shape = 21, size=2, color='red', stroke = 1) + \n",
    "              labs(title=title,x=x, y=y) +\n",
    "              scale_y_continuous(limits=c(min_axis, max_axis)) +\n",
    "              scale_x_continuous(limits=c(min_axis, max_axis)) +\n",
    "              theme_classic()\n",
    "        }} else {\n",
    "        plot_pcs = function(pca_final, x, y, title=\"\") {\n",
    "          ggplot(pca_final, aes_string(x=x, y=y)) + geom_point(aes(color=${label_col}), size=1) + \n",
    "              labs(title=title,x=x, y=y) +\n",
    "              scale_y_continuous(limits=c(min_axis, max_axis)) +\n",
    "              scale_x_continuous(limits=c(min_axis, max_axis)) +\n",
    "              theme_classic()\n",
    "        }}\n",
    "    unit = 4\n",
    "    n_col = 4\n",
    "    n_row = ceiling(k / n_col)\n",
    "    plots = lapply(1:(${plot_K}-1), function(i) plot_pcs(pca_final, paste0(\"PC\",i), paste0(\"PC\",i+1), dat$meta))\n",
    "    png('${_output[0]}', width = unit * n_col, height = unit * n_row, unit='in', res=300)\n",
    "    do.call(gridExtra::grid.arrange, c(plots, list(ncol = n_col, nrow = n_row)))\n",
    "    dev.off()\n",
    "    # Create scree plot\n",
    "    PVE <- f$values\n",
    "    PVE <- round(PVE/sum(PVE), 2)\n",
    "    PVEplot <- qplot(c(1:k), PVE) + geom_line() + xlab(\"Principal Component\") + ylab(\"PVE\") + ggtitle(\"Scree Plot\") + ylim(0, 1) +scale_x_discrete(limits=factor(1:k))\n",
    "    PVE_cum <- cumsum(PVE)/sum(PVE)\n",
    "    cumPVEplot <- qplot(c(1:k), cumsum(PVE)) + geom_line() + xlab(\"Principal Component\") + ylab(\"PVE\") + ggtitle(\"Cumulative PVE Plot\") + ylim(0, 1) + scale_x_discrete(limits=factor(1:k))\n",
    "    png('${_output[1]}', width = 8, height = 4, unit='in', res=300)\n",
    "    grid.arrange(PVEplot, cumPVEplot, nrow = 1)\n",
    "    dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Detect outliers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "##### Calculate Mahalanobis distance per population and report outliers\n",
    "[detect_outliers]\n",
    "# Set the probability to remove outliers eg 0.95 or 0.997\n",
    "parameter: prob = 0.997\n",
    "# Robust Mahalanobis to outliers\n",
    "parameter: robust = True\n",
    "parameter: pca_result = path\n",
    "input: pca_result\n",
    "output: distance=f'{_input:n}.{name}.mahalanobis.rds',\n",
    "        removed_outliers=f'{_input:n}.{name}.outliers',\n",
    "        analysis_summary=f'{_input:n}.{name}.analysis_summary.md',\n",
    "        qqplot_mahalanobis=f'{_input:n}.{name}.mahalanobis_qq.png',\n",
    "        hist_mahalanobis=f'{_input:n}.{name}.mahalanobis_hist.png'\n",
    "task: trunk_workers = 1, walltime = walltime, mem = mem, cores = 1, tags = f'{step_name}_{_output[0]:bn}'\n",
    "bash: container = container_lmm, expand = \"${ }\"\n",
    "    echo '''---\n",
    "    theme: base-theme\n",
    "    style: |\n",
    "      img {\n",
    "        height: 80%;\n",
    "        display: block;\n",
    "        margin-left: auto;\n",
    "        margin-right: auto;\n",
    "      }\n",
    "    ---    \n",
    "    ''' > ${_output[2]}\n",
    "    \n",
    "R: container = container_lmm, expand= \"${ }\", stderr = f'{_output[0]}.stderr', stdout = f'{_output[0]}.stdout'\n",
    "    # Load required libraries\n",
    "    library(dplyr)\n",
    "    library(ggplot2)\n",
    "    library(gridExtra)\n",
    "  \n",
    "    # invert a known covariance matrix but allow them to be numerically singular matrix (still assuming full rank)\n",
    "    robust_inv = function(s) {\n",
    "        tryCatch(solve(s), error=function(cond) solve(Matrix::nearPD(s)$mat))\n",
    "    }\n",
    "\n",
    "    # Calculate mahalanobis distance  \n",
    "    calc_mahalanobis_dist = function(x, m, s, name = '', prob=${prob}) {\n",
    "        pc <- x %>%\n",
    "          select(\"IID\",\"FID\", starts_with(\"PC\"))\n",
    "        mu_pc <- pc %>%\n",
    "        select(starts_with(\"PC\"))\n",
    "        mu = t(t(mu_pc) - m)\n",
    "        pc$mahal = rowSums(mu %*% robust_inv(s) * mu)\n",
    "        pc$p <- pchisq(pc$mahal, df=nrow(s), lower.tail=FALSE)\n",
    "        manh_dis_sq_cutoff = quantile(pc$mahal, probs=prob,na.rm=TRUE)\n",
    "        # Obtain outliers\n",
    "        outliers = pc[(pc$mahal > manh_dis_sq_cutoff),]\n",
    "        d_summary = paste0(capture.output(summary(pc$mahal)), collapse = '\\n')\n",
    "        msg = paste('#', name, \"result summary\\n## Mahalanobis distance summary:\\n```\\n\", d_summary, \"\\n```\\n\", \n",
    "            paste(\"The cut-off for outlier removal is set to:\", manh_dis_sq_cutoff, \"and the number of individuals to remove is:\", nrow(outliers),\"\\n\"),\n",
    "            paste(\"The new sample size after outlier removal is:\",nrow(pc) - nrow(outliers),\"\\n\"))\n",
    "        #\n",
    "        outliers <- outliers %>%\n",
    "        mutate(FID = IID) %>% \n",
    "        select(FID, IID)%>% \n",
    "        filter(!is.na(IID))\n",
    "      list(pc=pc, manh_dis_sq_cutoff=manh_dis_sq_cutoff, msg=msg, outliers=outliers)\n",
    "    }\n",
    "\n",
    "    dat = readRDS(${_input:r})\n",
    "    if (is.list(dat$pc_mean)) {\n",
    "      pops = names(dat$pc_mean)\n",
    "      pop_group = split(dat$pc_scores, f = dat$pc_scores$${pop_col})\n",
    "      res = lapply(pops, function(p) calc_mahalanobis_dist(pop_group[[p]], dat$${\"pc_mean\" if not robust else \"pc_median\"}[[p]], dat$pc_cov[[p]], name = paste(dat$meta, p)))\n",
    "      names(res) = pops\n",
    "      res = list(\n",
    "          msg = do.call(paste, c(lapply(pops, function(p) res[[p]]$msg), sep = \"\\n\")),\n",
    "          manh_dis_sq_cutoff = cbind(pops, sapply(pops, function(p) res[[p]]$manh_dis_sq_cutoff)),\n",
    "          outliers = do.call(rbind, c(lapply(pops, function(p) res[[p]]$outliers))),\n",
    "          pc = do.call(rbind, c(lapply(pops, function(p) res[[p]]$pc)))\n",
    "          )\n",
    "    } else {\n",
    "      res = calc_mahalanobis_dist(dat$pc_scores, dat$${\"pc_mean\" if not robust else \"pc_median\"}, dat$pc_cov, name = dat$meta)\n",
    "    }\n",
    "      \n",
    "    write(res$msg, ${_output[2]:r})   \n",
    "    # Plot mahalanobis\n",
    "    k = dat$k\n",
    "    png('${_output[3]}', width = 4, height = 4, unit='in', res=300)\n",
    "    qqplot(qchisq(ppoints(100), df=k), res$pc$mahal, main = expression(\"Mahalanobis\" * ~D^2 * \" vs. quantiles of\" * ~ chi[k]^2), xlab = expression(chi[2]^2 * \", probability points = 100\"), ylab = expression(D^2), pch=16)\n",
    "    dev.off() \n",
    "    png('${_output[4]}', width = 4, height = 4, unit='in', res=300)\n",
    "    ggplot(res$pc, aes(x=mahal)) + geom_histogram(aes(y = ..count..), binwidth = 0.5, colour = \"#1F3552\", fill = \"#4271AE\") + scale_x_continuous(name = \"Mahalanobis distance\") + theme_classic()\n",
    "    dev.off()\n",
    "  \n",
    "    # Save results and outliers\n",
    "    saveRDS(res,${_output[0]:r})\n",
    "    write.table(res$outliers, ${_output[1]:r}, sep=\"\\t\", quote=FALSE, row.names=FALSE, col.names=FALSE)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.1"
  },
  "sos": {
   "kernels": [
    [
     "Bash",
     "calysto_bash",
     "Bash",
     "#E6EEFF",
     ""
    ],
    [
     "SoS",
     "sos",
     "",
     "",
     "sos"
    ]
   ],
   "version": "0.22.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
