<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="ipynb_website:version" content="0.9.7" />
<meta name="viewport" content="width=device-width, initial-scale=1" />

<link rel="stylesheet" type="text/css" href="../css/jt.css">

<link rel="stylesheet" type="text/css" href="../css/toc2.css">

<link href="../site_libs/jqueryui-1.11.4/jquery-ui.css">
<link rel="stylesheet" href="../site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<link rel="stylesheet" href="../site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jqueryui/1.9.1/jquery-ui.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>

<link rel="stylesheet"
      href="../site_libs/highlightjs/null.min.css"
      type="text/css" />

<script src="../site_libs/highlightjs/highlight.pack.js"></script>
<script>hljs.initHighlightingOnLoad();</script>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>

<script src="../js/doc_toc.js"></script>
<script src="../js/docs.js"></script>

<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.2/MathJax.js?config=TeX-MML-AM_CHTML"></script>
<script>
    MathJax.Hub.Config({
        extensions: ["tex2jax.js"],
        jax: ["input/TeX", "output/HTML-CSS"],
        tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true
        },
        "HTML-CSS": {
            preferredFont: "TeX",
            availableFonts: ["TeX"],
            styles: {
                scale: 110,
                ".MathJax_Display": {
                    "font-size": "110%",
                }
            }
        }
    });
</script>
<script>
function filterDataFrame(id) {
    var input = document.getElementById("search_" + id);
    var filter = input.value.toUpperCase();
    var table = document.getElementById("dataframe_" + id);
    var tr = table.getElementsByTagName("tr");
    // Loop through all table rows, and hide those who don't match the search query
    for (var i = 1; i < tr.length; i++) {
        for (var j = 0; j < tr[i].cells.length; ++j) {
            var matched = false;
            if (tr[i].cells[j].innerHTML.toUpperCase().indexOf(filter) != -1) {
                tr[i].style.display = "";
                matched = true
                break;
            }
            if (!matched)
                tr[i].style.display = "none";
        }
    }
}
function sortDataFrame(id, n, dtype) {
    var table = document.getElementById("dataframe_" + id);
    var tb = table.tBodies[0]; // use `<tbody>` to ignore `<thead>` and `<tfoot>` rows
    var tr = Array.prototype.slice.call(tb.rows, 0); // put rows into array
    if (dtype === 'numeric') {
        var fn = function(a, b) { 
            return parseFloat(a.cells[n].textContent) <= parseFloat(b.cells[n].textContent) ? -1 : 1;
        }
    } else {
        var fn = function(a, b) {
            var c = a.cells[n].textContent.trim().localeCompare(b.cells[n].textContent.trim()); 
            return c > 0 ? 1 : (c < 0 ? -1 : 0) }
    }
    var isSorted = function(array, fn) {
        if (array.length < 2)
            return 1;
        var direction = fn(array[0], array[1]); 
        for (var i = 1; i < array.length - 1; ++i) {
            var d = fn(array[i], array[i+1]);
            if (d == 0)
                continue;
            else if (direction == 0)
                direction = d;
            else if (direction != d)
                return 0;
            }
        return direction;
    }
    var sorted = isSorted(tr, fn);
    if (sorted == 1 || sorted == -1) {
        // if sorted already, reverse it
        for(var i = tr.length - 1; i >= 0; --i)
            tb.appendChild(tr[i]); // append each row in order
    } else {
        tr = tr.sort(fn);
        for(var i = 0; i < tr.length; ++i)
            tb.appendChild(tr[i]); // append each row in order
    }
}
</script>

<script>
$( document ).ready(function(){
            var cfg={'threshold':3,     // depth of toc (number of levels)
             'number_sections': false,
             'toc_cell': false,          // useless here
             'toc_window_display': true, // display the toc window
             "toc_section_display": "block", // display toc contents in the window
             'sideBar':true,       // sidebar or floating window
             'navigate_menu':false       // navigation menu (only in liveNotebook -- do not change)
            }
            var st={};                  // some variables used in the script
            st.rendering_toc_cell = false;
            st.config_loaded = false;
            st.extension_initialized=false;
            st.nbcontainer_marginleft = $('#notebook-container').css('margin-left')
            st.nbcontainer_marginright = $('#notebook-container').css('margin-right')
            st.nbcontainer_width = $('#notebook-container').css('width')
            st.oldTocHeight = undefined
            st.cell_toc = undefined;
            st.toc_index=0;
            // fire the main function with these parameters
            table_of_contents(cfg, st);
            var file=GWASDict[$("h1:first").attr("id")];
            $("#toc-level0 a").css("color","#126dce");
            $('a[href="#'+$("h1:first").attr("id")+'"]').hide()
            var docs=GWASArray;
            var docs_map=GWASArrayMap;
            var pos=GWASArray.indexOf(file);
            for (var a=pos;a>=0;a--){
                  $('<li><a href="'+docs[a]+'.html"><font color="#073642"><b>'+docs_map[docs[a]].replace(/_/g," ")+'</b></font></a></li>').insertBefore("#toc-level0 li:eq(0)");
            }
            $('a[href="'+file+'.html'+'"]').css("color","#126dce");
            for (var a=pos+1;a<docs.length;a++){
                  $(".toc #toc-level0").append('<li><a href="'+docs[a]+'.html"><font color="#073642"><b>'+docs_map[docs[a]].replace(/_/g," ")+'</b></font></a></li>');
            }
            // $("#toc-header").hide(); // comment out because it prevents search bar from displaying
    });
</script>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');
  // mark it active
  menuAnchor.parent().addClass('active');
  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>
<div class="container-fluid main-container">
<!-- tabsets -->
<script src="../site_libs/navigation-1.1/tabsets.js"></script>
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>



<title>Some computational genomics workflows</title>

<style type = "text/css">
body {
  font-family: "Droid Sans";
  padding-top: 66px;
  padding-bottom: 40px;
}
</style>
</head>

<body>
<div tabindex="-1" id="notebook" class="border-box-sizing">
<div class="container" id="notebook-container">

<!-- code folding -->

<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">Some computational genomics workflows</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
<li>
  <a href="../index.html">Overview</a>
</li>
        
      </ul>
        
<ul class="nav navbar-nav navbar-right">
<li>
   <a href="http://github.com/cumc/bioworkflows"> <span class="fa fa-github"></span> </a>
</li>
</ul>
        
      </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="GWAS-data-QC-workflow">GWAS data QC workflow<a class="anchor-link" href="#GWAS-data-QC-workflow">&#182;</a></h1><p>This workflow implements some prelimary data QC steps for PLINK input files.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Overview">Overview<a class="anchor-link" href="#Overview">&#182;</a></h2><p>This notebook includes workflow for</p>
<ul>
<li>Computer kinship matrix in sample and estimate related individuals</li>
<li>Genotype and sample QC: by MAF, missing data and HWE</li>
<li>LD pruning</li>
</ul>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Run-this-workflow">Run this workflow<a class="anchor-link" href="#Run-this-workflow">&#182;</a></h2><p>Depending on the context of your problem, the workflow can be executed in two ways:</p>
<ol>
<li>Run <code>merge_plink</code> if necessary, to merge all samples first; then run <code>king</code> to perform kinship estimate and finally <code>qc</code> to do addition QC</li>
<li>When you have a separate data-set for kinship estimate different from your genotype of interest, you can run <code>king</code>, followed by <code>qc</code>.</li>
</ol>
<p>In both cases, you should use the <code>*.related_remove</code> output from <code>king</code> as the <code>--remove_samples</code> parameter input for <code>qc</code> step.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Minimal-working-example">Minimal working example<a class="anchor-link" href="#Minimal-working-example">&#182;</a></h2><p>FIXME: first specify which of the 2 scenarios this example is for, then show how to run it.</p>
<h3 id="First-scenario:-estimate-kinship">First scenario: estimate kinship<a class="anchor-link" href="#First-scenario:-estimate-kinship">&#182;</a></h3>
<pre><code>sos run ~/bioworkflows/GWAS/GWAS_QC.ipynb king\
    --cwd ~/output \
    --genoFile ~/MWE_AD/rename_chr22.bed \
    --name first \ 
    --kinship 0.05</code></pre>
<h3 id="First-scenario:-do-qc">First scenario: do qc<a class="anchor-link" href="#First-scenario:-do-qc">&#182;</a></h3>
<pre><code>sos run ~/bioworkflows/GWAS/GWAS_QC.ipynb qc\
    --cwd ~/output \
    --genoFile ~/MWE_AD/rename_chr22.bed \
    --remove_samples ~/output/rename_chr22.first.related_remove \
    --maf_filter 0.5 \
    --geno_filter 0.2 \
    --mind_filter 0.1 \
    --hwe_filter 0.0 \
    --name first \
    --window 50 \
    --shift 10 \
    --r2 0.5</code></pre>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[1]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">GWAS_QC</span><span class="o">.</span><span class="n">ipynb</span> <span class="o">-</span><span class="n">h</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

    <div class="prompt"></div>


<div class="output_subarea output_stream output_stdout output_text">
<pre>usage: sos run GWAS_QC.ipynb [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &#34;sos run -h&#34; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  king
  qc
  merge_plink

Global Workflow Options:
  --cwd VAL (as path, required)
                        the output directory for generated files
  --name VAL (as str, required)
                        A string to identify your analysis run
  --job-size 1 (as int)
                        For cluster jobs, number commands to run per job
  --walltime 5h
                        Wall clock time expected
  --mem 16G
                        Memory expected
  --numThreads 20 (as int)
                        Number of threads
  --container-lmm &#39;statisticalgenetics/lmm:1.8&#39;
                        Software container option

Sections
  king:                 Inference of relationships in the sample to remove
                        closely related individuals
    Workflow Options:
      --kinship 0.0625 (as float)
                        Filter based on kinship coefficient higher than a number
                        (e.g first degree 0.25, second degree 0.125, third
                        degree 0.0625)
      --genoFile VAL (as path, required)
                        Plink binary file
  qc_1:                 Filter SNPs and select individuals
    Workflow Options:
      --remove-samples . (as path)
                        The path to the file that contains the list of samples
                        to remove (format FID, IID)
      --keep-samples . (as path)
                        The path to the file that contains the list of samples
                        to keep (format FID, IID)
      --keep-variants . (as path)
                        The path to the file that contains the list of variants
                        to keep
      --maf-filter 0.01 (as float)
                        MAF filter to use
      --geno-filter 0.01 (as float)
                        Maximum missingess per-variant
      --mind-filter 0.02 (as float)
                        Maximum missingness per-sample
      --hwe-filter 5e-08 (as float)
                        HWE filter
      --genoFile  paths

                        Plink binary files
  qc_2:                 LD prunning and remove related individuals (both ind of
                        a pair)
    Workflow Options:
      --window 50 (as int)
                        Window size
      --shift 10 (as int)
                        Shift window every 10 snps
      --r2 0.1 (as float)
  qc_3:                 Merge all the .bed files into one bed file
    Workflow Options:
      --merged-prefix &#39;&#39;
  merge_plink:
    Workflow Options:
      --merged-prefix VAL (as str, required)
      --genoFile  paths

</pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># the output directory for generated files</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># A string to identify your analysis run</span>
<span class="kn">parameter:</span> <span class="kp">name</span> <span class="o">=</span> <span class="n">f</span><span class="s2">&quot;{cwd:b}&quot;</span>
<span class="c1"># For cluster jobs, number commands to run per job</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">1</span>
<span class="c1"># Wall clock time expected</span>
<span class="kn">parameter:</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s2">&quot;5h&quot;</span>
<span class="c1"># Memory expected</span>
<span class="kn">parameter:</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s2">&quot;16G&quot;</span>
<span class="c1"># Number of threads</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">20</span>
<span class="c1"># Software container option</span>
<span class="kn">parameter:</span> <span class="n">container_lmm</span> <span class="o">=</span> <span class="s1">&#39;statisticalgenetics/lmm:1.8&#39;</span>
<span class="c1"># use this function to edit memory string for PLINK input</span>
<span class="kn">from</span> <span class="nn">sos.utils</span> <span class="kn">import</span> <span class="n">expand_size</span>
<span class="n">cwd</span> <span class="o">=</span> <span class="n">f</span><span class="s2">&quot;{cwd:a}&quot;</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Estimate-kinship-in-the-sample">Estimate kinship in the sample<a class="anchor-link" href="#Estimate-kinship-in-the-sample">&#182;</a></h2><p>The output is a list of related individuals, as well as the kinship matrix</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># Inference of relationships in the sample to remove closely related individuals</span>
<span class="p">[</span><span class="n">king_1</span><span class="p">]</span>
<span class="c1"># Plink binary file</span>
<span class="kn">parameter:</span> <span class="n">genoFile</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">input:</span> <span class="n">genoFile</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{cwd}/{_input:bn}.{name}.kin0&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{step_name}_{_output:bn}&#39;</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container_lmm</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output}.stdout&#39;</span>
    <span class="n">plink2</span> \
      <span class="o">--</span><span class="n">bfile</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">n</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">make</span><span class="o">-</span><span class="n">king</span><span class="o">-</span><span class="n">table</span> \
      <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">threads</span> <span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">memory</span> <span class="err">$</span><span class="p">{</span><span class="nb">int</span><span class="p">(</span><span class="n">expand_size</span><span class="p">(</span><span class="kp">mem</span><span class="p">)</span> <span class="o">*</span> <span class="mf">0.9</span><span class="p">)}</span>
      
<span class="p">[</span><span class="n">king_2</span><span class="p">]</span>
<span class="c1"># Filter based on kinship coefficient higher than a number (e.g first degree 0.25, second degree 0.125, third degree 0.0625)</span>
<span class="kn">parameter:</span> <span class="n">kinship</span> <span class="o">=</span> <span class="mf">0.0625</span>
<span class="c1"># If set to true, the unrelated individuals in a family will be kept without being reported. </span>
<span class="c1"># Otherwise (use `--no-maximize-unrelated`) the entire family will be removed</span>
<span class="kn">parameter:</span> <span class="n">maximize_unrelated</span> <span class="o">=</span> <span class="bp">True</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{_input:n}.related_id&#39;</span>
<span class="kn">R:</span>  <span class="n">container</span><span class="o">=</span><span class="n">container_lmm</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output}.stdout&#39;</span>
    <span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">igraph</span><span class="p">)</span>
    <span class="c1"># Remove related individuals while keeping maximum number of individuals</span>
    <span class="c1"># this function is simplified from: </span>
    <span class="c1"># https://rdrr.io/cran/plinkQC/src/R/utils.R</span>
    <span class="c1">#&#39; @param relatedness [data.frame] containing pair-wise relatedness estimates</span>
    <span class="c1">#&#39; (in column [relatednessRelatedness]) for individual 1 (in column</span>
    <span class="c1">#&#39; [relatednessIID1] and individual 2 (in column [relatednessIID1]). Columns</span>
    <span class="c1">#&#39; relatednessIID1, relatednessIID2 and relatednessRelatedness have to present,</span>
    <span class="c1">#&#39; while additional columns such as family IDs can be present. Default column</span>
    <span class="c1">#&#39; names correspond to column names in output of plink --genome</span>
    <span class="c1">#&#39; (\url{https://www.cog-genomics.org/plink/1.9/ibd}). All original</span>
    <span class="c1">#&#39; columns for pair-wise highIBDTh fails will be returned in fail_IBD.</span>
    <span class="c1">#&#39; @param relatednessTh [double] Threshold for filtering related individuals.</span>
    <span class="c1">#&#39; Individuals, whose pair-wise relatedness estimates are greater than this</span>
    <span class="c1">#&#39; threshold are considered related.</span>
    <span class="n">relatednessFilter</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">relatedness</span><span class="p">,</span> 
                                  <span class="n">relatednessTh</span><span class="p">,</span>
                                  <span class="n">relatednessIID1</span><span class="o">=</span><span class="s2">&quot;IID1&quot;</span><span class="p">,</span> 
                                  <span class="n">relatednessIID2</span><span class="o">=</span><span class="s2">&quot;IID2&quot;</span><span class="p">,</span>
                                  <span class="n">relatednessRelatedness</span><span class="o">=</span><span class="s2">&quot;KINSHIP&quot;</span><span class="p">)</span> <span class="p">{</span>
        <span class="c1"># format data</span>
        <span class="k">if</span> <span class="p">(</span><span class="err">!</span><span class="p">(</span><span class="n">relatednessIID1</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">names</span><span class="p">(</span><span class="n">relatedness</span><span class="p">)))</span> <span class="p">{</span>
            <span class="n">stop</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">&quot;Column&quot;</span><span class="p">,</span> <span class="n">relatednessIID1</span><span class="p">,</span> <span class="s2">&quot;for relatedness not found!&quot;</span><span class="p">))</span>
        <span class="p">}</span>
        <span class="k">if</span> <span class="p">(</span><span class="err">!</span><span class="p">(</span><span class="n">relatednessIID2</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">names</span><span class="p">(</span><span class="n">relatedness</span><span class="p">)))</span> <span class="p">{</span>
            <span class="n">stop</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">&quot;Column&quot;</span><span class="p">,</span> <span class="n">relatednessIID1</span><span class="p">,</span> <span class="s2">&quot;for relatedness not found!&quot;</span><span class="p">))</span>
        <span class="p">}</span>
        <span class="k">if</span> <span class="p">(</span><span class="err">!</span><span class="p">(</span><span class="n">relatednessRelatedness</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">names</span><span class="p">(</span><span class="n">relatedness</span><span class="p">)))</span> <span class="p">{</span>
            <span class="n">stop</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">&quot;Column&quot;</span><span class="p">,</span> <span class="n">relatednessRelatedness</span><span class="p">,</span>
                       <span class="s2">&quot;for relatedness not found!&quot;</span><span class="p">))</span>
        <span class="p">}</span>

        <span class="n">iid1_index</span> <span class="o">&lt;-</span> <span class="n">which</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">relatedness</span><span class="p">)</span> <span class="o">==</span> <span class="n">relatednessIID1</span><span class="p">)</span>
        <span class="n">iid2_index</span> <span class="o">&lt;-</span> <span class="n">which</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">relatedness</span><span class="p">)</span> <span class="o">==</span> <span class="n">relatednessIID2</span><span class="p">)</span>

        <span class="n">relatedness</span><span class="p">[,</span><span class="n">iid1_index</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">character</span><span class="p">(</span><span class="n">relatedness</span><span class="p">[,</span><span class="n">iid1_index</span><span class="p">])</span>
        <span class="n">relatedness</span><span class="p">[,</span><span class="n">iid2_index</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">character</span><span class="p">(</span><span class="n">relatedness</span><span class="p">[,</span><span class="n">iid2_index</span><span class="p">])</span>

        <span class="n">relatedness_names</span> <span class="o">&lt;-</span> <span class="n">names</span><span class="p">(</span><span class="n">relatedness</span><span class="p">)</span>
        <span class="n">names</span><span class="p">(</span><span class="n">relatedness</span><span class="p">)[</span><span class="n">iid1_index</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="s2">&quot;IID1&quot;</span>
        <span class="n">names</span><span class="p">(</span><span class="n">relatedness</span><span class="p">)[</span><span class="n">iid2_index</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="s2">&quot;IID2&quot;</span>
        <span class="n">names</span><span class="p">(</span><span class="n">relatedness</span><span class="p">)[</span><span class="n">names</span><span class="p">(</span><span class="n">relatedness</span><span class="p">)</span> <span class="o">==</span> <span class="n">relatednessRelatedness</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="s2">&quot;M&quot;</span>

        <span class="c1"># Remove symmetric IID rows</span>
        <span class="n">relatedness_original</span> <span class="o">&lt;-</span> <span class="n">relatedness</span>
        <span class="n">relatedness</span> <span class="o">&lt;-</span> <span class="n">dplyr</span><span class="p">::</span><span class="n">select_</span><span class="p">(</span><span class="n">relatedness</span><span class="p">,</span> <span class="o">~</span><span class="n">IID1</span><span class="p">,</span> <span class="o">~</span><span class="n">IID2</span><span class="p">,</span> <span class="o">~</span><span class="n">M</span><span class="p">)</span>

        <span class="n">sortedIDs</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">t</span><span class="p">(</span><span class="nb">apply</span><span class="p">(</span><span class="n">relatedness</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">pair</span><span class="p">)</span> <span class="p">{</span>
            <span class="n">c</span><span class="p">(</span><span class="n">sort</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="n">pair</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">pair</span><span class="p">[</span><span class="mi">2</span><span class="p">])))</span>
            <span class="p">})),</span> <span class="n">stringsAsFactors</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
        <span class="n">keepIndex</span> <span class="o">&lt;-</span> <span class="n">which</span><span class="p">(</span><span class="err">!</span><span class="n">duplicated</span><span class="p">(</span><span class="n">sortedIDs</span><span class="p">))</span>

        <span class="n">relatedness_original</span> <span class="o">&lt;-</span> <span class="n">relatedness_original</span><span class="p">[</span><span class="n">keepIndex</span><span class="p">,]</span>
        <span class="n">relatedness</span> <span class="o">&lt;-</span> <span class="n">relatedness</span><span class="p">[</span><span class="n">keepIndex</span><span class="p">,]</span>

        <span class="c1"># individuals with at least one pair-wise comparison &gt; relatednessTh</span>
        <span class="c1"># return NULL to failIDs if no one fails the relatedness check</span>
        <span class="n">highRelated</span> <span class="o">&lt;-</span> <span class="n">dplyr</span><span class="p">::</span><span class="n">filter_</span><span class="p">(</span><span class="n">relatedness</span><span class="p">,</span> <span class="o">~</span><span class="n">M</span> <span class="o">&gt;</span> <span class="n">relatednessTh</span><span class="p">)</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">nrow</span><span class="p">(</span><span class="n">highRelated</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="p">{</span>
            <span class="k">return</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">relatednessFails</span><span class="o">=</span><span class="n">NULL</span><span class="p">,</span> <span class="n">failIDs</span><span class="o">=</span><span class="n">NULL</span><span class="p">))</span>
        <span class="p">}</span>

        <span class="c1"># all samples with related individuals</span>
        <span class="n">allRelated</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="n">highRelated</span><span class="err">$</span><span class="n">IID1</span><span class="p">,</span> <span class="n">highRelated</span><span class="err">$</span><span class="n">IID2</span><span class="p">)</span>
        <span class="n">uniqueIIDs</span> <span class="o">&lt;-</span> <span class="n">unique</span><span class="p">(</span><span class="n">allRelated</span><span class="p">)</span>

        <span class="c1"># Further selection of samples with relatives in cohort</span>
        <span class="n">multipleRelative</span> <span class="o">&lt;-</span> <span class="n">unique</span><span class="p">(</span><span class="n">allRelated</span><span class="p">[</span><span class="n">duplicated</span><span class="p">(</span><span class="n">allRelated</span><span class="p">)])</span>
        <span class="n">singleRelative</span> <span class="o">&lt;-</span> <span class="n">uniqueIIDs</span><span class="p">[</span><span class="err">!</span><span class="n">uniqueIIDs</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">multipleRelative</span><span class="p">]</span>

        <span class="n">highRelatedMultiple</span> <span class="o">&lt;-</span> <span class="n">highRelated</span><span class="p">[</span><span class="n">highRelated</span><span class="err">$</span><span class="n">IID1</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">multipleRelative</span> <span class="o">|</span>
                                            <span class="n">highRelated</span><span class="err">$</span><span class="n">IID2</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">multipleRelative</span><span class="p">,]</span>
        <span class="n">highRelatedSingle</span> <span class="o">&lt;-</span> <span class="n">highRelated</span><span class="p">[</span><span class="n">highRelated</span><span class="err">$</span><span class="n">IID1</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">singleRelative</span> <span class="o">&amp;</span>
                                           <span class="n">highRelated</span><span class="err">$</span><span class="n">IID2</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">singleRelative</span><span class="p">,]</span>

        <span class="c1"># Only one related samples per individual</span>
        <span class="k">if</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">singleRelative</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span> <span class="p">{</span>
          <span class="c1"># randomly choose one to exclude</span>
          <span class="n">failIDs_single</span> <span class="o">&lt;-</span> <span class="n">highRelatedSingle</span><span class="p">[,</span><span class="mi">1</span><span class="p">]</span>
            
        <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
          <span class="n">failIDs_single</span> <span class="o">&lt;-</span> <span class="n">NULL</span>
        <span class="p">}</span>
  
        <span class="c1"># An individual has multiple relatives</span>
        <span class="k">if</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">multipleRelative</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span> <span class="p">{</span>
            <span class="n">relatedPerID</span> <span class="o">&lt;-</span> <span class="n">lapply</span><span class="p">(</span><span class="n">multipleRelative</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="p">{</span>
                <span class="n">tmp</span> <span class="o">&lt;-</span> <span class="n">highRelatedMultiple</span><span class="p">[</span><span class="n">rowSums</span><span class="p">(</span>
                    <span class="n">cbind</span><span class="p">(</span><span class="n">highRelatedMultiple</span><span class="err">$</span><span class="n">IID1</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">x</span><span class="p">,</span>
                          <span class="n">highRelatedMultiple</span><span class="err">$</span><span class="n">IID2</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">x</span><span class="p">))</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span>
                <span class="n">rel</span> <span class="o">&lt;-</span> <span class="n">unique</span><span class="p">(</span><span class="n">unlist</span><span class="p">(</span><span class="n">tmp</span><span class="p">))</span>
                <span class="k">return</span><span class="p">(</span><span class="n">rel</span><span class="p">)</span>
            <span class="p">})</span>
            <span class="n">names</span><span class="p">(</span><span class="n">relatedPerID</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">multipleRelative</span>

            <span class="n">keepIDs_multiple</span> <span class="o">&lt;-</span> <span class="n">lapply</span><span class="p">(</span><span class="n">relatedPerID</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="p">{</span>
                <span class="n">pairwise</span> <span class="o">&lt;-</span> <span class="n">t</span><span class="p">(</span><span class="n">combn</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
                <span class="n">index</span> <span class="o">&lt;-</span> <span class="p">(</span><span class="n">highRelatedMultiple</span><span class="err">$</span><span class="n">IID1</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">pairwise</span><span class="p">[,</span><span class="mi">1</span><span class="p">]</span> <span class="o">&amp;</span>
                              <span class="n">highRelatedMultiple</span><span class="err">$</span><span class="n">IID2</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">pairwise</span><span class="p">[,</span><span class="mi">2</span><span class="p">])</span> <span class="o">|</span>
                    <span class="p">(</span><span class="n">highRelatedMultiple</span><span class="err">$</span><span class="n">IID1</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">pairwise</span><span class="p">[,</span><span class="mi">2</span><span class="p">]</span> <span class="o">&amp;</span>
                         <span class="n">highRelatedMultiple</span><span class="err">$</span><span class="n">IID2</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">pairwise</span><span class="p">[,</span><span class="mi">1</span><span class="p">])</span>
                <span class="n">combination</span> <span class="o">&lt;-</span> <span class="n">highRelatedMultiple</span><span class="p">[</span><span class="n">index</span><span class="p">,]</span>
                <span class="n">combination_graph</span> <span class="o">&lt;-</span> <span class="n">igraph</span><span class="p">::</span><span class="n">graph_from_data_frame</span><span class="p">(</span><span class="n">combination</span><span class="p">,</span>
                                                                   <span class="n">directed</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
                <span class="n">all_iv_set</span> <span class="o">&lt;-</span> <span class="n">igraph</span><span class="p">::</span><span class="n">ivs</span><span class="p">(</span><span class="n">combination_graph</span><span class="p">)</span>
                <span class="n">length_iv_set</span> <span class="o">&lt;-</span> <span class="n">sapply</span><span class="p">(</span><span class="n">all_iv_set</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="n">length</span><span class="p">(</span><span class="n">x</span><span class="p">))</span>

                <span class="k">if</span> <span class="p">(</span><span class="nb">all</span><span class="p">(</span><span class="n">length_iv_set</span> <span class="o">==</span> <span class="mi">1</span><span class="p">))</span> <span class="p">{</span>
                    <span class="c1"># check how often they occurr elsewhere</span>
                    <span class="n">occurrence</span> <span class="o">&lt;-</span> <span class="n">sapply</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="nb">id</span><span class="p">)</span> <span class="p">{</span>
                        <span class="nb">sum</span><span class="p">(</span><span class="n">sapply</span><span class="p">(</span><span class="n">relatedPerID</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">idlist</span><span class="p">)</span> <span class="nb">id</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">idlist</span><span class="p">))</span>
                    <span class="p">})</span>
                    <span class="c1"># if occurrence the same everywhere, pick the first, else keep</span>
                    <span class="c1"># the one with minimum occurrence elsewhere</span>
                    <span class="k">if</span> <span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">occurrence</span><span class="p">))</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span> <span class="p">{</span>
                        <span class="n">nonRelated</span> <span class="o">&lt;-</span> <span class="n">sort</span><span class="p">(</span><span class="n">x</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
                    <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
                        <span class="n">nonRelated</span> <span class="o">&lt;-</span> <span class="n">names</span><span class="p">(</span><span class="n">occurrence</span><span class="p">)[</span><span class="n">which</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">occurrence</span><span class="p">)]</span>
                    <span class="p">}</span>
                <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
                    <span class="n">nonRelated</span> <span class="o">&lt;-</span> <span class="n">all_iv_set</span><span class="p">[</span><span class="n">which</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">length_iv_set</span><span class="p">)]</span>
                <span class="p">}</span>
                <span class="k">return</span><span class="p">(</span><span class="n">nonRelated</span><span class="p">)</span>
            <span class="p">})</span>
            <span class="n">keepIDs_multiple</span> <span class="o">&lt;-</span> <span class="n">unique</span><span class="p">(</span><span class="n">unlist</span><span class="p">(</span><span class="n">keepIDs_multiple</span><span class="p">))</span>
            <span class="n">failIDs_multiple</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="n">multipleRelative</span><span class="p">[</span><span class="err">!</span><span class="n">multipleRelative</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span>
                                                       <span class="n">keepIDs_multiple</span><span class="p">])</span>
        <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
            <span class="n">failIDs_multiple</span> <span class="o">&lt;-</span> <span class="n">NULL</span>
        <span class="p">}</span>
        <span class="n">allFailIIDs</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="n">failIDs_single</span><span class="p">,</span> <span class="n">failIDs_multiple</span><span class="p">)</span>
        <span class="n">relatednessFails</span> <span class="o">&lt;-</span> <span class="n">lapply</span><span class="p">(</span><span class="n">allFailIIDs</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="nb">id</span><span class="p">)</span> <span class="p">{</span>
            <span class="n">fail_inorder</span> <span class="o">&lt;-</span> <span class="n">relatedness_original</span><span class="err">$</span><span class="n">IID1</span> <span class="o">==</span> <span class="nb">id</span> <span class="o">&amp;</span>
                <span class="n">relatedness_original</span><span class="err">$</span><span class="n">M</span> <span class="o">&gt;</span> <span class="n">relatednessTh</span>
            <span class="n">fail_inreverse</span> <span class="o">&lt;-</span> <span class="n">relatedness_original</span><span class="err">$</span><span class="n">IID2</span> <span class="o">==</span> <span class="nb">id</span> <span class="o">&amp;</span>
                <span class="n">relatedness_original</span><span class="err">$</span><span class="n">M</span> <span class="o">&gt;</span> <span class="n">relatednessTh</span>
            <span class="k">if</span> <span class="p">(</span><span class="nb">any</span><span class="p">(</span><span class="n">fail_inreverse</span><span class="p">))</span> <span class="p">{</span>
                <span class="n">inreverse</span> <span class="o">&lt;-</span> <span class="n">relatedness_original</span><span class="p">[</span><span class="n">fail_inreverse</span><span class="p">,</span> <span class="p">]</span>
                <span class="n">id1</span> <span class="o">&lt;-</span> <span class="n">iid1_index</span>
                <span class="n">id2</span> <span class="o">&lt;-</span> <span class="n">iid2_index</span>
                <span class="n">inreverse</span><span class="p">[,</span><span class="n">c</span><span class="p">(</span><span class="n">id1</span><span class="p">,</span> <span class="n">id2</span><span class="p">)]</span> <span class="o">&lt;-</span> <span class="n">inreverse</span><span class="p">[,</span><span class="n">c</span><span class="p">(</span><span class="n">id2</span><span class="p">,</span> <span class="n">id1</span><span class="p">)]</span>
                <span class="n">names</span><span class="p">(</span><span class="n">inreverse</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">relatedness_names</span>
            <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
                <span class="n">inreverse</span> <span class="o">&lt;-</span> <span class="n">NULL</span>
            <span class="p">}</span>
            <span class="n">inorder</span> <span class="o">&lt;-</span> <span class="n">relatedness_original</span><span class="p">[</span><span class="n">fail_inorder</span><span class="p">,</span> <span class="p">]</span>
            <span class="n">names</span><span class="p">(</span><span class="n">inorder</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">relatedness_names</span>
            <span class="k">return</span><span class="p">(</span><span class="n">rbind</span><span class="p">(</span><span class="n">inorder</span><span class="p">,</span> <span class="n">inreverse</span><span class="p">))</span>
        <span class="p">})</span>
        <span class="n">relatednessFails</span> <span class="o">&lt;-</span> <span class="n">do</span><span class="o">.</span><span class="n">call</span><span class="p">(</span><span class="n">rbind</span><span class="p">,</span> <span class="n">relatednessFails</span><span class="p">)</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">nrow</span><span class="p">(</span><span class="n">relatednessFails</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="p">{</span>
            <span class="n">relatednessFails</span> <span class="o">&lt;-</span> <span class="n">NULL</span>
            <span class="n">failIDs</span> <span class="o">&lt;-</span> <span class="n">NULL</span>
        <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
            <span class="n">names</span><span class="p">(</span><span class="n">relatednessFails</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">relatedness_names</span>
            <span class="n">rownames</span><span class="p">(</span><span class="n">relatednessFails</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="mi">1</span><span class="p">:</span><span class="n">nrow</span><span class="p">(</span><span class="n">relatednessFails</span><span class="p">)</span>
            <span class="n">uniqueFails</span> <span class="o">&lt;-</span> <span class="n">relatednessFails</span><span class="p">[</span><span class="err">!</span><span class="n">duplicated</span><span class="p">(</span><span class="n">relatednessFails</span><span class="p">[,</span><span class="n">iid1_index</span><span class="p">]),]</span>
            <span class="n">failIDs</span> <span class="o">&lt;-</span> <span class="n">uniqueFails</span><span class="p">[,</span><span class="n">iid1_index</span><span class="p">]</span>
        <span class="p">}</span>
        <span class="k">return</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">relatednessFails</span><span class="o">=</span><span class="n">relatednessFails</span><span class="p">,</span> <span class="n">failIDs</span><span class="o">=</span><span class="n">failIDs</span><span class="p">))</span>
    <span class="p">}</span>
    
  
    <span class="c1"># main code</span>
    <span class="n">kin0</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">},</span> <span class="n">header</span><span class="o">=</span><span class="n">F</span><span class="p">,</span> <span class="n">stringsAsFactor</span><span class="o">=</span><span class="n">F</span><span class="p">)</span>
    <span class="n">colnames</span><span class="p">(</span><span class="n">kin0</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;FID1&quot;</span><span class="p">,</span><span class="s2">&quot;ID1&quot;</span><span class="p">,</span><span class="s2">&quot;FID2&quot;</span><span class="p">,</span><span class="s2">&quot;ID2&quot;</span><span class="p">,</span><span class="s2">&quot;NSNP&quot;</span><span class="p">,</span><span class="s2">&quot;HETHET&quot;</span><span class="p">,</span><span class="s2">&quot;IBS0&quot;</span><span class="p">,</span><span class="s2">&quot;KINSHIP&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="s2">&quot;TRUE&quot;</span> <span class="k">if</span> <span class="n">maximize_unrelated</span> <span class="k">else</span> <span class="s2">&quot;FALSE&quot;</span><span class="p">})</span> <span class="p">{</span>
        <span class="n">rel</span> <span class="o">&lt;-</span> <span class="n">relatednessFilter</span><span class="p">(</span><span class="n">kin0</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">kinship</span><span class="p">},</span> <span class="s2">&quot;ID1&quot;</span><span class="p">,</span> <span class="s2">&quot;ID2&quot;</span><span class="p">,</span> <span class="s2">&quot;KINSHIP&quot;</span><span class="p">)</span><span class="err">$</span><span class="n">failIDs</span>
        <span class="n">tmp1</span> <span class="o">&lt;-</span> <span class="n">kin0</span><span class="p">[,</span><span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span>
        <span class="n">tmp2</span> <span class="o">&lt;-</span> <span class="n">kin0</span><span class="p">[,</span><span class="mi">3</span><span class="p">:</span><span class="mi">4</span><span class="p">]</span>
        <span class="n">colnames</span><span class="p">(</span><span class="n">tmp1</span><span class="p">)</span> <span class="o">=</span> <span class="n">colnames</span><span class="p">(</span><span class="n">tmp2</span><span class="p">)</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;FID&quot;</span><span class="p">,</span> <span class="s2">&quot;ID&quot;</span><span class="p">)</span>
        <span class="c1"># Get the family ID for these rels so there are two columns FID and IID in the output</span>
        <span class="n">lookup</span> <span class="o">&lt;-</span> <span class="n">dplyr</span><span class="p">::</span><span class="n">distinct</span><span class="p">(</span><span class="n">rbind</span><span class="p">(</span><span class="n">tmp1</span><span class="p">,</span><span class="n">tmp2</span><span class="p">))</span>
        <span class="n">dat</span> <span class="o">&lt;-</span> <span class="n">lookup</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">lookup</span><span class="p">[,</span><span class="mi">2</span><span class="p">]</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">rel</span><span class="p">),]</span>
    <span class="p">}</span> <span class="k">else</span> <span class="p">{</span><span class="n">s</span>
        <span class="n">rel</span> <span class="o">&lt;-</span> <span class="n">kin0</span> <span class="o">%&gt;%</span> <span class="nb">filter</span><span class="p">(</span><span class="n">KINSHIP</span> <span class="o">&gt;=</span> <span class="err">$</span><span class="p">{</span><span class="n">kinship</span><span class="p">})</span>
        <span class="n">IID</span> <span class="o">&lt;-</span> <span class="n">sort</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">unlist</span><span class="p">(</span><span class="n">rel</span><span class="p">[,</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;ID1&quot;</span><span class="p">,</span> <span class="s2">&quot;ID2&quot;</span><span class="p">)])))</span>
        <span class="n">dat</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">IID</span><span class="p">)</span>
        <span class="n">dat</span> <span class="o">&lt;-</span> <span class="n">dat</span> <span class="o">%&gt;%</span>
            <span class="n">mutate</span><span class="p">(</span><span class="n">FID</span> <span class="o">=</span> <span class="n">IID</span><span class="p">)</span> <span class="o">%&gt;%</span>
            <span class="n">select</span><span class="p">(</span><span class="n">FID</span><span class="p">,</span> <span class="n">IID</span><span class="p">)</span>
    <span class="p">}</span>
    <span class="n">cat</span><span class="p">(</span><span class="s2">&quot;There are&quot;</span><span class="p">,</span> <span class="n">nrow</span><span class="p">(</span><span class="n">dat</span><span class="p">),</span><span class="s2">&quot;related individuals using a kinship threshold of ${kinship}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">write</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">dat</span><span class="p">,</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">},</span> <span class="n">quote</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span> <span class="n">col</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Genotype-and-sample-QC">Genotype and sample QC<a class="anchor-link" href="#Genotype-and-sample-QC">&#182;</a></h2><p>QC the genetic data based on MAF, sample and variant missigness and Hardy-Weinberg Equilibrium (HWE).</p>
<p>In this step you may also provide a list of samples to keep, for example in the case when you would like to subset a sample based on their ancestries to perform independent analyses on each of these groups.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[2]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># Filter SNPs and select individuals </span>
<span class="p">[</span><span class="n">qc_1</span> <span class="p">(</span><span class="n">basic</span> <span class="n">QC</span> <span class="n">filters</span><span class="p">)]</span>
<span class="c1"># The path to the file that contains the list of samples to remove (format FID, IID)</span>
<span class="kn">parameter:</span> <span class="n">remove_samples</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)</span>
<span class="c1"># The path to the file that contains the list of samples to keep (format FID, IID)</span>
<span class="kn">parameter:</span> <span class="n">keep_samples</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)</span>
<span class="c1"># The path to the file that contains the list of variants to keep</span>
<span class="kn">parameter:</span> <span class="n">keep_variants</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)</span>
<span class="c1"># minimum MAF filter to use. Notice that PLINK default is 0.01</span>
<span class="kn">parameter:</span> <span class="n">maf_filter</span> <span class="o">=</span> <span class="mf">0.01</span>
<span class="c1"># maximum MAF filter to use</span>
<span class="kn">parameter:</span> <span class="n">maf_max_filter</span> <span class="o">=</span> <span class="mf">0.0</span>
<span class="c1"># Maximum missingess per-variant</span>
<span class="kn">parameter:</span> <span class="n">geno_filter</span> <span class="o">=</span> <span class="mf">0.01</span>
<span class="c1"># Maximum missingness per-sample</span>
<span class="kn">parameter:</span> <span class="n">mind_filter</span> <span class="o">=</span> <span class="mf">0.02</span>
<span class="c1"># HWE filter </span>
<span class="kn">parameter:</span> <span class="n">hwe_filter</span> <span class="o">=</span> <span class="mf">5e-08</span>
<span class="c1"># Plink binary files</span>
<span class="kn">parameter:</span> <span class="n">genoFile</span> <span class="o">=</span> <span class="n">paths</span>

<span class="n">fail_if</span><span class="p">(</span><span class="ow">not</span> <span class="p">(</span><span class="n">keep_samples</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="ow">or</span> <span class="n">keep_samples</span> <span class="o">==</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)),</span> <span class="n">msg</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;Cannot find ``{keep_samples}``&#39;</span><span class="p">)</span>
<span class="n">fail_if</span><span class="p">(</span><span class="ow">not</span> <span class="p">(</span><span class="n">keep_variants</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="ow">or</span> <span class="n">keep_variants</span> <span class="o">==</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)),</span> <span class="n">msg</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;Cannot find ``{keep_variants}``&#39;</span><span class="p">)</span>
<span class="n">fail_if</span><span class="p">(</span><span class="ow">not</span> <span class="p">(</span><span class="n">remove_samples</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="ow">or</span> <span class="n">remove_samples</span> <span class="o">==</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)),</span> <span class="n">msg</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;Cannot find ``{remove_samples}``&#39;</span><span class="p">)</span>

<span class="kn">input:</span> <span class="n">genoFile</span><span class="p">,</span> <span class="n">group_by</span><span class="o">=</span><span class="mi">1</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{cwd}/cache/{_input:bn}.{name}.filtered{&quot;.extracted&quot; if keep_variants.is_file() else &quot;&quot;}.bed&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{step_name}_{_output:bn}&#39;</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container_lmm</span><span class="p">,</span> <span class="n">volumes</span><span class="o">=</span><span class="p">[</span><span class="n">f</span><span class="s1">&#39;{cwd}:{cwd}&#39;</span><span class="p">],</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.stdout&#39;</span>
    <span class="n">plink2</span> \
      <span class="o">--</span><span class="n">bfile</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">n</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--maf </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">maf_filter</span><span class="p">)</span> <span class="k">if</span> <span class="n">maf_filter</span> <span class="o">&gt;=</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--max-maf </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">maf_max_filter</span><span class="p">)</span> <span class="k">if</span> <span class="n">maf_max_filter</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--geno </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">geno_filter</span><span class="p">)</span> <span class="k">if</span> <span class="n">geno_filter</span> <span class="o">&gt;=</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--hwe </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">hwe_filter</span><span class="p">)</span> <span class="k">if</span> <span class="n">hwe_filter</span> <span class="o">&gt;=</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--mind </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">mind_filter</span><span class="p">)</span> <span class="k">if</span> <span class="n">mind_filter</span> <span class="o">&gt;=</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--keep </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">keep_samples</span><span class="p">)</span> <span class="k">if</span> <span class="n">keep_samples</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--remove </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">remove_samples</span><span class="p">)</span> <span class="k">if</span> <span class="n">remove_samples</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--extract </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">keep_variants</span><span class="p">)</span> <span class="k">if</span> <span class="n">keep_variants</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">make</span><span class="o">-</span><span class="n">bed</span> \
      <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">threads</span> <span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">memory</span> <span class="err">$</span><span class="p">{</span><span class="nb">int</span><span class="p">(</span><span class="n">expand_size</span><span class="p">(</span><span class="kp">mem</span><span class="p">)</span> <span class="o">*</span> <span class="mf">0.9</span><span class="p">)}</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[1]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># LD prunning and remove related individuals (both ind of a pair)</span>
<span class="p">[</span><span class="n">qc_2</span> <span class="p">(</span><span class="n">LD</span> <span class="n">pruning</span><span class="p">)]</span>
<span class="c1"># Window size</span>
<span class="kn">parameter:</span> <span class="n">window</span> <span class="o">=</span> <span class="mi">50</span>
<span class="c1"># Shift window every 10 snps</span>
<span class="kn">parameter:</span> <span class="n">shift</span> <span class="o">=</span> <span class="mi">10</span>
<span class="kn">parameter:</span> <span class="n">r2</span> <span class="o">=</span> <span class="mf">0.1</span>
<span class="kn">output:</span> <span class="n">bed</span><span class="o">=</span><span class="n">f</span><span class="s1">&#39;{_input:n}.prune.bed&#39;</span><span class="p">,</span> <span class="n">prune</span><span class="o">=</span><span class="n">f</span><span class="s1">&#39;{_input:n}.prune.in&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{step_name}_{_output[0]:bn}&#39;</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container_lmm</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[0]:n}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[0]:n}.stdout&#39;</span>
    <span class="n">plink2</span> \
    <span class="o">--</span><span class="n">bfile</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">n</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">indep</span><span class="o">-</span><span class="n">pairwise</span> <span class="err">$</span><span class="p">{</span><span class="n">window</span><span class="p">}</span> <span class="err">$</span><span class="p">{</span><span class="n">shift</span><span class="p">}</span> <span class="err">$</span><span class="p">{</span><span class="n">r2</span><span class="p">}</span>  \
    <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="s2">&quot;prune&quot;</span><span class="p">]:</span><span class="n">nn</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">threads</span> <span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">memory</span> <span class="err">$</span><span class="p">{</span><span class="nb">int</span><span class="p">(</span><span class="n">expand_size</span><span class="p">(</span><span class="kp">mem</span><span class="p">)</span> <span class="o">*</span> <span class="mf">0.9</span><span class="p">)}</span>
   
    <span class="n">plink2</span> \
    <span class="o">--</span><span class="n">bfile</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">n</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">extract</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="s1">&#39;prune&#39;</span><span class="p">]}</span> \
    <span class="o">--</span><span class="n">make</span><span class="o">-</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="s1">&#39;bed&#39;</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">threads</span> <span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">memory</span> <span class="err">$</span><span class="p">{</span><span class="nb">int</span><span class="p">(</span><span class="n">expand_size</span><span class="p">(</span><span class="kp">mem</span><span class="p">)</span> <span class="o">*</span> <span class="mf">0.9</span><span class="p">)}</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># Merge all the .bed files into one bed file </span>
<span class="p">[</span><span class="n">qc_3</span> <span class="p">(</span><span class="n">merge</span> <span class="nb">all</span> <span class="n">files</span><span class="p">)]</span>
<span class="kn">parameter:</span> <span class="n">merged_prefix</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="n">merged_prefix</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_input[0]:bn}.merged&#39;</span> <span class="k">if</span> <span class="n">merged_prefix</span> <span class="o">==</span> <span class="s1">&#39;&#39;</span> <span class="k">else</span> <span class="n">merged_prefix</span>
<span class="n">sos_run</span><span class="p">(</span><span class="s2">&quot;merge_plink&quot;</span><span class="p">,</span> <span class="n">merged_prefix</span><span class="o">=</span><span class="n">merged_prefix</span><span class="p">,</span> <span class="n">genoFile</span><span class="o">=</span><span class="n">_input</span><span class="p">[</span><span class="s1">&#39;bed&#39;</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="n">merge_plink</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">merged_prefix</span> <span class="o">=</span> <span class="nb">str</span>
<span class="kn">parameter:</span> <span class="n">genoFile</span> <span class="o">=</span> <span class="n">paths</span>
<span class="n">skip_if</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">genoFile</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span>
<span class="kn">input:</span> <span class="n">genoFile</span><span class="p">,</span> <span class="n">group_by</span> <span class="o">=</span> <span class="s1">&#39;all&#39;</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s2">&quot;{_input[0]:d}/{merged_prefix}.bed&quot;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{step_name}_{_output:bn}&#39;</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container_lmm</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.stdout&#39;</span>
    <span class="n">echo</span> <span class="o">-</span><span class="n">e</span> <span class="err">$</span><span class="p">{</span><span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)[:</span><span class="o">-</span><span class="mi">4</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">:]])}</span> <span class="o">|</span> <span class="n">sed</span> <span class="s1">&#39;s/ /</span><span class="se">\n</span><span class="s1">/g&#39;</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">merge_list</span>
    <span class="n">plink2</span> \
    <span class="o">--</span><span class="n">bfile</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">merge</span><span class="o">-</span><span class="nb">list</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">merge_list</span> \
    <span class="o">--</span><span class="n">make</span><span class="o">-</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">threads</span> <span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">memory</span> <span class="err">$</span><span class="p">{</span><span class="nb">int</span><span class="p">(</span><span class="n">expand_size</span><span class="p">(</span><span class="kp">mem</span><span class="p">)</span> <span class="o">*</span> <span class="mf">0.9</span><span class="p">)}</span>
</pre></div>

    </div>
</div>
</div>

</div>
<hr>
&copy 2020-current Gao Wang and others, Department of Neurology at Columbia University
<p><small>Exported from <a href="http://github.com/cumc/bioworkflows/blob/e440166b2b950d521f451dea4d680a8290f1d802/GWAS/GWAS_QC.ipynb"><code>GWAS/GWAS_QC.ipynb</code></a> committed by Gao Wang on Wed May 19 03:33:36 2021 <a href="http://github.com/cumc/bioworkflows/commit/e440166b2b950d521f451dea4d680a8290f1d802">revision 28, e440166</a> <a href="https://stephenslab.github.io/ipynb-website/notes.html#Note-about-commit-ids"><span class="fa fa-question-circle"></span></a></small></p>
</div>
</div>
</body>
</html>
