<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="ipynb_website:version" content="0.9.7" />
<meta name="viewport" content="width=device-width, initial-scale=1" />

<link rel="stylesheet" type="text/css" href="../css/jt.css">

<link rel="stylesheet" type="text/css" href="../css/toc2.css">

<link href="../site_libs/jqueryui-1.11.4/jquery-ui.css">
<link rel="stylesheet" href="../site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<link rel="stylesheet" href="../site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jqueryui/1.9.1/jquery-ui.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>

<link rel="stylesheet"
      href="../site_libs/highlightjs/null.min.css"
      type="text/css" />

<script src="../site_libs/highlightjs/highlight.pack.js"></script>
<script>hljs.initHighlightingOnLoad();</script>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>

<script src="../js/doc_toc.js"></script>
<script src="../js/docs.js"></script>

<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.2/MathJax.js?config=TeX-MML-AM_CHTML"></script>
<script>
    MathJax.Hub.Config({
        extensions: ["tex2jax.js"],
        jax: ["input/TeX", "output/HTML-CSS"],
        tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true
        },
        "HTML-CSS": {
            preferredFont: "TeX",
            availableFonts: ["TeX"],
            styles: {
                scale: 110,
                ".MathJax_Display": {
                    "font-size": "110%",
                }
            }
        }
    });
</script>
<script>
function filterDataFrame(id) {
    var input = document.getElementById("search_" + id);
    var filter = input.value.toUpperCase();
    var table = document.getElementById("dataframe_" + id);
    var tr = table.getElementsByTagName("tr");
    // Loop through all table rows, and hide those who don't match the search query
    for (var i = 1; i < tr.length; i++) {
        for (var j = 0; j < tr[i].cells.length; ++j) {
            var matched = false;
            if (tr[i].cells[j].innerHTML.toUpperCase().indexOf(filter) != -1) {
                tr[i].style.display = "";
                matched = true
                break;
            }
            if (!matched)
                tr[i].style.display = "none";
        }
    }
}
function sortDataFrame(id, n, dtype) {
    var table = document.getElementById("dataframe_" + id);
    var tb = table.tBodies[0]; // use `<tbody>` to ignore `<thead>` and `<tfoot>` rows
    var tr = Array.prototype.slice.call(tb.rows, 0); // put rows into array
    if (dtype === 'numeric') {
        var fn = function(a, b) { 
            return parseFloat(a.cells[n].textContent) <= parseFloat(b.cells[n].textContent) ? -1 : 1;
        }
    } else {
        var fn = function(a, b) {
            var c = a.cells[n].textContent.trim().localeCompare(b.cells[n].textContent.trim()); 
            return c > 0 ? 1 : (c < 0 ? -1 : 0) }
    }
    var isSorted = function(array, fn) {
        if (array.length < 2)
            return 1;
        var direction = fn(array[0], array[1]); 
        for (var i = 1; i < array.length - 1; ++i) {
            var d = fn(array[i], array[i+1]);
            if (d == 0)
                continue;
            else if (direction == 0)
                direction = d;
            else if (direction != d)
                return 0;
            }
        return direction;
    }
    var sorted = isSorted(tr, fn);
    if (sorted == 1 || sorted == -1) {
        // if sorted already, reverse it
        for(var i = tr.length - 1; i >= 0; --i)
            tb.appendChild(tr[i]); // append each row in order
    } else {
        tr = tr.sort(fn);
        for(var i = 0; i < tr.length; ++i)
            tb.appendChild(tr[i]); // append each row in order
    }
}
</script>

<script>
$( document ).ready(function(){
            var cfg={'threshold':3,     // depth of toc (number of levels)
             'number_sections': false,
             'toc_cell': false,          // useless here
             'toc_window_display': true, // display the toc window
             "toc_section_display": "block", // display toc contents in the window
             'sideBar':true,       // sidebar or floating window
             'navigate_menu':false       // navigation menu (only in liveNotebook -- do not change)
            }
            var st={};                  // some variables used in the script
            st.rendering_toc_cell = false;
            st.config_loaded = false;
            st.extension_initialized=false;
            st.nbcontainer_marginleft = $('#notebook-container').css('margin-left')
            st.nbcontainer_marginright = $('#notebook-container').css('margin-right')
            st.nbcontainer_width = $('#notebook-container').css('width')
            st.oldTocHeight = undefined
            st.cell_toc = undefined;
            st.toc_index=0;
            // fire the main function with these parameters
            table_of_contents(cfg, st);
            var file=GWASDict[$("h1:first").attr("id")];
            $("#toc-level0 a").css("color","#126dce");
            $('a[href="#'+$("h1:first").attr("id")+'"]').hide()
            var docs=GWASArray;
            var docs_map=GWASArrayMap;
            var pos=GWASArray.indexOf(file);
            for (var a=pos;a>=0;a--){
                  $('<li><a href="'+docs[a]+'.html"><font color="#073642"><b>'+docs_map[docs[a]].replace(/_/g," ")+'</b></font></a></li>').insertBefore("#toc-level0 li:eq(0)");
            }
            $('a[href="'+file+'.html'+'"]').css("color","#126dce");
            for (var a=pos+1;a<docs.length;a++){
                  $(".toc #toc-level0").append('<li><a href="'+docs[a]+'.html"><font color="#073642"><b>'+docs_map[docs[a]].replace(/_/g," ")+'</b></font></a></li>');
            }
            // $("#toc-header").hide(); // comment out because it prevents search bar from displaying
    });
</script>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');
  // mark it active
  menuAnchor.parent().addClass('active');
  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>
<div class="container-fluid main-container">
<!-- tabsets -->
<script src="../site_libs/navigation-1.1/tabsets.js"></script>
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>



<title>Some computational genomics workflows</title>

<style type = "text/css">
body {
  font-family: "Droid Sans";
  padding-top: 66px;
  padding-bottom: 40px;
}
</style>
</head>

<body>
<div tabindex="-1" id="notebook" class="border-box-sizing">
<div class="container" id="notebook-container">

<!-- code folding -->

<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">Some computational genomics workflows</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
<li>
  <a href="../index.html">Overview</a>
</li>
        
      </ul>
        
<ul class="nav navbar-nav navbar-right">
<li>
   <a href="http://github.com/cumc/bioworkflows"> <span class="fa fa-github"></span> </a>
</li>
</ul>
        
      </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="LMM/GLMM-analyses-for-UK-Biobank-data">LMM/GLMM analyses for UK Biobank data<a class="anchor-link" href="#LMM/GLMM-analyses-for-UK-Biobank-data">&#182;</a></h1><p>This notebook implements pipelines for analyzing binary and quantitative traits association using BOLT-LMM (version 2.3.4), fastGWA, REGENIE and SAIGE software.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Aim">Aim<a class="anchor-link" href="#Aim">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>This pipeline was initially developed to perfom genetic association analysis using various LMM methods on UK Biobank imputed data of ~500K invidivuals, although it can be used to analyze other studies.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Input-data">Input data<a class="anchor-link" href="#Input-data">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<ol>
<li>Genotype file for constructing the GRM (genetic relationship matrix) formated as a plink binary file <code>(.bed/.bim/.fam)</code> <ul>
<li><code>--bfile=prefix</code></li>
</ul>
</li>
<li>Imputed genotype dosages in <code>bgen</code> format (<code>.bgen</code>, <code>.bgi</code>, <code>.sample</code>)<ul>
<li><code>--bgenFile</code> and <code>--sampleFile</code></li>
</ul>
</li>
<li>Phenotype file (white space delimited file with column headers, first two columns should be FID and IID) specify files by options <code>--phenoFile</code> and the phenotype to be analized by <code>--phenoCol</code></li>
<li>Covariates file (same format as phenoFile) specify them by <code>--covarFile</code> for qualitative covariates use <code>--covarCol</code> and for quantitative <code>--qCovarCol</code>. If <code>--covarFile</code> is not specified then phenotype file will be used as covariate file. To specify an array of covariates you can use bash tricks, eg <code>--qCovarCol PC{1:20}</code></li>
</ol>
<p>Note: reference genome used <strong>GRCh37/hg19</strong>.</p>
<h2 id="Software-specific-inputs">Software specific inputs<a class="anchor-link" href="#Software-specific-inputs">&#182;</a></h2><h3 id="BoltLMM-additional-input">BoltLMM additional input<a class="anchor-link" href="#BoltLMM-additional-input">&#182;</a></h3><ul>
<li>Reference genetic maps, provided on BoltLMM website<ul>
<li><code>--geneticMapFile=tables/genetic_map_hg##.txt.gz</code></li>
</ul>
</li>
<li>Reference LD scores, provided on BoltLMM website</li>
<li>Use <code>--covarMaxLevels</code> to specify the number of categories of a qualitative covariate. </li>
</ul>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Output">Output<a class="anchor-link" href="#Output">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Our pipeline generates</p>
<ol>
<li>Summary statistics file for each variant analyzed</li>
<li>QQ and Manhattan plots for these summary statistics</li>
</ol>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="References">References<a class="anchor-link" href="#References">&#182;</a></h2><p>To learn more about each of the specific methods applied in this pipeline please refer to the published papers and software documentation:</p>
<ol>
<li><a href="http://dx.doi.org/10.1038/ng.3190">Bolt-LMM</a> and <a href="https://alkesgroup.broadinstitute.org/BOLT-LMM">documentation</a></li>
<li><a href="http://dx.doi.org/10.1038/s41588-019-0530-8">FastGWA</a> and <a href="https://cnsgenomics.com/software/gcta/#Overview">documentation</a></li>
<li><a href="https://www.biorxiv.org/content/10.1101/2020.06.19.162354v2">REGENIE</a> and <a href="https://rgcgithub.github.io/regenie/">documentation</a></li>
<li><a href="http://dx.doi.org/10.1038/s415">SAIGE</a> and <a href="https://github.com/weizhouUMICH/SAIGE">documentation</a></li>
<li><a href="https://github.com/hanchenphd/GMMAT">GMMAT</a> and <a href="https://github.com/hanchenphd/GMMAT/blob/master/inst/doc/GMMAT.pdf">documentation</a></li>
</ol>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Command-interface">Command interface<a class="anchor-link" href="#Command-interface">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[1]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">LMM</span><span class="o">.</span><span class="n">ipynb</span> <span class="o">-</span><span class="n">h</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

    <div class="prompt"></div>


<div class="output_subarea output_stream output_stdout output_text">
<pre>usage: sos run LMM.ipynb [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &#34;sos run -h&#34; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  boltlmm
  gcta
  fastGWA
  PLINK_QC
  regenie
  regenie_burden
  SAIGE

Global Workflow Options:
  --cwd VAL (as path, required)
                        the output directory for generated files
  --sampleFile . (as path)
                        Path to sample file
  --bfile VAL (as path, required)
                        Genotype files in plink binary this is used for
                        computing the GRM
  --genoFile  paths

                        Path to bgen or bed files
  --phenoFile VAL (as path, required)
                        Phenotype file for quantitative trait (BMI)
  --phenoCol VAL VAL ... (as type, required)
                        Phenotype to be analyzed (specify the column)
  --covarFile . (as path)
                        Covariate file path. Will use phenoFile if empty
  --formatFile . (as path)
                        Summary statisticss format file path used for unifying
                        output column names. Will not unify names if empty
  --covarCol  (as list)
                        Qualitative covariates to be used in the analysis
  --qCovarCol  (as list)
                        Quantitative covariates to be used in the analysis
  --numThreads 2 (as int)
                        Specific number of threads to use
  --bgenMinMAF 0.001 (as float)
                        Minimum MAF to be used
  --bgenMinINFO 0.8 (as float)
                        Mimimum info score to be used
  --job-size 1 (as int)
                        For cluster jobs, number commands to run per job
  --container-lmm &#39;statisticalgenetics/lmm:1.8&#39;
                        The container with the lmm software. Can be either a
                        dockerhub image or a singularity `sif` file. Default is
                        set to using dockerhub image
  --container-marp &#39;gaow/marp&#39;

Sections
  boltlmm_1:            Run BOLT analysis
    Workflow Options:
      --covarMaxLevels VAL (as int, required)
                        Maximum categories of covariates allowed
      --LDscoresFile VAL (as path, required)
                        Path to LDscore file for reference population
      --geneticMapFile VAL (as path, required)
                        Path to genetic map file used to interpolate genetic map
                        coordinates from SNP physical (base pair) positions
      --lmm-option lmm
                        LMM option: lmm, lmmInfOnly, and lmmForceNonInf
  gcta_1:               Partition the GRM into 100 parts and allocate 8GB memory
                        to each job
    Workflow Options:
      --parts 100 (as int)
                        Number of parts the GRM calculation is to be partitioned
  gcta_2:               Merge all the parts together (Linux, Mac)
  gcta_3:               Make a sparse GRM from the merged full-dense GRM
  fastGWA_1:            fastGWA mixed model (based on the sparse GRM generated
                        above)
    Workflow Options:
      --grmFile  path(f&#39;{cwd}/{bfile:bn}.grm.sp&#39;)

  PLINK_QC:             Select the SNPs and samples to be used based on maf,
                        geno, hwe and mind options
    Workflow Options:
      --maf-filter 0.0 (as float)
      --geno-filter 0.0 (as float)
      --hwe-filter 0.0 (as float)
      --mind-filter 0.0 (as float)
  regenie_1:            Run REGENIE step 1: fitting the null
    Workflow Options:
      --bsize 400 (as int)
                        Size of the genotype blocks to be used
      --lowmem-prefix  cwd

                        Path to temporarily store block predictions
      --trait bt
                        Specify that traits are binary with
                        0=control,1=case,NA=missing (default is quantitative)
  regenie_2:            Run REGENIE step 2: association analysis
    Workflow Options:
      --bsize 400 (as int)
                        Size of the genotype blocks to be used
      --minMAC VAL (as int, required)
                        Mimimum allele count to be used
      --trait bt
  regenie_burden_1:     Run REGENIE_BURDEN step 1: fitting the null Similar to
                        the REGENIE step 1, but add &#39;regenie_burden&#39; instead of
                        &#39;regenie&#39; as a sufix
    Workflow Options:
      --bsize 400 (as int)
                        Size of the genotype blocks to be used
      --lowmem-prefix  cwd

                        Path to temporarily store block predictions
      --trait bt
                        Specify that traits are binary with
                        0=control,1=case,NA=missing (default is quantitative)
  regenie_burden_2:     Run regenie for burden tests
    Workflow Options:
      --trait bt
                        Specify that traits are binary with
                        0=control,1=case,NA=missing (default is quantitative)
      --bsize 400 (as int)
                        Size of the genotype blocks to be used
      --anno-file VAL (as path, required)
                        Annotation file format: variantID, gene and functional
                        annotation (space/tab delimited)
      --set-list VAL (as path, required)
                        This file lists variants within each set/gene to use
                        when building masks. Format: set/gene name, chromosome,
                        physical pos set/gene, then by a comma-separated list of
                        variants included in the set/gene.
      --keep-gene VAL (as path, required)
                        Select specific genes/sets to test
      --mask-file VAL (as path, required)
                        Allele frequency file. format: variantId, alternative
                        allele frequency parameter: aaf_file = path Select the
                        annotations to be used in the mask file. format: mask#
                        annotatio type
      --aaf-bins 0.05 (as float)
                        Select the upper MAF to generate masks
      --build-mask max
                        The way in which the alternative alleles are counted
  SAIGE_1:              Fit SAIGE null model
    Workflow Options:
      --trait-type VAL (as str, required)
                        trait type, eg &#39;binary&#39; or &#39;quantitative&#39;
      --loco TRUE
                        Whether to use LOCO or not
      --sampleCol IID
                        Name of the sample column
      --script-path /Users/zhoujiayi/software/bin/step1_fitNULLGLMM.R (as path)
                        Path specific to SAIGE script
      --invNormalize FALSE
                        Inverse normalization only for non-normal quantitative
                        traits
  SAIGE_2:              Compute SAIGE statistics
    Workflow Options:
      --bgenMinMAC 4 (as int)
                        Mimimum allele count to be used
      --af-caco TRUE
                        Specify whether to output allele frequencies in cases
                        and controls
      --script-path /Users/zhoujiayi/software/bin/step2_SPAtests.R (as path)
                        Path specific to SAIGE script
  regenie_burden_3:     Merge results and log files Skipping the first two rows
                        of the input file
    Workflow Options:
      --[no-]reverse-log-p (default to False)
  boltlmm_2, fastGWA_2, SAIGE_3, regenie_3: Merge results and log files
    Workflow Options:
      --[no-]reverse-log-p (default to False)
  boltlmm_3, fastGWA_3, SAIGE_4, regenie_4: Manhattan and QQ plots using `qqman`
    Workflow Options:
      --bp POS
                        Column name for BP
      --pval P
                        Column name for p-value
      --snp SNP
                        Column name for SNP
      --p-filter &#39;0.05&#39;
                        Plot only on p-values smaller than this
      --ylim 0 (as int)
                        ylim set to 0 to use maximum -log10(p) in data
  boltlmm_4, fastGWA_4, SAIGE_5, regenie_5: Generate analysis report: HTML file,
                        and optionally PPTX file
</pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Global-parameter-setting">Global parameter setting<a class="anchor-link" href="#Global-parameter-setting">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[2]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># the output directory for generated files</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># Path to sample file</span>
<span class="kn">parameter:</span> <span class="n">sampleFile</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)</span>
<span class="c1"># Genotype files in plink binary this is used for computing the GRM</span>
<span class="kn">parameter:</span> <span class="n">bfile</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># Path to bgen or bed files </span>
<span class="kn">parameter:</span> <span class="n">genoFile</span> <span class="o">=</span> <span class="n">paths</span>
<span class="c1"># Phenotype file for quantitative trait (BMI)</span>
<span class="kn">parameter:</span> <span class="n">phenoFile</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># Phenotype to be analyzed (specify the column)</span>
<span class="kn">parameter:</span> <span class="n">phenoCol</span> <span class="o">=</span> <span class="nb">list</span>
<span class="c1"># Covariate file path. Will use phenoFile if empty</span>
<span class="kn">parameter:</span> <span class="n">covarFile</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)</span>
<span class="c1"># Summary statisticss format file path used for unifying output column names. Will not unify names if empty</span>
<span class="kn">parameter:</span> <span class="n">formatFile</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)</span>
<span class="c1"># Qualitative covariates to be used in the analysis</span>
<span class="kn">parameter:</span> <span class="n">covarCol</span> <span class="o">=</span> <span class="p">[]</span>
<span class="c1"># Quantitative covariates to be used in the analysis</span>
<span class="kn">parameter:</span> <span class="n">qCovarCol</span> <span class="o">=</span> <span class="p">[]</span>
<span class="c1"># Specific number of threads to use</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">2</span>
<span class="c1"># Minimum MAF to be used</span>
<span class="kn">parameter:</span> <span class="n">bgenMinMAF</span> <span class="o">=</span> <span class="mf">0.001</span>
<span class="c1"># Mimimum info score to be used</span>
<span class="kn">parameter:</span> <span class="n">bgenMinINFO</span> <span class="o">=</span> <span class="mf">0.8</span>
<span class="c1"># For cluster jobs, number commands to run per job</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">1</span>
<span class="c1"># The container with the lmm software. Can be either a dockerhub image or a singularity `sif` file.</span>
<span class="c1"># Default is set to using dockerhub image</span>
<span class="kn">parameter:</span> <span class="n">container_lmm</span> <span class="o">=</span> <span class="s1">&#39;statisticalgenetics/lmm:1.9&#39;</span>
<span class="kn">parameter:</span> <span class="n">container_marp</span> <span class="o">=</span> <span class="s1">&#39;gaow/marp&#39;</span>
<span class="k">if</span> <span class="ow">not</span> <span class="n">covarFile</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>
    <span class="n">covarFile</span> <span class="o">=</span> <span class="n">phenoFile</span>
<span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;{cwd:a}&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Illustration-with-minimal-working-examples">Illustration with minimal working examples<a class="anchor-link" href="#Illustration-with-minimal-working-examples">&#182;</a></h2>
<pre><code>JOB_OPT='-j 2'</code></pre>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="BOLT-LMM-example-command">BOLT-LMM example command<a class="anchor-link" href="#BOLT-LMM-example-command">&#182;</a></h3><p>On a minimal working example (MWE) dataset (about 1min to complete the analysis),</p>

<pre><code>sos run LMM.ipynb boltlmm \
    --cwd output \
    --bfile data/genotypes.bed \
    --sampleFile data/imputed_genotypes.sample \
    --genoFile data/imputed_genotypes_chr*.bgen \
    --phenoFile data/phenotypes.txt \
    --formatFile data/boltlmm_template.yml \
    --LDscoresFile data/LDSCORE.1000G_EUR.tab.gz \
    --geneticMapFile data/genetic_map_hg19_withX.txt.gz \
    --phenoCol BMI \
    --covarCol SEX \
    --covarMaxLevels 10 \
    --qCovarCol AGE \
    --numThreads 5 \
    --bgenMinMAF 0.001 \
    --bgenMinINFO 0.1 \
    --lmm-option none \
    --p-filter 1 \
    $JOB_OPT</code></pre>
<p>Please note that the command above is only meant to demonstrate the usage of the pipeline. Data will be generated to a folder called <code>output</code>. We set <code>--lmm-option</code> to <code>none</code> to not run LMM on this minimal data-set. The <code>--pval</code> column name <code>P_LINREG</code> for QQ/Manhattan plot is also p-value from conventional linear regression. In practice we will definitely want to use one of the LMM options in BoltLMM. Default is <code>lmm</code> switch in <code>bolt</code> if you don't specify <code>--lmm-option</code>.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="fastGWA-example-command">fastGWA example command<a class="anchor-link" href="#fastGWA-example-command">&#182;</a></h3><p>On a minimal working example (MWE) dataset (analysis completes almost instantly),</p>

<pre><code>sos run LMM.ipynb fastGWA \
    --cwd output \
    --bfile data/genotypes.bed \
    --sampleFile data/imputed_genotypes.sample \
    --genoFile data/imputed_genotypes_chr*.bgen \
    --phenoFile data/phenotypes.txt \
    --formatFile data/fastGWA_template.yml \
    --phenoCol BMI \
    --covarCol SEX \
    --qCovarCol AGE \
    --numThreads 1 \
    --bgenMinMAF 0.001 \
    --bgenMinINFO 0.1 \
    --parts 2 \
    --p-filter 1 \
    $JOB_OPT</code></pre>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h4 id="REGENIE-example-command">REGENIE example command<a class="anchor-link" href="#REGENIE-example-command">&#182;</a></h4><p>On a minimal working example (MWE) dataset,</p>

<pre><code>sos run LMM.ipynb regenie \
    --cwd output \
    --bfile data/genotypes21_22.bed \
    --maf-filter 0.001 \
    --sampleFile data/imputed_genotypes.sample \
    --genoFile data/imputed_genotypes_chr*.bgen \
    --phenoFile data/phenotypes.txt \
    --formatFile data/regenie_template.yml \
    --phenoCol ASTHMA T2D\
    --covarCol SEX \
    --qCovarCol AGE \
    --numThreads 8 \
    --bsize 1000 \
    --trait bt \
    --minMAC 4 \
    --bgenMinMAF 0.05 \
    --bgenMinINFO 0.8 \
    --reverse_log_p \
    --p-filter 1 \
    $JOB_OPT</code></pre>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="REGENIE-burden-example-command">REGENIE burden example command<a class="anchor-link" href="#REGENIE-burden-example-command">&#182;</a></h3><p>On a minimal working example (MWE) dataset,</p>

<pre><code>sos run LMM.ipynb regenie_burden \
    --cwd output \
    --bfile genotypes_21_22_plink.exome.bed \
    --genoFile ukb23155_c22_b0_v1.plink.exome.filtered.bed \
    --phenoFile phenotype_burden.txt\
    --phenoCol ASTHMA T2D \
    --formatFile data/regenie_template.yml \
    --covarCol SEX \
    --qCovarCol AGE \
    --numThreads 8 \
    --bsize 1000 \
    --anno_file annotation_file.txt\
    --set_list set_list_file_chr22.txt \
    --mask_file mask_file.txt \
    --keep_gene keep_gene.txt\
    --aaf_bins 0.05 \
    --trait bt \
    --build_mask max</code></pre>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="SAIGE-example-command">SAIGE example command<a class="anchor-link" href="#SAIGE-example-command">&#182;</a></h3><p>On a minimal working example (MWE) dataset,</p>

<pre><code>sos run LMM.ipynb SAIGE \
    --cwd output \
    --bfile data/genotypes.bed \
    --sampleFile data/imputed_genotypes.sample \
    --genoFile data/imputed_genotypes_chr*.bgen \
    --phenoFile data/phenotypes.txt \
    --phenoCol BMI \
    --covarCol SEX \
    --qCovarCol AGE \
    --numThreads 4 \
    --bgenMinMAF 0.001 \
    --bgenMinINFO 0.1 \
    --trait_type quantitative \
    --pval p.val \
    --bp POS \
    --p-filter 1 \
    $JOB_OPT</code></pre>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Input-using-bed-format">Input using <code>bed</code> format<a class="anchor-link" href="#Input-using-bed-format">&#182;</a></h3><p>sos run LMM.ipynb fastGWA \
    --cwd output \
    --bfile data/genotypes.bed \
    --genoFile data/genotypes21_22.bed \
    --phenoFile data/phenotypes.txt \
    --formatFile data/fastGWA_template.yml \
    --phenoCol BMI \
    --covarCol SEX \
    --qCovarCol AGE \
    --numThreads 1 \
    --bgenMinMAF 0.001 \
    --bgenMinINFO 0.1 \
    --parts 2 \
    --p-filter 1 \
    $JOB_OPT</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="GMMAT-example-command">GMMAT example command<a class="anchor-link" href="#GMMAT-example-command">&#182;</a></h3><p>sos run LMM.ipynb GMMAT \
    --cwd gemma_output \
    --bfile 100K_chr22.bed \
    --genoFile 100K_chr22.bed \
    --phenoFile MWE_pheno_new1.txt \
    --formatFile gmmat_template.yml \
    --phenoCol AD \
    --covarCol SEX \
    --covarMaxLevels 10 \
    --qCovarCol AGE \
    --numThreads 5 \
    --bgenMinMAF 0.001 \
    --bgenMinINFO 0.1 \
    --lmm-option none \
    --p-filter 1 \
    --geno_filter 0.0005 \
    --nperbatch 10 \
    $JOB_OPT</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Run-workflow-on-a-cluster">Run workflow on a cluster<a class="anchor-link" href="#Run-workflow-on-a-cluster">&#182;</a></h3><p>The shell variable <code>JOB_OPT</code> was set to <code>-j 2</code>. That is, run 2 jobs in parallel on a local computer (each using 5 threads due to <code>--numThreads 5</code>).</p>
<p>On cluster we use a job template, and configure <code>JOB_OPT</code> as follows:</p>

<pre><code>JOB_OPT="-c farnam.yml -q farnam -J 40"</code></pre>
<p>Here we use task queue <code>farnam</code> configured in file <code>farnam.yml</code>. We allow for at most 40 jobs in the cluster job queue.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="BoltLMM-workflow-implementation">BoltLMM workflow implementation<a class="anchor-link" href="#BoltLMM-workflow-implementation">&#182;</a></h2><p>To install from source code follow instructions here: <a href="https://data.broadinstitute.org/alkesgroup/BOLT-LMM/#x1-70002.2">https://data.broadinstitute.org/alkesgroup/BOLT-LMM/#x1-70002.2</a></p>

<pre><code>- On Linux machine a binary executable is provided and can be used.
- Supporting files such as LD score file and genetic map file can be found [in the installation bundle](https://data.broadinstitute.org/alkesgroup/BOLT-LMM/downloads/BOLT-LMM_v2.3.4.tar.gz).
- For a complete description on bolt commands go to: http://manpages.ubuntu.com/manpages/eoan/en/man1/bolt.1.html.

</code></pre>
<p><strong>A note for developers</strong>: it is important to have input and output for each step. Input files and output files are best derived from one another.</p>
<p>BOLT-LMM software computes statistics for testing association between phenotypes and genotypes using a linear mixed model</p>

<pre><code>--bfile = accepts genotype files in PLINK binary format (.fam, .bed, .bim)
--geneticMapFile = Oxford-format file for interpolating genetic distances: tables/genetic_map_hg##.txt.gz
--phenoFile = phenotype file (header required; FII and IID must be first two columns)
--phenoCol = phenotype columns header
--covarFile = covariate file (header required; FII and IID must be first two columns)
--covarCol = categorical covariate column(s); for &gt;1, use multiple --covarCol and/or {i:j} expansion
--qcovarCol = quantitative covariate column(s); for  &gt;1, use multiple --qCovarCol and/or {i:j} expansion
--lmm = compute assoc stats under the inf model and with Bayesian non-inf prior (VB approx), if power gain expected
--modelSnps = file(s) listing SNPs to use in model (i.e., GRM) (default: use all non-excluded SNPs)
--LDscoresFile = LD Scores for calibration of Bayesian assoc stats: tables/LDSCORE.1000G_EUR.tab.g
--numThreads = number of computational threads
--statsFile = output file for assoc stats at PLINK genotypes
--bgenFile = file(s) containing Oxford BGEN-format genotypes to test for association
--sampleFile = file containing Oxford sample file corresponding to BGEN file(s)
--bgenMinMAF = MAF threshold on Oxford BGEN-format genotypes; lower-MAF SNPs will be ignored
--bgenMinINFO = INFO threshold on Oxford BGEN-format genotypes; lower-INFO SNPs will be ignored
--statsFileBgenSNPs = output file for assoc stats at BGEN-format genotypes</code></pre>
<p>It is important to know that BOLT-LMMv2.3.4 accepts bgen files only in 8bit formatting as stated below:</p>
<p><em>WARNING: The BGEN format comprises a few sub-formats; we have only implemented support for the versions (and specific data layouts) used in the UK Biobank N=150K and N=500K releases. In particular, for BGEN v1.2, BOLT-LMM currently only supports the 8-bit encoding used for the UK Biobank N=500K data. (Starting with BOLT-LMM v2.3.3, missing values in BGEN v1.2 data are now allowed.)</em></p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># Run BOLT analysis</span>
<span class="p">[</span><span class="n">boltlmm_1</span><span class="p">]</span>
<span class="c1"># Maximum categories of covariates allowed </span>
<span class="kn">parameter:</span> <span class="n">covarMaxLevels</span> <span class="o">=</span> <span class="nb">int</span>
<span class="c1"># Path to LDscore file for reference population</span>
<span class="kn">parameter:</span> <span class="n">LDscoresFile</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># Path to genetic map file used to interpolate genetic map coordinates from SNP physical (base pair) positions</span>
<span class="kn">parameter:</span> <span class="n">geneticMapFile</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># LMM option: lmm, lmmInfOnly, and lmmForceNonInf</span>
<span class="kn">parameter:</span> <span class="n">lmm_option</span> <span class="o">=</span> <span class="s1">&#39;lmm&#39;</span>
<span class="kn">depends:</span> <span class="n">LDscoresFile</span><span class="p">,</span> <span class="n">geneticMapFile</span>
<span class="kn">input:</span> <span class="n">genoFile</span><span class="p">,</span> <span class="n">group_by</span> <span class="o">=</span> <span class="mi">1</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{cwd}/cache/{_input:bn}.{phenoFile:bn}_{phenoCol[0]}.boltlmm.snp_stats.gz&#39;</span>
<span class="n">file_options</span><span class="o">=</span><span class="n">f</span><span class="s2">&quot;--bfile {bfile:n} --bgenFile={_input} --bgenMinMAF={bgenMinMAF} --bgenMinINFO={bgenMinINFO} --sampleFile={sampleFile} --statsFileBgenSnps={_output} --statsFile={_output:nn}.ref_stats.gz &quot;</span> <span class="k">if</span> <span class="n">_input</span><span class="o">.</span><span class="n">suffix</span> <span class="o">==</span> <span class="s2">&quot;.bgen&quot;</span> <span class="k">else</span> <span class="n">f</span><span class="s2">&quot;--bfile={_input:n} --statsFile={_output} &quot;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;48h&#39;</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;60G&#39;</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{step_name}_{_output:bn}&#39;</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container_lmm</span><span class="p">,</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.stdout&#39;</span><span class="p">,</span> <span class="n">volumes</span> <span class="o">=</span> <span class="p">[</span><span class="n">f</span><span class="s2">&quot;{cwd:a}:{cwd:a}&quot;</span><span class="p">]</span>
    <span class="n">bolt</span> \
    <span class="o">--</span><span class="n">phenoFile</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">phenoFile</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">phenoCol</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">phenoCol</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span> \
    <span class="o">--</span><span class="n">covarFile</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">covarFile</span><span class="p">}</span> \
    <span class="err">$</span><span class="p">{</span><span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s1">&#39;--covarCol=</span><span class="si">%s</span><span class="s1"> &#39;</span> <span class="o">%</span> <span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">covarCol</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">])}</span> \
    <span class="o">--</span><span class="n">covarMaxLevels</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">covarMaxLevels</span><span class="p">}</span> \
    <span class="err">$</span><span class="p">{</span><span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s1">&#39;--qCovarCol=</span><span class="si">%s</span><span class="s1"> &#39;</span> <span class="o">%</span> <span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">qCovarCol</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">])}</span> \
    <span class="o">--</span><span class="n">LDscoresFile</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">LDscoresFile</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">geneticMapFile</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">geneticMapFile</span><span class="p">}</span> \
    <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--&#39;</span> <span class="o">+</span> <span class="n">lmm_option</span><span class="p">)</span> <span class="k">if</span> <span class="n">lmm_option</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;lmm&#39;</span><span class="p">,</span> <span class="s1">&#39;lmmInfOnly&#39;</span><span class="p">,</span> <span class="s1">&#39;lmmForceNonInf&#39;</span><span class="p">]</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">}</span> \
    <span class="err">$</span><span class="p">{</span><span class="n">file_options</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">numThreads</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">verboseStats</span> 

<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="kp">active</span> <span class="o">=</span> <span class="p">(</span><span class="n">_index</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span>
    <span class="c1"># remove redundant reference summary stats file</span>
    <span class="n">rm</span> <span class="o">-</span><span class="n">f</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">nn</span><span class="p">}</span><span class="o">.</span><span class="n">ref_stats</span><span class="o">.</span><span class="n">gz</span>

<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="kp">active</span> <span class="o">=</span> <span class="p">(</span><span class="n">_index</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span>
    <span class="c1"># rename reference summary stats file</span>
    <span class="k">if</span> <span class="p">[</span> <span class="o">-</span><span class="n">f</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">nn</span><span class="p">}</span><span class="o">.</span><span class="n">ref_stats</span><span class="o">.</span><span class="n">gz</span> <span class="p">];</span> <span class="n">then</span>
      <span class="n">mv</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">nn</span><span class="p">}</span><span class="o">.</span><span class="n">ref_stats</span><span class="o">.</span><span class="n">gz</span> <span class="err">$</span><span class="p">{</span><span class="n">cwd</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">phenoFile</span><span class="p">:</span><span class="n">bn</span><span class="p">}</span><span class="n">_</span><span class="err">$</span><span class="p">{</span><span class="n">phenoCol</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span><span class="o">.</span><span class="n">boltlmm</span><span class="o">.</span><span class="n">ref_stats</span><span class="o">.</span><span class="n">gz</span>
    <span class="k">else</span>
       <span class="n">echo</span> <span class="s2">&quot;File does not exist.&quot;</span>
    <span class="n">fi</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="fastGWA-workflow-implementation">fastGWA workflow implementation<a class="anchor-link" href="#fastGWA-workflow-implementation">&#182;</a></h2><p>Installation instructions can be found in <a href="https://cnsgenomics.com/software/gcta/#Download">https://cnsgenomics.com/software/gcta/#Download</a>. On Linux machine a binary executable is provided and can be used.</p>
<p>Documentation: <a href="https://cnsgenomics.com/software/gcta/#fastGWA">https://cnsgenomics.com/software/gcta/#fastGWA</a></p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Step-1:-Creation-of-the-GRM">Step 1: Creation of the GRM<a class="anchor-link" href="#Step-1:-Creation-of-the-GRM">&#182;</a></h3><p>The GRM only needs to be created once for all the phenotypes to analyze with the same genotypic data. In this step the GRM calculation is divided into multiple parts for a faster computational time.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># Partition the GRM into 100 parts and allocate 8GB memory to each job</span>
<span class="p">[</span><span class="n">gcta_1</span><span class="p">]</span>
<span class="c1"># Number of parts the GRM calculation is to be partitioned</span>
<span class="kn">parameter:</span> <span class="n">parts</span> <span class="o">=</span> <span class="mi">100</span>
<span class="n">part_number</span> <span class="o">=</span> <span class="p">[</span><span class="n">f</span><span class="s1">&#39;{parts}_{format(x+1, &quot;0&quot; + str(len(str(parts))))}&#39;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">parts</span><span class="p">)]</span>
<span class="kn">input:</span> <span class="n">bfile</span><span class="p">,</span> <span class="n">for_each</span> <span class="o">=</span> <span class="s1">&#39;part_number&#39;</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{cwd}/cache/{_input:bn}.part_{_part_number}.grm.bin&#39;</span><span class="p">,</span> 
        <span class="n">f</span><span class="s1">&#39;{cwd}/cache/{_input:bn}.part_{_part_number}.grm.N.bin&#39;</span><span class="p">,</span> 
        <span class="n">f</span><span class="s1">&#39;{cwd}/cache/{_input:bn}.part_{_part_number}.grm.id&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;48h&#39;</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;48G&#39;</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{step_name}_{_output[0]:bn}&#39;</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container_lmm</span><span class="p">,</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[0]:n}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[0]:n}.stdout&#39;</span>
    <span class="n">gcta64</span> \
    <span class="o">--</span><span class="n">bfile</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">make</span><span class="o">-</span><span class="n">grm</span><span class="o">-</span><span class="n">part</span> <span class="err">$</span><span class="p">{</span><span class="n">parts</span><span class="p">}</span> <span class="err">$</span><span class="p">{</span><span class="n">_index</span><span class="o">+</span><span class="mi">1</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">thread</span><span class="o">-</span><span class="n">num</span> <span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">nnn</span><span class="p">}</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Step-2:-Combine-all-the-GRM-parts-into-one-file">Step 2: Combine all the GRM parts into one file<a class="anchor-link" href="#Step-2:-Combine-all-the-GRM-parts-into-one-file">&#182;</a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># Merge all the parts together (Linux, Mac)</span>
<span class="p">[</span><span class="n">gcta_2</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">group_by</span> <span class="o">=</span> <span class="s1">&#39;all&#39;</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{cwd}/{bfile:bn}.grm.bin&#39;</span><span class="p">,</span> 
        <span class="n">f</span><span class="s1">&#39;{cwd}/{bfile:bn}.grm.N.bin&#39;</span><span class="p">,</span> 
        <span class="n">f</span><span class="s1">&#39;{cwd}/{bfile:bn}.grm.id&#39;</span> 
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;2h&#39;</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;6G&#39;</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{step_name}_{_output[0]:bn}&#39;</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container_lmm</span><span class="p">,</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[0]:n}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[0]:n}.stdout&#39;</span>
    <span class="c1"># here input is results all parts each having 3 items. We need to get the corresponding every other 3 items</span>
    <span class="n">cat</span> <span class="err">$</span><span class="p">{</span><span class="n">paths</span><span class="p">(</span><span class="n">_input</span><span class="p">[::</span><span class="mi">3</span><span class="p">])}</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span>
    <span class="n">cat</span> <span class="err">$</span><span class="p">{</span><span class="n">paths</span><span class="p">(</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">::</span><span class="mi">3</span><span class="p">])}</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">1</span><span class="p">]}</span>
    <span class="n">cat</span> <span class="err">$</span><span class="p">{</span><span class="n">paths</span><span class="p">(</span><span class="n">_input</span><span class="p">[</span><span class="mi">2</span><span class="p">::</span><span class="mi">3</span><span class="p">])}</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">2</span><span class="p">]}</span>
    <span class="c1">#rm ${paths(_input)}</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Step-3:-Make-a-sparse-GRM-to-be-used-in-the-association-analyses">Step 3: Make a sparse GRM to be used in the association analyses<a class="anchor-link" href="#Step-3:-Make-a-sparse-GRM-to-be-used-in-the-association-analyses">&#182;</a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># Make a sparse GRM from the merged full-dense GRM</span>
<span class="p">[</span><span class="n">gcta_3</span><span class="p">]</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{cwd}/{bfile:bn}.grm.sp&#39;</span> 
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;48h&#39;</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;48G&#39;</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{step_name}_{_output:bn}&#39;</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container_lmm</span><span class="p">,</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output}.stdout&#39;</span>
    <span class="n">gcta64</span> <span class="o">--</span><span class="n">grm</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">nn</span><span class="p">}</span> <span class="o">--</span><span class="n">make</span><span class="o">-</span><span class="n">bK</span><span class="o">-</span><span class="n">sparse</span> <span class="mf">0.05</span> <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">nn</span><span class="p">}</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Step-4:-Run-the-single-variant-association-analysis-using-FastGWA">Step 4: Run the single variant association analysis using FastGWA<a class="anchor-link" href="#Step-4:-Run-the-single-variant-association-analysis-using-FastGWA">&#182;</a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># fastGWA mixed model (based on the sparse GRM generated above)</span>
<span class="p">[</span><span class="n">fastGWA_1</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">grmFile</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;{cwd}/{bfile:bn}.grm.sp&#39;</span><span class="p">)</span>
<span class="kn">depends:</span> <span class="n">grmFile</span>
<span class="c1"># extract and prepare phenotype &amp; covariate files</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="kn">as</span> <span class="nn">pd</span>
<span class="n">dat</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">phenoFile</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">delim_whitespace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">phenoCol</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
    <span class="n">dat</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;{cwd}/{phenoFile:bn}.fastGWA_phenotype&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;FID&#39;</span><span class="p">,</span> <span class="s1">&#39;IID&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">phenoCol</span><span class="p">)</span>
<span class="n">dat</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">covarFile</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">delim_whitespace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">covarCol</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
    <span class="n">dat</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;{cwd}/{phenoFile:bn}.fastGWA_covar&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;FID&#39;</span><span class="p">,</span> <span class="s1">&#39;IID&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">covarCol</span><span class="p">)</span>
<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">qCovarCol</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
    <span class="n">dat</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;{cwd}/{phenoFile:bn}.fastGWA_qcovar&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;FID&#39;</span><span class="p">,</span> <span class="s1">&#39;IID&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">qCovarCol</span><span class="p">)</span>

<span class="kn">input:</span> <span class="n">genoFile</span><span class="p">,</span> <span class="n">group_by</span> <span class="o">=</span> <span class="mi">1</span>
<span class="n">input_options</span> <span class="o">=</span> <span class="n">f</span><span class="s2">&quot;--bgen {_input} --info {bgenMinINFO} --sample {sampleFile}&quot;</span> <span class="k">if</span> <span class="n">_input</span><span class="o">.</span><span class="n">suffix</span> <span class="o">==</span> <span class="s2">&quot;.bgen&quot;</span> <span class="k">else</span> <span class="n">f</span><span class="s2">&quot;--bfile {_input:n}&quot;</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{cwd}/cache/{_input:bnn}.{phenoFile:bn}.fastGWA.gz&#39;</span>
<span class="n">fail_if</span><span class="p">(</span><span class="ow">not</span> <span class="n">path</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;{_input}.bgi&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="ow">and</span> <span class="n">_input</span><span class="o">.</span><span class="n">suffix</span> <span class="o">==</span> <span class="s1">&#39;.bgen&#39;</span><span class="p">,</span> <span class="n">msg</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;Cannot find file ``{_input}.bgi``. Please generate it using command ``bgenix -g {_input} -index``.&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="n">_input</span><span class="o">.</span><span class="n">suffix</span> <span class="o">==</span> <span class="s2">&quot;.bgen&quot;</span> <span class="k">else</span> <span class="n">f</span><span class="s2">&quot;continue&quot;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;48h&#39;</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;5G&#39;</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{step_name}_{_output:bn}&#39;</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container_lmm</span><span class="p">,</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.stdout&#39;</span>
    <span class="n">gcta64</span> \
    <span class="err">$</span><span class="p">{</span><span class="n">input_options</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">grm</span><span class="o">-</span><span class="n">sparse</span> <span class="err">$</span><span class="p">{</span><span class="n">grmFile</span><span class="p">:</span><span class="n">nn</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">maf</span> <span class="err">$</span><span class="p">{</span><span class="n">bgenMinMAF</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">fastGWA</span><span class="o">-</span><span class="n">mlm</span> \
    <span class="o">--</span><span class="n">pheno</span> <span class="err">$</span><span class="p">{</span><span class="n">cwd</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">phenoFile</span><span class="p">:</span><span class="n">bn</span><span class="p">}</span><span class="o">.</span><span class="n">fastGWA_phenotype</span> \
    <span class="o">--</span><span class="n">qcovar</span> <span class="err">$</span><span class="p">{</span><span class="n">cwd</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">covarFile</span><span class="p">:</span><span class="n">bn</span><span class="p">}</span><span class="o">.</span><span class="n">fastGWA_qcovar</span> \
    <span class="o">--</span><span class="n">covar</span> <span class="err">$</span><span class="p">{</span><span class="n">cwd</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">covarFile</span><span class="p">:</span><span class="n">bn</span><span class="p">}</span><span class="o">.</span><span class="n">fastGWA_covar</span> \
    <span class="o">--</span><span class="n">threads</span> <span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">nn</span><span class="p">}</span>\
    <span class="o">&amp;&amp;</span> <span class="n">gzip</span> <span class="o">-</span><span class="n">f</span> <span class="o">--</span><span class="n">best</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Output from each step:</p>
<ol>
<li><strong>gcta_1 for x number of parts (in the example above x=100, so this step will create 400 files):</strong></li>
</ol>
<ul>
<li>test.part_{_part_number}.grm.bin</li>
<li>test.part_{_part_number}.grm.N.bin </li>
<li>test.part_{_part_number}.grm.id</li>
<li>test.part_{_part_number}.log (the program creates the log file so there is no need for .stderr and .stdout)</li>
</ul>
<ol>
<li><strong>gcta_2 this step creates 5 output files:</strong></li>
</ol>
<ul>
<li>test.grm.bin (it is a binary file which contains the lower triangle elements of the GRM)</li>
<li>test.grm.N.bin (it is a binary file which contains the number of SNPs used to calculate the GRM)</li>
<li>test.grm.id (no header line; columns are family ID and individual ID, see above)</li>
<li>test.grm.stderr</li>
<li>test.grm.stdout</li>
</ul>
<ol>
<li><strong>gcta_3 this step creates 3 output files:</strong></li>
</ol>
<ul>
<li>test.grm.sp (sparse GRM made from the dense GRM)</li>
<li>test.grm.sp.stderr</li>
<li>test.grm.sp.stdout</li>
</ul>
<ol>
<li><strong>fastGWA this step creates 2 output files per chromosome</strong></li>
</ol>
<ul>
<li>test{chr1:22}.fastGWA</li>
<li>test{chr1:22}.fastGWA.log</li>
</ul>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="REGENIE-workflow-implementation">REGENIE workflow implementation<a class="anchor-link" href="#REGENIE-workflow-implementation">&#182;</a></h1>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Documentation can be found <a href="https://rgcgithub.github.io/regenie/">here</a>. Binary and quantitative traits should be analyzed separately.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># Select the SNPs and samples to be used based on maf, geno, hwe and mind options</span>
<span class="p">[</span><span class="n">PLINK_QC</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">maf_filter</span> <span class="o">=</span> <span class="mf">0.0</span>
<span class="kn">parameter:</span> <span class="n">geno_filter</span> <span class="o">=</span> <span class="mf">0.0</span>
<span class="kn">parameter:</span> <span class="n">hwe_filter</span> <span class="o">=</span> <span class="mf">0.0</span>
<span class="kn">parameter:</span> <span class="n">mind_filter</span> <span class="o">=</span> <span class="mf">0.0</span>
<span class="kn">input:</span> <span class="n">bfile</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{cwd}/cache/{bfile:bn}.qc_pass.id&#39;</span><span class="p">,</span> <span class="n">f</span><span class="s1">&#39;{cwd}/cache/{bfile:bn}.qc_pass.snplist&#39;</span> 
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;10h&#39;</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;30G&#39;</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{step_name}_{_output[0]:bn}&#39;</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container_lmm</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[0]:n}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[0]:n}.stdout&#39;</span> 
    <span class="n">plink2</span> \
      <span class="o">--</span><span class="n">bfile</span> <span class="err">$</span><span class="p">{</span><span class="n">bfile</span><span class="p">:</span><span class="n">n</span><span class="p">}</span> <span class="o">--</span><span class="n">mac</span> <span class="mi">1</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--maf </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">maf_filter</span><span class="p">)</span> <span class="k">if</span> <span class="n">maf_filter</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">}</span> <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--geno </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">geno_filter</span><span class="p">)</span> <span class="k">if</span> <span class="n">geno_filter</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">}</span> <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--hwe </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">hwe_filter</span><span class="p">)</span> <span class="k">if</span> <span class="n">hwe_filter</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">}</span> <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--mind </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">mind_filter</span><span class="p">)</span> <span class="k">if</span> <span class="n">mind_filter</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">write</span><span class="o">-</span><span class="n">snplist</span> <span class="o">--</span><span class="n">write</span><span class="o">-</span><span class="n">samples</span> <span class="o">--</span><span class="n">no</span><span class="o">-</span><span class="nb">id</span><span class="o">-</span><span class="n">header</span> \
      <span class="o">--</span><span class="n">threads</span> <span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Step-1:-fitting-the-null">Step 1: fitting the null<a class="anchor-link" href="#Step-1:-fitting-the-null">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># Run REGENIE step 1: fitting the null</span>
<span class="p">[</span><span class="n">regenie_1</span><span class="p">,</span><span class="n">regenie_burden_1</span><span class="p">]</span>
<span class="c1"># Size of the genotype blocks to be used </span>
<span class="kn">parameter:</span> <span class="n">bsize</span> <span class="o">=</span> <span class="mi">400</span>
<span class="c1"># Path to temporarily store block predictions</span>
<span class="kn">parameter:</span> <span class="n">lowmem_dir</span> <span class="o">=</span> <span class="n">cwd</span>
<span class="c1"># Specify that traits are binary with 0=control,1=case,NA=missing (default is quantitative)</span>
<span class="kn">parameter:</span> <span class="n">trait</span> <span class="o">=</span> <span class="s1">&#39;bt&#39;</span>
<span class="c1"># extract and prepare phenotype &amp; covariate files</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="kn">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="kn">as</span> <span class="nn">np</span>
<span class="n">dat</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">phenoFile</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">delim_whitespace</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">str</span><span class="p">)</span>
<span class="n">dat</span> <span class="o">=</span> <span class="n">dat</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">to_replace</span> <span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">value</span> <span class="o">=</span><span class="s2">&quot;NA&quot;</span><span class="p">)</span>
<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">phenoCol</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>    
    <span class="n">dat</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;{cwd}/{phenoFile:bn}.regenie_phenotype&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;FID&#39;</span><span class="p">,</span> <span class="s1">&#39;IID&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">phenoCol</span><span class="p">)</span>
<span class="n">dat</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">covarFile</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">delim_whitespace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">covarCol</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="nb">len</span><span class="p">(</span><span class="n">qCovarCol</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
    <span class="n">dat</span> <span class="o">=</span> <span class="n">dat</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="n">covarCol</span><span class="p">)</span>
    <span class="n">dat</span> <span class="o">=</span> <span class="n">dat</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="n">qCovarCol</span><span class="p">)</span>
    <span class="n">dat</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">to_replace</span> <span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">value</span> <span class="o">=</span><span class="s2">&quot;NA&quot;</span><span class="p">)</span>
    <span class="n">dat1</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">dat</span><span class="p">,</span> <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;FID&#39;</span><span class="p">,</span><span class="s1">&#39;IID&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">covarCol</span><span class="p">)</span>
    <span class="n">dat1</span> <span class="o">=</span> <span class="n">dat1</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
    <span class="n">dat2</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">dat</span><span class="p">,</span> <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;IID&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">qCovarCol</span><span class="p">)</span>
    <span class="n">merged_left</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">left</span><span class="o">=</span><span class="n">dat1</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="n">dat2</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;left&#39;</span><span class="p">,</span> <span class="n">left_on</span><span class="o">=</span><span class="s1">&#39;IID&#39;</span><span class="p">,</span> <span class="n">right_on</span><span class="o">=</span><span class="s1">&#39;IID&#39;</span><span class="p">)</span>
    <span class="n">merged_left</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;{cwd}/{phenoFile:bn}.regenie_covar&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
<span class="kn">depends:</span> <span class="n">f</span><span class="s1">&#39;{cwd}/cache/{bfile:bn}.qc_pass.snplist&#39;</span><span class="p">,</span> <span class="n">f</span><span class="s1">&#39;{cwd}/cache/{bfile:bn}.qc_pass.id&#39;</span>
<span class="kn">input:</span> <span class="n">geno</span> <span class="o">=</span> <span class="n">bfile</span><span class="p">,</span> <span class="n">pheno</span> <span class="o">=</span> <span class="n">f</span><span class="s2">&quot;{cwd}/{phenoFile:bn}.regenie_phenotype&quot;</span><span class="p">,</span> <span class="n">covar</span> <span class="o">=</span> <span class="n">f</span><span class="s2">&quot;{cwd}/{phenoFile:bn}.regenie_covar&quot;</span><span class="p">,</span> <span class="n">qc</span> <span class="o">=</span> <span class="n">output_from</span><span class="p">(</span><span class="s2">&quot;PLINK_QC&quot;</span><span class="p">)</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{cwd}/{phenoFile:bn}_&#39;</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">phenoCol</span><span class="p">])</span> <span class="o">+</span> <span class="s1">&#39;.regenie_pred.list&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;12h&#39;</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;15G&#39;</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{step_name}_{_output[0]:bn}&#39;</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container_lmm</span><span class="p">,</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.stdout&#39;</span><span class="p">,</span> <span class="n">volumes</span> <span class="o">=</span> <span class="p">[</span><span class="n">f</span><span class="s2">&quot;{lowmem_dir:a}:{lowmem_dir:a}&quot;</span><span class="p">]</span>
    <span class="n">regenie</span> \
      <span class="o">--</span><span class="n">step</span> <span class="mi">1</span> \
      <span class="o">--</span><span class="n">bed</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="s2">&quot;geno&quot;</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">phenoFile</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="s2">&quot;pheno&quot;</span><span class="p">]}</span> \
      <span class="o">--</span><span class="n">covarFile</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="s2">&quot;covar&quot;</span><span class="p">]}</span> \
      <span class="o">--</span><span class="n">keep</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="s2">&quot;qc&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]}</span> \
      <span class="o">--</span><span class="n">extract</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="s2">&quot;qc&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">]}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--&#39;</span> <span class="o">+</span> <span class="n">trait</span><span class="p">)</span> <span class="k">if</span> <span class="n">trait</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;bt&#39;</span><span class="p">]</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">bsize</span> <span class="err">$</span><span class="p">{</span><span class="n">bsize</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">lowmem</span> <span class="o">--</span><span class="n">lowmem</span><span class="o">-</span><span class="n">prefix</span> <span class="err">$</span><span class="p">{</span><span class="n">lowmem_dir</span><span class="p">:</span><span class="n">a</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">bn</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">threads</span> <span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">nn</span><span class="p">}</span><span class="o">.</span><span class="n">regenie</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Step-2:-association-analysis">Step 2: association analysis<a class="anchor-link" href="#Step-2:-association-analysis">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># Run REGENIE step 2: association analysis</span>
<span class="p">[</span><span class="n">regenie_2</span><span class="p">]</span>
<span class="c1"># Size of the genotype blocks to be used </span>
<span class="kn">parameter:</span> <span class="n">bsize</span> <span class="o">=</span> <span class="mi">400</span>
<span class="c1"># Mimimum allele count to be used</span>
<span class="kn">parameter:</span> <span class="n">minMAC</span> <span class="o">=</span> <span class="nb">int</span>
<span class="kn">parameter:</span> <span class="n">trait</span> <span class="o">=</span> <span class="s1">&#39;bt&#39;</span>
<span class="kn">input:</span> <span class="n">genoFile</span><span class="p">,</span> <span class="n">group_by</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">group_with</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">info</span><span class="o">=</span><span class="p">[(</span><span class="n">path</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;{cwd}/{phenoFile:bn}_&#39;</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">phenoCol</span><span class="p">])</span> <span class="o">+</span> <span class="s1">&#39;.regenie_pred.list&#39;</span><span class="p">))]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">genoFile</span><span class="p">))</span>
<span class="n">input_options</span> <span class="o">=</span> <span class="n">f</span><span class="s2">&quot;--bgen {_input} --sample {sampleFile}&quot;</span> <span class="k">if</span> <span class="n">_input</span><span class="o">.</span><span class="n">suffix</span> <span class="o">==</span> <span class="s2">&quot;.bgen&quot;</span> <span class="k">else</span> <span class="n">f</span><span class="s2">&quot;--bed {_input:n}&quot;</span>
<span class="kn">output:</span> <span class="p">[</span><span class="n">f</span><span class="s1">&#39;{cwd}/cache/{_input:bn}_&#39;</span><span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">phenoCol</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="o">+</span> <span class="s1">&#39;.regenie.gz&#39;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">phenoCol</span><span class="p">))]</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;12h&#39;</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;15G&#39;</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{step_name}_{_output[0]:bn}&#39;</span>
<span class="kn">bash:</span><span class="n">container</span><span class="o">=</span><span class="n">container_lmm</span><span class="p">,</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{cwd}/cache/{_input:bn}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{cwd}/cache/{_input:bn}.stdout&#39;</span><span class="p">,</span> <span class="n">volumes</span> <span class="o">=</span> <span class="p">[</span><span class="n">f</span><span class="s2">&quot;{cwd:a}:{cwd:a}&quot;</span><span class="p">]</span>
    <span class="nb">set</span> <span class="o">-</span><span class="n">e</span>
    <span class="n">regenie</span> \
     <span class="o">--</span><span class="n">step</span> <span class="mi">2</span> \
     <span class="err">$</span><span class="p">{</span><span class="n">input_options</span><span class="p">}</span> \
     <span class="o">--</span><span class="n">phenoFile</span> <span class="err">$</span><span class="p">{</span><span class="n">cwd</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">phenoFile</span><span class="p">:</span><span class="n">bn</span><span class="p">}</span><span class="o">.</span><span class="n">regenie_phenotype</span> \
     <span class="o">--</span><span class="n">covarFile</span> <span class="err">$</span><span class="p">{</span><span class="n">cwd</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">covarFile</span><span class="p">:</span><span class="n">bn</span><span class="p">}</span><span class="o">.</span><span class="n">regenie_covar</span> \
     <span class="o">--</span><span class="n">phenoColList</span> <span class="err">$</span><span class="p">{</span><span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">phenoCol</span><span class="p">)}</span> \
     <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--&#39;</span> <span class="o">+</span> <span class="n">trait</span><span class="p">)</span> <span class="k">if</span> <span class="n">trait</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;bt&#39;</span><span class="p">]</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">}</span> \
     <span class="o">--</span><span class="n">firth</span> <span class="mf">0.01</span> <span class="o">--</span><span class="n">approx</span> \
     <span class="o">--</span><span class="n">pred</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="o">.</span><span class="n">info</span><span class="p">}</span> \
     <span class="o">--</span><span class="n">bsize</span> <span class="err">$</span><span class="p">{</span><span class="n">bsize</span><span class="p">}</span> \
     <span class="o">--</span><span class="n">minMAC</span> <span class="err">$</span><span class="p">{</span><span class="n">minMAC</span><span class="p">}</span> \
     <span class="o">--</span><span class="n">minINFO</span> <span class="err">$</span><span class="p">{</span><span class="n">bgenMinINFO</span><span class="p">}</span>\
     <span class="o">--</span><span class="n">split</span> \
     <span class="o">--</span><span class="n">threads</span> <span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span> \
     <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">{</span><span class="n">cwd</span><span class="p">}</span><span class="o">/</span><span class="n">cache</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">bn</span><span class="p">}</span> <span class="o">&amp;&amp;</span> \
     <span class="n">gzip</span> <span class="o">-</span><span class="n">f</span> <span class="o">--</span><span class="n">best</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Regenie-burden-test">Regenie burden test<a class="anchor-link" href="#Regenie-burden-test">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[1]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># Run regenie for burden tests</span>
<span class="p">[</span><span class="n">regenie_burden_2</span><span class="p">]</span>
<span class="c1"># Specify that traits are binary with 0=control,1=case,NA=missing (default is quantitative)</span>
<span class="kn">parameter:</span> <span class="n">trait</span> <span class="o">=</span> <span class="s1">&#39;bt&#39;</span>
<span class="c1"># Size of the genotype blocks to be used </span>
<span class="kn">parameter:</span> <span class="n">bsize</span> <span class="o">=</span> <span class="mi">400</span>
<span class="c1"># Annotation file format: variantID, gene and functional annotation (space/tab delimited)</span>
<span class="kn">parameter:</span> <span class="n">anno_file</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># This file lists variants within each set/gene to use when building masks. Format: set/gene name, chromosome, physical pos set/gene, then by a comma-separated list of variants included in the set/gene.</span>
<span class="kn">parameter:</span> <span class="n">set_list</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># Select specific genes/sets to test</span>
<span class="kn">parameter:</span> <span class="n">keep_gene</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># Allele frequency file. format: variantId, alternative allele frequency</span>
<span class="c1">#parameter: aaf_file = path</span>
<span class="c1"># Select the annotations to be used in the mask file. format: mask# annotatio type</span>
<span class="kn">parameter:</span> <span class="n">mask_file</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># Select the upper MAF to generate masks</span>
<span class="kn">parameter:</span> <span class="n">aaf_bins</span> <span class="o">=</span> <span class="mf">0.05</span>
<span class="c1"># The way in which the alternative alleles are counted</span>
<span class="kn">parameter:</span> <span class="n">build_mask</span> <span class="o">=</span> <span class="s1">&#39;max&#39;</span>
<span class="kn">input:</span> <span class="n">genoFile</span><span class="p">,</span> <span class="n">group_by</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">group_with</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">info</span><span class="o">=</span><span class="p">[(</span><span class="n">path</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;{cwd}/{phenoFile:bn}_&#39;</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">phenoCol</span><span class="p">])</span> <span class="o">+</span> <span class="s1">&#39;.regenie_pred.list&#39;</span><span class="p">))]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">genoFile</span><span class="p">))</span>
<span class="n">input_options</span> <span class="o">=</span> <span class="n">f</span><span class="s2">&quot;--bgen {_input} --sample {sampleFile}&quot;</span> <span class="k">if</span> <span class="n">_input</span><span class="o">.</span><span class="n">suffix</span> <span class="o">==</span> <span class="s2">&quot;.bgen&quot;</span> <span class="k">else</span> <span class="n">f</span><span class="s2">&quot;--bed {_input:n}&quot;</span>
<span class="kn">output:</span> <span class="p">[</span><span class="n">f</span><span class="s1">&#39;{cwd}/cache/{_input:bn}_burden_&#39;</span><span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">phenoCol</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="o">+</span> <span class="s1">&#39;.regenie.gz&#39;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">phenoCol</span><span class="p">))]</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;48h&#39;</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;15G&#39;</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{step_name}_{_output[0]:bn}&#39;</span>
<span class="kn">bash:</span><span class="n">container</span><span class="o">=</span><span class="n">container_lmm</span><span class="p">,</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{cwd}/cache/{_input:bn}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{cwd}/cache/{_input:bn}.stdout&#39;</span><span class="p">,</span> <span class="n">volumes</span> <span class="o">=</span> <span class="p">[</span><span class="n">f</span><span class="s2">&quot;{cwd:a}:{cwd:a}&quot;</span><span class="p">]</span>
    <span class="nb">set</span> <span class="o">-</span><span class="n">e</span>
    <span class="n">regenie</span> \
      <span class="o">--</span><span class="n">step</span> <span class="mi">2</span> \
      <span class="err">$</span><span class="p">{</span><span class="n">input_options</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">phenoFile</span> <span class="err">$</span><span class="p">{</span><span class="n">cwd</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">phenoFile</span><span class="p">:</span><span class="n">bn</span><span class="p">}</span><span class="o">.</span><span class="n">regenie_phenotype</span> \
      <span class="o">--</span><span class="n">covarFile</span> <span class="err">$</span><span class="p">{</span><span class="n">cwd</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">covarFile</span><span class="p">:</span><span class="n">bn</span><span class="p">}</span><span class="o">.</span><span class="n">regenie_covar</span> \
      <span class="o">--</span><span class="n">phenoColList</span> <span class="err">$</span><span class="p">{</span><span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">phenoCol</span><span class="p">)}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--&#39;</span> <span class="o">+</span> <span class="n">trait</span><span class="p">)</span> <span class="k">if</span> <span class="n">trait</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;bt&#39;</span><span class="p">]</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">firth</span> <span class="o">--</span><span class="n">approx</span> \
      <span class="o">--</span><span class="n">pred</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="o">.</span><span class="n">info</span><span class="p">}</span> \
      <span class="o">--</span><span class="nb">set</span><span class="o">-</span><span class="nb">list</span> <span class="err">$</span><span class="p">{</span><span class="n">set_list</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">extract</span><span class="o">-</span><span class="n">sets</span> <span class="err">$</span><span class="p">{</span><span class="n">keep_gene</span><span class="p">}</span>\
      <span class="o">--</span><span class="n">aaf</span><span class="o">-</span><span class="n">bins</span> <span class="err">$</span><span class="p">{</span><span class="n">aaf_bins</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">write</span><span class="o">-</span><span class="n">mask</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--build-mask &#39;</span> <span class="o">+</span> <span class="n">build_mask</span><span class="p">)</span> <span class="k">if</span> <span class="n">build_mask</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;max&#39;</span><span class="p">,</span><span class="s1">&#39;sum&#39;</span><span class="p">,</span><span class="s1">&#39;comphet&#39;</span><span class="p">]</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">bsize</span> <span class="err">$</span><span class="p">{</span><span class="n">bsize</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">check</span><span class="o">-</span><span class="n">burden</span><span class="o">-</span><span class="n">files</span> \
      <span class="o">--</span><span class="n">out</span>  <span class="err">$</span><span class="p">{</span><span class="n">cwd</span><span class="p">}</span><span class="o">/</span><span class="n">cache</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">bn</span><span class="p">}</span><span class="n">_burden</span> <span class="o">&amp;&amp;</span> \
      <span class="n">gzip</span> <span class="o">-</span><span class="n">f</span> <span class="o">--</span><span class="n">best</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="SAIGE-workflow-implementation">SAIGE workflow implementation<a class="anchor-link" href="#SAIGE-workflow-implementation">&#182;</a></h2><p>We need to create a conda enviroment for the installation of SAIGE in Yale's HRC cluster. Instructions in <a href="https://github.com/weizhouUMICH/SAIGE">https://github.com/weizhouUMICH/SAIGE</a></p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Step-1:-fitting-the-null">Step 1: fitting the null<a class="anchor-link" href="#Step-1:-fitting-the-null">&#182;</a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># Fit SAIGE null model</span>
<span class="p">[</span><span class="n">SAIGE_1</span><span class="p">]</span>
<span class="c1"># trait type, eg &#39;binary&#39; or &#39;quantitative&#39;</span>
<span class="kn">parameter:</span> <span class="n">trait_type</span> <span class="o">=</span> <span class="nb">str</span>
<span class="c1"># Whether to use LOCO or not</span>
<span class="kn">parameter:</span> <span class="n">loco</span> <span class="o">=</span> <span class="s1">&#39;TRUE&#39;</span>
<span class="c1"># Name of the sample column</span>
<span class="kn">parameter:</span> <span class="n">sampleCol</span><span class="o">=</span><span class="s1">&#39;IID&#39;</span>
<span class="c1">#Path specific to SAIGE script</span>
<span class="kn">parameter:</span> <span class="n">script_path</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;~/software/bin/step1_fitNULLGLMM.R&#39;</span><span class="p">)</span>
<span class="c1"># Inverse normalization only for non-normal quantitative traits</span>
<span class="kn">parameter:</span> <span class="n">invNormalize</span> <span class="o">=</span> <span class="s1">&#39;FALSE&#39;</span>
<span class="kn">input:</span> <span class="n">bfile</span><span class="p">,</span> <span class="n">phenoFile</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{cwd}/{bfile:bn}.{phenoFile:bn}.SAIGE.rda&#39;</span><span class="p">,</span> <span class="n">f</span><span class="s1">&#39;{cwd}/{bfile:bn}.{phenoFile:bn}.SAIGE.varianceRatio.txt&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;48h&#39;</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;60G&#39;</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{step_name}_{_output[0]:bn}&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[0]:n}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[0]:n}.stdout&#39;</span><span class="p">,</span> <span class="n">template_name</span><span class="o">=</span><span class="s1">&#39;conda&#39;</span><span class="p">,</span> <span class="n">env_name</span><span class="o">=</span><span class="s1">&#39;RSAIGE&#39;</span>
    <span class="n">Rscript</span> <span class="err">$</span><span class="p">{</span><span class="n">script_path</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">plinkFile</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">phenoFile</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]}</span> \
        <span class="o">--</span><span class="n">phenoCol</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">phenoCol</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span> \
        <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--covarColList=&#39;</span> <span class="o">+</span> <span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">covarCol</span> <span class="o">+</span> <span class="n">qCovarCol</span><span class="p">))</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">covarCol</span> <span class="o">+</span> <span class="n">qCovarCol</span><span class="p">)</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">sampleIDColinphenoFile</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">sampleCol</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">traitType</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">trait_type</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">outputPrefix</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">nThreads</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">LOCO</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">loco</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">invNormalize</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">invNormalize</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">IsOverwriteVarianceRatioFile</span><span class="o">=</span><span class="n">TRUE</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Step-2:-perform-single-variant-association-test">Step 2: perform single variant association test<a class="anchor-link" href="#Step-2:-perform-single-variant-association-test">&#182;</a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># Compute SAIGE statistics</span>
<span class="p">[</span><span class="n">SAIGE_2</span><span class="p">]</span>
<span class="c1"># Mimimum allele count to be used</span>
<span class="kn">parameter:</span> <span class="n">bgenMinMAC</span> <span class="o">=</span> <span class="mi">4</span>
<span class="c1">#Specify whether to output allele frequencies in cases and controls</span>
<span class="kn">parameter:</span> <span class="n">af_caco</span> <span class="o">=</span> <span class="s1">&#39;TRUE&#39;</span>
<span class="c1">#Path specific to SAIGE script</span>
<span class="kn">parameter:</span> <span class="n">script_path</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;~/software/bin/step2_SPAtests.R&#39;</span><span class="p">)</span>
<span class="c1"># Fix SAIGE non-standard sample file input</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="kn">as</span> <span class="nn">pd</span>
<span class="n">dat</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">sampleFile</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">skiprows</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="o">==</span> <span class="mi">1</span><span class="p">,</span> <span class="n">delim_whitespace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="n">dat</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;{cwd}/{sampleFile:bn}.SAIGE_sample&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">dat</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">0</span><span class="p">]])</span>

<span class="kn">input:</span> <span class="n">for_each</span><span class="o">=</span><span class="s1">&#39;genoFile&#39;</span>
<span class="n">input_options</span> <span class="o">=</span> <span class="n">f</span><span class="s2">&quot;--bgenFile={_genoFile} --bgenFileIndex=${_genoFile}.bgi --sampleFile=${cwd}/{sampleFile:bn}.SAIGE_sample --minInfo=${bgenMinINFO}&quot;</span> <span class="k">if</span> <span class="n">_input</span><span class="o">.</span><span class="n">suffix</span> <span class="o">==</span> <span class="s2">&quot;.bgen&quot;</span> <span class="k">else</span> <span class="n">f</span><span class="s2">&quot;--plinkFile={_input:n}&quot;</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{cwd}/cache/{_genoFile:bn}.{phenoFile:bn}.SAIGE.gz&#39;</span>
<span class="n">fail_if</span><span class="p">(</span><span class="ow">not</span> <span class="n">path</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;{_genoFile}.bgi&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="ow">and</span> <span class="n">_input</span><span class="o">.</span><span class="n">suffix</span> <span class="o">==</span> <span class="s1">&#39;.bgen&#39;</span><span class="p">,</span> <span class="n">msg</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;Cannot find file ``{_genoFile}.bgi``. Please generate it using command ``bgenix -g {_genoFile} -index``.&#39;</span><span class="p">)</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;48h&#39;</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;60G&#39;</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{step_name}_{_output:bn}&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.stdout&#39;</span><span class="p">,</span> <span class="n">template_name</span><span class="o">=</span><span class="s1">&#39;conda&#39;</span><span class="p">,</span> <span class="n">env_name</span><span class="o">=</span><span class="s1">&#39;RSAIGE&#39;</span>
    <span class="n">Rscript</span> <span class="err">$</span><span class="p">{</span><span class="n">script_path</span><span class="p">}</span> \
        <span class="err">$</span><span class="p">{</span><span class="n">input_options</span><span class="p">}</span>\
        <span class="o">--</span><span class="n">minMAF</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">bgenMinMAF</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">minMAC</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">bgenMinMAC</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">GMMATmodelFile</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span> \
        <span class="o">--</span><span class="n">varianceRatioFile</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]}</span> \
        <span class="o">--</span><span class="n">SAIGEOutputFile</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">numLinesOutput</span><span class="o">=</span><span class="mi">2</span> \
        <span class="o">--</span><span class="n">IsOutputAFinCaseCtrl</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">af_caco</span><span class="p">}</span> \
        <span class="o">&amp;&amp;</span> <span class="n">sed</span> <span class="s1">&#39;1 s/rsid //&#39;</span> <span class="o">-</span><span class="n">i</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span> \
        <span class="o">&amp;&amp;</span> <span class="n">gzip</span> <span class="o">-</span><span class="n">f</span> <span class="o">--</span><span class="n">best</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span> \
        <span class="o">&amp;&amp;</span> <span class="n">mv</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">bgen</span><span class="o">.</span><span class="n">txt</span><span class="o">.</span><span class="n">gz</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">}</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Output from each step:</p>
<p><strong>From step 1</strong></p>
<ol>
<li><p>Model file: <code>${_output}.rda</code></p>
</li>
<li><p>Association result file for the subset of randomly selected markers: <code>${_output}.results.txt</code></p>
</li>
<li><p>Variance ratio file: <code>${_output}.varianceRatio.txt</code></p>
</li>
</ol>
<p><strong>From step 2</strong></p>
<ol>
<li>A file with association results for each chromosome (Note: this are given in regard to Allele 2)</li>
</ol>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="GMMAT-workflow-implementation">GMMAT workflow implementation<a class="anchor-link" href="#GMMAT-workflow-implementation">&#182;</a></h2><p>Documentation can be found <a href="https://github.com/hanchenphd/GMMAT/blob/master/inst/doc/GMMAT.pdf">here</a></p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Step-1:-Creation-of-the-GRM">Step 1: Creation of the GRM<a class="anchor-link" href="#Step-1:-Creation-of-the-GRM">&#182;</a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1">#Calculate standardized GRM using GEMMA</span>
<span class="p">[</span><span class="n">gemma_grm</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">grmFile</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;{cwd}/{bfile:bn}.sXX.txt&#39;</span><span class="p">)</span>
<span class="kn">input:</span> <span class="n">bfile</span>
<span class="kn">output:</span> <span class="n">grmFile</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;2h&#39;</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;6G&#39;</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{step_name}_{_output[0]:bn}&#39;</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="s1">&#39;/mnt/mfs/statgen/containers/lmm.sif&#39;</span><span class="p">,</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.stdout&#39;</span> 
    <span class="n">gemma</span> \
    <span class="o">-</span><span class="n">bfile</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">n</span><span class="p">}</span> \
    <span class="o">-</span><span class="n">gk</span> <span class="mi">2</span> \
    <span class="o">-</span><span class="n">o</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">bnn</span><span class="p">}</span> \
    <span class="o">-</span><span class="n">outdir</span> <span class="err">$</span><span class="p">{</span><span class="n">grmFile</span><span class="p">:</span><span class="n">d</span><span class="p">}</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Step-2:-Fitting-the-null">Step 2: Fitting the null<a class="anchor-link" href="#Step-2:-Fitting-the-null">&#182;</a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># Run GMMAT step1: fit a GLMM with covariate adjustment and random effects to account for population structure and family or cryptic relatedness</span>
<span class="p">[</span><span class="n">GMMAT_null</span><span class="p">]</span>
<span class="c1">#use the standardized GRM file generated in gemma steps</span>
<span class="kn">parameter:</span> <span class="n">grmFile</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;{cwd}/{bfile:bn}.sXX.txt&#39;</span><span class="p">)</span>
<span class="c1">#a colum in the  data frame data, indicaing e id of samples</span>
<span class="kn">parameter:</span> <span class="n">phenoCol</span> <span class="o">=</span> <span class="s1">&#39;AD&#39;</span>
<span class="kn">parameter:</span> <span class="n">idCol</span> <span class="o">=</span> <span class="s1">&#39;IID&#39;</span>
<span class="kn">input:</span> <span class="n">phenoFile</span><span class="p">,</span> <span class="n">f</span><span class="s1">&#39;{bfile:n}.fam&#39;</span><span class="p">,</span> <span class="n">grmFile</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{cwd}/{bfile:bn}.{phenoFile:bn}.GMMAT.rds&#39;</span>
<span class="kn">R:</span> <span class="n">container</span><span class="o">=</span><span class="s1">&#39;/mnt/mfs/statgen/containers/lmm.sif&#39;</span><span class="p">,</span><span class="n">expand</span><span class="o">=</span><span class="s1">&#39;${ }&#39;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[0]:n}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[0]:n}.stdout&#39;</span>
    <span class="n">library</span><span class="p">(</span><span class="s1">&#39;GMMAT&#39;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s1">&#39;data.table&#39;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s1">&#39;dplyr&#39;</span><span class="p">)</span>
   
    <span class="c1">#Prepare phenotype and covariates in an R data frame</span>
    <span class="n">pheno</span> <span class="o">=</span> <span class="n">fread</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span> <span class="n">header</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span>
    <span class="c1">#Pheno are currently coded as 1, 2, and -9, need to recoded as 0,1,and NA</span>
    <span class="n">pheno</span><span class="err">$$</span><span class="p">{</span><span class="n">phenoCol</span><span class="p">}</span><span class="o">=</span>  <span class="n">recode</span><span class="p">(</span><span class="n">pheno</span><span class="err">$$</span><span class="p">{</span><span class="n">phenoCol</span><span class="p">},</span> <span class="sb">`2`</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="sb">`1`</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="sb">`-9`</span> <span class="o">=</span> <span class="n">NULL</span><span class="p">)</span>
    <span class="c1">#Prepare GRM file in a R data frame</span>
    <span class="n">GRM</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">fread</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">2</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span> <span class="n">header</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">))</span>
    <span class="c1">#Extract IIDs from .fam file</span>
    <span class="n">id_vector</span> <span class="o">=</span> <span class="n">fread</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span> <span class="n">header</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">)[,</span><span class="mi">2</span><span class="p">]</span>
    <span class="c1">#make the GRM colnames and rownames using the actual IID</span>
    <span class="n">colnames</span><span class="p">(</span><span class="n">GRM</span><span class="p">)</span> <span class="o">=</span> <span class="n">t</span><span class="p">(</span><span class="n">id_vector</span><span class="p">)</span>
    <span class="n">rownames</span><span class="p">(</span><span class="n">GRM</span><span class="p">)</span> <span class="o">=</span> <span class="n">t</span><span class="p">(</span><span class="n">id_vector</span><span class="p">)</span>

    <span class="n">fit_null</span> <span class="o">=</span> <span class="n">glmmkin</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">phenoCol</span><span class="p">}</span> <span class="o">~</span> <span class="err">$</span><span class="p">{</span><span class="s2">&quot;+&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">covarCol</span> <span class="o">+</span> <span class="n">qCovarCol</span><span class="p">)}</span> <span class="p">,</span> 
                           <span class="n">data</span> <span class="o">=</span> <span class="n">pheno</span><span class="p">,</span> 
                           <span class="n">kins</span> <span class="o">=</span> <span class="n">GRM</span><span class="p">,</span> 
                           <span class="nb">id</span> <span class="o">=</span> <span class="s1">&#39;${idCol}&#39;</span><span class="p">,</span>
                           <span class="n">family</span> <span class="o">=</span> <span class="n">binomial</span><span class="p">(</span><span class="n">link</span> <span class="o">=</span> <span class="s2">&quot;logit&quot;</span><span class="p">))</span>
    <span class="n">saveRDS</span><span class="p">(</span><span class="n">fit_null</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">})</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Step-3:-Perform-single-variant-score-test">Step 3: Perform single variant score test<a class="anchor-link" href="#Step-3:-Perform-single-variant-score-test">&#182;</a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1">#Run GMMAT step2: single variant score test(based on the null model built above)</span>
<span class="p">[</span><span class="n">GMMAT</span><span class="p">]</span>
<span class="c1">#the maximum rate alllowed for a variant to be included</span>
<span class="kn">parameter:</span> <span class="n">geno_filter</span> <span class="o">=</span> <span class="mf">0.01</span>
<span class="c1">#how many SNPs should be tested in a batch</span>
<span class="kn">parameter:</span> <span class="n">nperbatch</span> <span class="o">=</span> <span class="mi">100</span>
<span class="kn">depends:</span>  <span class="n">f</span><span class="s1">&#39;{cwd}/{bfile:bn}.{phenoFile:bn}.GMMAT.rds&#39;</span>
<span class="kn">input:</span> <span class="n">genoFile</span><span class="p">,</span> <span class="n">group_by</span> <span class="o">=</span> <span class="mi">1</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{cwd}/{_input:bn}.{phenoFile:bn}.gmmat.score.txt.gz&#39;</span>
<span class="kn">R:</span>  <span class="n">container</span><span class="o">=</span><span class="s1">&#39;/mnt/mfs/statgen/containers/lmm.sif&#39;</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span><span class="s1">&#39;${ }&#39;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[0]:n}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[0]:n}.stdout&#39;</span>
    <span class="n">library</span><span class="p">(</span><span class="s1">&#39;GMMAT&#39;</span><span class="p">)</span>
    <span class="n">null_model</span> <span class="o">=</span> <span class="n">readRDS</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_depends</span><span class="p">:</span><span class="n">r</span><span class="p">})</span>
    <span class="n">glmm</span><span class="o">.</span><span class="n">score</span><span class="p">(</span><span class="n">null_model</span><span class="p">,</span> 
               <span class="n">infile</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">nr</span><span class="p">},</span> 
               <span class="n">outfile</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">nr</span><span class="p">},</span> 
               <span class="n">MAF</span><span class="o">.</span><span class="n">range</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">bgenMinMAF</span><span class="p">},</span><span class="mi">1</span><span class="p">),</span> 
               <span class="n">miss</span><span class="o">.</span><span class="n">cutoff</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">geno_filter</span><span class="p">},</span>
               <span class="n">nperbatch</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">nperbatch</span><span class="p">})</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container_lmm</span><span class="p">,</span><span class="n">expand</span><span class="o">=</span><span class="s1">&#39;${ }&#39;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[0]:n}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[0]:n}.stdout&#39;</span>
    <span class="n">gzip</span> <span class="err">$</span><span class="p">{</span><span class="n">cwd</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">bn</span><span class="p">}</span><span class="o">.</span><span class="err">$</span><span class="p">{</span><span class="n">phenoFile</span><span class="p">:</span><span class="n">bn</span><span class="p">}</span><span class="o">.</span><span class="n">gmmat</span><span class="o">.</span><span class="n">score</span><span class="o">.</span><span class="n">txt</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Output from each step:</p>
<p><strong>From step 1</strong></p>
<p>Model file:<code>${_output}.rda</code></p>
<p><strong>From step 2</strong></p>
<p>A file with association results(score satistics, variance of score, score test P value) for each chromosome (Note: this are given in regard to Allele 2</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Merge-results">Merge results<a class="anchor-link" href="#Merge-results">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[1]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># Merge results and log files</span>
<span class="p">[</span><span class="n">boltlmm_2</span><span class="p">,</span> <span class="n">fastGWA_2</span><span class="p">,</span> <span class="n">SAIGE_3</span><span class="p">,</span> <span class="n">regenie_3</span><span class="p">,</span> <span class="n">regenie_burden_3</span><span class="p">,</span> <span class="n">GMMAT_1</span><span class="p">]</span>
<span class="kn">parameter:</span><span class="n">reverse_log_p</span> <span class="o">=</span> <span class="bp">False</span>
<span class="kn">depends:</span> <span class="n">formatFile</span>
<span class="kn">input:</span> <span class="n">group_by</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="n">i</span><span class="p">::</span><span class="nb">len</span><span class="p">(</span><span class="n">phenoCol</span><span class="p">)]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">phenoCol</span><span class="p">))],</span> <span class="n">group_with</span><span class="o">=</span><span class="s1">&#39;phenoCol&#39;</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{cwd}/{phenoFile:bn}_{_phenoCol}.{step_name.rsplit(&quot;_&quot;,1)[0]}.snp_stats.gz&#39;</span><span class="p">,</span>
        <span class="n">f</span><span class="s1">&#39;{cwd}/{phenoFile:bn}_{_phenoCol}.{step_name.rsplit(&quot;_&quot;,1)[0]}.snp_counts.txt&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;1h&#39;</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;20G&#39;</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{step_name}_{_output[0]:bn}&#39;</span>
<span class="kn">python:</span> <span class="n">container</span><span class="o">=</span><span class="n">container_lmm</span><span class="p">,</span> <span class="n">expand</span> <span class="o">=</span><span class="s1">&#39;${ }&#39;</span>
    <span class="kn">import</span> <span class="nn">gzip</span>
    <span class="kn">import</span> <span class="nn">pandas</span> <span class="kn">as</span> <span class="nn">pd</span>
    <span class="k">if</span> <span class="err">$</span><span class="p">{</span><span class="n">formatFile</span><span class="o">.</span><span class="n">is_file</span><span class="p">()}:</span>
        <span class="n">output</span> <span class="o">=</span> <span class="s1">&#39;${_output[0]:n}&#39;</span> <span class="o">+</span> <span class="s1">&#39;_original_columns&#39;</span> <span class="o">+</span> <span class="s1">&#39;${_output[0]:x}&#39;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">output</span> <span class="o">=</span> <span class="s1">&#39;${_output[0]}&#39;</span>
   
    <span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s1">&#39;gzip&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">delim_whitespace</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">quotechar</span><span class="o">=</span><span class="s1">&#39;&quot;&#39;</span><span class="p">,</span> <span class="n">comment</span><span class="o">=</span><span class="s1">&#39;#&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="p">[</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">,}]],</span> <span class="n">ignore_index</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
    <span class="n">data</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s1">&#39;gzip&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="bp">True</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="bp">False</span><span class="p">)</span>
    <span class="c1"># unify output format</span>
    <span class="k">if</span> <span class="err">$</span><span class="p">{</span><span class="n">formatFile</span><span class="o">.</span><span class="n">is_file</span><span class="p">()}</span> <span class="ow">or</span> <span class="err">$</span><span class="p">{</span><span class="n">reverse_log_p</span><span class="p">}:</span>
        <span class="n">sumstats</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">output</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s1">&#39;gzip&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">delim_whitespace</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">quotechar</span><span class="o">=</span><span class="s1">&#39;&quot;&#39;</span><span class="p">)</span>  
        <span class="k">if</span> <span class="err">$</span><span class="p">{</span><span class="n">formatFile</span><span class="o">.</span><span class="n">is_file</span><span class="p">()}:</span>
            <span class="kn">import</span> <span class="nn">yaml</span>
            <span class="n">config</span> <span class="o">=</span> <span class="n">yaml</span><span class="o">.</span><span class="n">safe_load</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">formatFile</span><span class="p">:</span><span class="n">r</span><span class="p">},</span> <span class="s1">&#39;r&#39;</span><span class="p">))</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">sumstats</span> <span class="o">=</span> <span class="n">sumstats</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span><span class="nb">list</span><span class="p">(</span><span class="n">config</span><span class="o">.</span><span class="n">values</span><span class="p">())]</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;According to ${formatFile}, input summary statistics should have the following columns: {list(config.values())}.&#39;</span><span class="p">)</span>
        <span class="n">sumstats</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">config</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
        <span class="k">if</span> <span class="err">$</span><span class="p">{</span><span class="n">reverse_log_p</span><span class="p">}:</span>
            <span class="n">sumstats</span><span class="p">[</span><span class="s1">&#39;P&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">sumstats</span><span class="p">[</span><span class="s1">&#39;P&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="mi">10</span><span class="o">**-</span><span class="n">row</span><span class="p">)</span>
        <span class="n">sumstats</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span> <span class="n">compression</span><span class="o">=</span><span class="s1">&#39;gzip&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="bp">True</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="bp">False</span><span class="p">)</span>        

<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container_lmm</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span><span class="s2">&quot;$( )&quot;</span>
    <span class="c1"># count result SNPs</span>
    <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="err">$</span><span class="p">(</span><span class="n">_input</span><span class="p">);</span> <span class="n">do</span> <span class="n">echo</span> <span class="s2">&quot;$f: `zcat $f | wc -l`&quot;</span><span class="p">;</span> <span class="n">done</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">(</span><span class="n">_output</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
    <span class="c1"># merge stderr and stdout files</span>
    <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="err">$</span><span class="p">(</span><span class="n">_input</span><span class="p">);</span> <span class="n">do</span> 
        <span class="k">for</span> <span class="n">ext</span> <span class="ow">in</span> <span class="n">stderr</span> <span class="n">stdout</span> <span class="n">log</span><span class="p">;</span> <span class="n">do</span>
            <span class="n">echo</span> <span class="s2">&quot;$f $ext:&quot;</span>
            <span class="n">cat</span> <span class="err">$</span><span class="p">{</span><span class="n">f</span><span class="o">%.</span><span class="n">gz</span><span class="p">}</span><span class="o">.</span><span class="err">$</span><span class="n">ext</span> <span class="mi">2</span><span class="o">&gt;/</span><span class="n">dev</span><span class="o">/</span><span class="n">null</span> <span class="o">||</span> <span class="n">true</span>
            <span class="n">rm</span> <span class="o">-</span><span class="n">f</span> <span class="err">$</span><span class="p">{</span><span class="n">f</span><span class="o">%.</span><span class="n">gz</span><span class="p">}</span><span class="o">.</span><span class="err">$</span><span class="n">ext</span> 
        <span class="n">done</span>
    <span class="n">done</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">(</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">)</span><span class="o">.</span><span class="n">log</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Manhattan-and-QQ-plots">Manhattan and QQ plots<a class="anchor-link" href="#Manhattan-and-QQ-plots">&#182;</a></h2><p>Before running the pipeline make sure you have installed the necessary packages. We use the <code>qqman</code> package from R: <a href="https://www.r-graph-gallery.com/101_Manhattan_plot.html">https://www.r-graph-gallery.com/101_Manhattan_plot.html</a></p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[3]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># Manhattan and QQ plots using `qqman`</span>
<span class="p">[</span><span class="n">boltlmm_3</span><span class="p">,</span> <span class="n">fastGWA_3</span><span class="p">,</span> <span class="n">SAIGE_4</span><span class="p">,</span> <span class="n">regenie_4</span><span class="p">,</span> <span class="n">regenie_burden_4</span><span class="p">,</span> <span class="n">GMMAT_2</span><span class="p">]</span>
<span class="c1"># Column name for BP</span>
<span class="kn">parameter:</span> <span class="n">bp</span> <span class="o">=</span> <span class="s1">&#39;POS&#39;</span>
<span class="c1"># Column name for p-value</span>
<span class="kn">parameter:</span> <span class="n">pval</span> <span class="o">=</span> <span class="s1">&#39;P&#39;</span>
<span class="c1"># Column name for SNP</span>
<span class="kn">parameter:</span> <span class="n">snp</span> <span class="o">=</span> <span class="s1">&#39;SNP&#39;</span>
<span class="c1"># Plot only on p-values smaller than this</span>
<span class="kn">parameter:</span> <span class="n">p_filter</span> <span class="o">=</span> <span class="s1">&#39;0.05&#39;</span>
<span class="c1"># ylim set to 0 to use maximum -log10(p) in data</span>
<span class="kn">parameter:</span> <span class="n">ylim</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\n\n</span><span class="s1">---</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="k">if</span> <span class="nb">any</span><span class="p">([</span><span class="s1">&#39;fastGWA&#39;</span> <span class="ow">in</span> <span class="n">step_name</span><span class="p">]):</span>
    <span class="n">heritability</span> <span class="o">=</span> <span class="n">get_output</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;grep Heritability {_input[0]:n}.log | head -1&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
<span class="kn">else:</span>
    <span class="n">heritability</span> <span class="o">=</span> <span class="bp">None</span>
<span class="kn">depends:</span> <span class="n">phenoFile</span>
<span class="kn">input:</span> <span class="n">group_by</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">group_with</span> <span class="o">=</span> <span class="s1">&#39;phenoCol&#39;</span>
<span class="kn">output:</span> <span class="n">manhattan</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_input[0]:nn}.manhattan.png&#39;</span><span class="p">,</span>
        <span class="n">qq</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_input[0]:nn}.qq.png&#39;</span><span class="p">,</span>
        <span class="n">annotated_manhattan</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_input[0]:nn}.manhattan_annotated.png&#39;</span><span class="p">,</span>
        <span class="n">analysis_summary</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_input[0]:nn}.analysis_summary.md&#39;</span><span class="p">,</span>
        <span class="n">plot_data</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_input[0]:nn}.plot_data.rds&#39;</span>       
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;3h&#39;</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;48G&#39;</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{step_name}_{_output[0]:bn}&#39;</span>    
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container_lmm</span><span class="p">,</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span>
    <span class="n">echo</span> <span class="s1">&#39;&#39;&#39;---</span>
<span class="s1">    theme: base-theme</span>
<span class="s1">    style: |</span>
<span class="s1">      img {</span>
<span class="s1">        height: 80%;</span>
<span class="s1">        display: block;</span>
<span class="s1">        margin-left: auto;</span>
<span class="s1">        margin-right: auto;</span>
<span class="s1">      }</span>
<span class="s1">    ---    </span>
<span class="s1">    &#39;&#39;&#39;</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">3</span><span class="p">]}</span>
    
<span class="kn">R:</span> <span class="n">container</span><span class="o">=</span><span class="n">container_lmm</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span><span class="s1">&#39;${ }&#39;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[0]:n}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[0]:n}.stdout&#39;</span>
    <span class="c1"># some summary statistics for phenotype</span>
    <span class="n">pheno</span> <span class="o">=</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">phenoFile</span><span class="p">:</span><span class="n">r</span><span class="p">},</span> <span class="n">header</span><span class="o">=</span><span class="n">T</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span><span class="err">$$</span><span class="p">{</span><span class="n">_phenoCol</span><span class="p">}</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">pheno</span><span class="p">))</span><span class="o">&gt;</span><span class="mi">2</span><span class="p">)</span> <span class="p">{</span>
      <span class="n">out</span> <span class="o">=</span> <span class="n">capture</span><span class="o">.</span><span class="n">output</span><span class="p">(</span><span class="n">summary</span><span class="p">(</span><span class="n">pheno</span><span class="p">))</span>
    <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
      <span class="n">out</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">pheno</span><span class="p">))</span>
      <span class="n">rownames</span><span class="p">(</span><span class="n">out</span><span class="p">)</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s1">&#39;n_ctrl&#39;</span><span class="p">,</span> <span class="s1">&#39;n_case&#39;</span><span class="p">)</span>
      <span class="n">out</span> <span class="o">=</span> <span class="n">out</span><span class="p">[,</span><span class="mi">2</span><span class="p">,</span><span class="n">drop</span><span class="o">=</span><span class="n">F</span><span class="p">]</span>
    <span class="p">}</span>
    <span class="n">write</span><span class="p">(</span><span class="s1">&#39;# ${_phenoCol} result summary</span><span class="se">\n</span><span class="s1">## Phenotype summary:</span><span class="se">\n</span><span class="s1">```&#39;</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">3</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span> <span class="n">append</span> <span class="o">=</span> <span class="n">T</span><span class="p">)</span>
    <span class="n">write</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">out</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">3</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span> <span class="n">append</span> <span class="o">=</span> <span class="n">T</span><span class="p">)</span>
    <span class="n">write</span><span class="p">(</span><span class="s1">&#39;${(&quot; Heritability is </span><span class="si">%s</span><span class="s1">&quot; </span><span class="si">% he</span><span class="s1">ritability) if heritability is not None else &#39;&#39;}&#39;</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">3</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span> <span class="n">append</span> <span class="o">=</span> <span class="n">T</span><span class="p">)</span>
    <span class="n">write</span><span class="p">(</span><span class="s2">&quot;```&quot;</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">3</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span> <span class="n">append</span> <span class="o">=</span> <span class="n">T</span><span class="p">)</span>

<span class="kn">R:</span> <span class="n">container</span><span class="o">=</span><span class="n">container_lmm</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span><span class="s1">&#39;${ }&#39;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[0]:n}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[0]:n}.stdout&#39;</span>
    <span class="n">library</span><span class="p">(</span><span class="s1">&#39;qqman&#39;</span><span class="p">)</span>
    <span class="n">data</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">gzfile</span><span class="p">(</span><span class="s1">&#39;${_input[0]}&#39;</span><span class="p">),</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="n">T</span><span class="p">)</span>
    <span class="k">lambda</span> <span class="o">&lt;-</span> <span class="n">median</span><span class="p">(</span><span class="n">qchisq</span><span class="p">(</span><span class="mi">1</span><span class="o">-</span><span class="n">data</span><span class="err">$$</span><span class="p">{</span><span class="n">pval</span><span class="p">},</span><span class="mi">1</span><span class="p">))</span><span class="o">/</span><span class="n">qchisq</span><span class="p">(</span><span class="mf">0.5</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">ifelse</span><span class="p">((</span><span class="err">$</span><span class="p">{</span><span class="n">ylim</span><span class="p">}</span> <span class="o">==</span> <span class="mi">0</span> <span class="o">&amp;&amp;</span> <span class="nb">min</span><span class="p">(</span><span class="n">data</span><span class="err">$$</span><span class="p">{</span><span class="n">pval</span><span class="p">},</span> <span class="n">na</span><span class="o">.</span><span class="n">rm</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span><span class="o">!=</span><span class="mi">0</span><span class="p">),</span> <span class="n">ylim</span> <span class="o">&lt;-</span> <span class="nb">abs</span><span class="p">(</span><span class="n">floor</span><span class="p">(</span><span class="n">log10</span><span class="p">(</span><span class="nb">min</span><span class="p">(</span><span class="n">data</span><span class="err">$$</span><span class="p">{</span><span class="n">pval</span><span class="p">},</span> <span class="n">na</span><span class="o">.</span><span class="n">rm</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)))),</span> <span class="n">ylim</span> <span class="o">&lt;-</span> <span class="nb">abs</span><span class="p">(</span><span class="n">floor</span><span class="p">(</span><span class="n">log10</span><span class="p">(</span><span class="mf">2.225074e-308</span><span class="p">))))</span>
    <span class="c1"># Creating manhattan plot</span>
    <span class="n">png</span><span class="p">(</span><span class="s1">&#39;${_output[0]}&#39;</span><span class="p">,</span> <span class="n">width</span> <span class="o">=</span> <span class="mi">6</span><span class="p">,</span> <span class="n">height</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span> <span class="n">unit</span><span class="o">=</span><span class="s1">&#39;in&#39;</span><span class="p">,</span> <span class="n">res</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">manhattan_plot</span> <span class="o">&lt;-</span> <span class="n">manhattan</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="nb">chr</span><span class="o">=</span><span class="s1">&#39;CHR&#39;</span><span class="p">,</span> <span class="n">bp</span><span class="o">=</span><span class="s1">&#39;${bp}&#39;</span><span class="p">,</span> <span class="n">snp</span><span class="o">=</span><span class="s1">&#39;${snp}&#39;</span><span class="p">,</span> <span class="n">p</span><span class="o">=</span><span class="s1">&#39;${pval}&#39;</span><span class="p">,</span> <span class="n">main</span> <span class="o">=</span> <span class="s1">&#39;Manhattan plot for ${_phenoCol} (${step_name.rsplit(&quot;_&quot;,1)[0]})&#39;</span><span class="p">,</span> <span class="n">ylim</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">ylim</span><span class="p">),</span> <span class="n">cex</span> <span class="o">=</span> <span class="mf">0.6</span><span class="p">,</span> 
    <span class="n">cex</span><span class="o">.</span><span class="n">axis</span> <span class="o">=</span> <span class="mf">0.9</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;blue4&quot;</span><span class="p">,</span> <span class="s2">&quot;orange3&quot;</span><span class="p">),</span> <span class="n">chrlabs</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">character</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="mi">1</span><span class="p">:</span><span class="mi">22</span><span class="p">)))</span>
    <span class="n">dev</span><span class="o">.</span><span class="n">off</span><span class="p">()</span>
    <span class="c1"># Creating qqplot</span>
    <span class="n">png</span><span class="p">(</span><span class="s1">&#39;${_output[1]}&#39;</span><span class="p">,</span> <span class="n">width</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">height</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">unit</span><span class="o">=</span><span class="s1">&#39;in&#39;</span><span class="p">,</span> <span class="n">res</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">qq_plot</span> <span class="o">&lt;-</span> <span class="n">qq</span><span class="p">(</span><span class="n">data</span><span class="err">$$</span><span class="p">{</span><span class="n">pval</span><span class="p">},</span> <span class="n">main</span> <span class="o">=</span> <span class="s1">&#39;QQ Plot for ${_phenoCol} (${step_name.rsplit(&quot;_&quot;,1)[0]})&#39;</span><span class="p">,</span> <span class="n">xlim</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">8</span><span class="p">),</span> <span class="n">ylim</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">ylim</span><span class="p">),</span> <span class="n">pch</span> <span class="o">=</span> <span class="mi">18</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="s2">&quot;blue4&quot;</span><span class="p">,</span> <span class="n">cex</span> <span class="o">=</span> <span class="mf">1.5</span><span class="p">,</span> <span class="n">las</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">dev</span><span class="o">.</span><span class="n">off</span><span class="p">()</span>
    <span class="n">write</span><span class="p">(</span><span class="s1">&#39;## p-value summary:&#39;</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">3</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span> <span class="n">append</span><span class="o">=</span><span class="n">T</span><span class="p">)</span>
    <span class="n">write</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">&quot;Genomic inflation factor is&quot;</span><span class="p">,</span> <span class="nb">round</span><span class="p">(</span><span class="k">lambda</span><span class="p">,</span><span class="mi">3</span><span class="p">),</span> <span class="s2">&quot;for&quot;</span><span class="p">,</span> <span class="n">nrow</span><span class="p">(</span><span class="n">data</span><span class="p">),</span> <span class="s2">&quot;variants analyzed.${sep}&quot;</span><span class="p">),</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">3</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span> <span class="n">append</span><span class="o">=</span><span class="n">T</span><span class="p">)</span>
    
  
<span class="kn">R:</span> <span class="n">container</span><span class="o">=</span><span class="n">container_lmm</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span><span class="s1">&#39;${ }&#39;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[2]:n}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[2]:n}.stdout&#39;</span>
    <span class="n">library</span><span class="p">(</span><span class="s1">&#39;dplyr&#39;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s1">&#39;ggrepel&#39;</span><span class="p">)</span>
    <span class="c1">#Load your data</span>
    <span class="n">data</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">gzfile</span><span class="p">(</span><span class="s1">&#39;${_input[0]}&#39;</span><span class="p">),</span><span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="n">T</span><span class="p">)</span>
    <span class="c1"># Create a subset of the data with variants with P&lt; 0.05 and arrange by chromosome number</span>
    <span class="c1"># https://danielroelfs.com/blog/how-i-create-manhattan-plots-using-ggplot/</span>
    <span class="n">sig</span><span class="o">.</span><span class="n">dat</span> <span class="o">&lt;-</span> <span class="n">data</span> <span class="o">%&gt;%</span> 
      <span class="n">subset</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">pval</span><span class="p">}</span> <span class="o">&lt;</span> <span class="err">$</span><span class="p">{</span><span class="n">p_filter</span><span class="p">})</span> <span class="o">%&gt;%</span>
      <span class="n">arrange</span> <span class="p">(</span><span class="n">CHR</span><span class="p">,</span> <span class="o">.</span><span class="n">by_group</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
    <span class="c1"># Add highlight and annotation information</span>
    <span class="c1">#mutate( is_highlight=ifelse(SNP %in% index_snps, &quot;yes&quot;, &quot;no&quot;)) %&gt;%</span>
    <span class="c1">#mutate( is_annotate=ifelse(-log10(P_BOLT_LMM)&gt;6, &quot;yes&quot;, &quot;no&quot;)) </span>
    <span class="c1"># Check the list of chromosomes (make sure the sex chr are at the end of the list)</span>
    <span class="c1"># Get the cumulative base pair position for each variant</span>
    <span class="n">nCHR</span> <span class="o">&lt;-</span> <span class="n">length</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">sig</span><span class="o">.</span><span class="n">dat</span><span class="err">$</span><span class="n">CHR</span><span class="p">))</span>
    <span class="n">sig</span><span class="o">.</span><span class="n">dat</span><span class="err">$</span><span class="n">BPcum</span> <span class="o">&lt;-</span> <span class="n">NA</span>
    <span class="n">s</span> <span class="o">&lt;-</span> <span class="mi">0</span>
    <span class="n">nbp</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">()</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="n">unique</span><span class="p">(</span><span class="n">sig</span><span class="o">.</span><span class="n">dat</span><span class="err">$</span><span class="n">CHR</span><span class="p">)){</span>
      <span class="n">nbp</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="nb">max</span><span class="p">(</span><span class="n">sig</span><span class="o">.</span><span class="n">dat</span><span class="p">[</span><span class="n">sig</span><span class="o">.</span><span class="n">dat</span><span class="err">$</span><span class="n">CHR</span> <span class="o">==</span> <span class="n">i</span><span class="p">,]</span><span class="err">$$</span><span class="p">{</span><span class="n">bp</span><span class="p">})</span>
      <span class="n">sig</span><span class="o">.</span><span class="n">dat</span><span class="p">[</span><span class="n">sig</span><span class="o">.</span><span class="n">dat</span><span class="err">$</span><span class="n">CHR</span> <span class="o">==</span> <span class="n">i</span><span class="p">,</span><span class="s2">&quot;BPcum&quot;</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="n">sig</span><span class="o">.</span><span class="n">dat</span><span class="p">[</span><span class="n">sig</span><span class="o">.</span><span class="n">dat</span><span class="err">$</span><span class="n">CHR</span> <span class="o">==</span> <span class="n">i</span><span class="p">,</span><span class="s2">&quot;${bp}&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">s</span>
      <span class="n">s</span> <span class="o">&lt;-</span> <span class="n">s</span> <span class="o">+</span> <span class="n">nbp</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
    <span class="p">}</span>

    <span class="c1"># Calculate the mid point for each chromosome for plotting the x-axis</span>
    <span class="c1"># Calculate the y-lim </span>

    <span class="n">axis</span><span class="o">.</span><span class="n">set</span> <span class="o">&lt;-</span> <span class="n">sig</span><span class="o">.</span><span class="n">dat</span> <span class="o">%&gt;%</span> 
      <span class="n">group_by</span><span class="p">(</span><span class="n">CHR</span><span class="p">)</span> <span class="o">%&gt;%</span> 
      <span class="n">summarize</span><span class="p">(</span><span class="n">center</span> <span class="o">=</span> <span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">BPcum</span><span class="p">)</span> <span class="o">+</span> <span class="nb">min</span><span class="p">(</span><span class="n">BPcum</span><span class="p">))</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
    <span class="k">if</span> <span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">ylim</span><span class="p">}</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="n">ylim</span> <span class="o">&lt;-</span> <span class="nb">abs</span><span class="p">(</span><span class="n">floor</span><span class="p">(</span><span class="n">log10</span><span class="p">(</span><span class="nb">min</span><span class="p">(</span><span class="n">sig</span><span class="o">.</span><span class="n">dat</span><span class="err">$$</span><span class="p">{</span><span class="n">pval</span><span class="p">}))))</span> <span class="o">+</span> <span class="mi">2</span> 
    <span class="n">sig</span> <span class="o">&lt;-</span> <span class="mf">5e-8</span>

    <span class="c1"># Now time to draw the manhattan plot without filtering the most significant signals</span>
    <span class="n">manhplot</span> <span class="o">&lt;-</span> <span class="n">ggplot</span><span class="p">(</span><span class="n">sig</span><span class="o">.</span><span class="n">dat</span><span class="p">,</span> <span class="n">aes</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="n">BPcum</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="o">-</span><span class="n">log10</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">pval</span><span class="p">}),</span> 
                                 <span class="n">color</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">factor</span><span class="p">(</span><span class="n">CHR</span><span class="p">),</span> <span class="n">size</span> <span class="o">=</span> <span class="o">-</span><span class="n">log10</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">pval</span><span class="p">})))</span> <span class="o">+</span>
      <span class="n">geom_point</span><span class="p">(</span><span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.75</span><span class="p">)</span> <span class="o">+</span>
      <span class="n">geom_hline</span><span class="p">(</span><span class="n">yintercept</span> <span class="o">=</span> <span class="o">-</span><span class="n">log10</span><span class="p">(</span><span class="n">sig</span><span class="p">),</span> <span class="n">color</span> <span class="o">=</span> <span class="s2">&quot;red1&quot;</span><span class="p">,</span> <span class="n">linetype</span> <span class="o">=</span> <span class="s2">&quot;dashed&quot;</span><span class="p">)</span> <span class="o">+</span> 
      <span class="n">scale_x_continuous</span><span class="p">(</span><span class="n">label</span> <span class="o">=</span> <span class="n">axis</span><span class="o">.</span><span class="n">set</span><span class="err">$</span><span class="n">CHR</span><span class="p">,</span> <span class="n">breaks</span> <span class="o">=</span> <span class="n">axis</span><span class="o">.</span><span class="n">set</span><span class="err">$</span><span class="n">center</span><span class="p">)</span> <span class="o">+</span>
      <span class="n">scale_y_continuous</span><span class="p">(</span><span class="n">expand</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">),</span> <span class="n">limits</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">ylim</span><span class="p">))</span> <span class="o">+</span>
      <span class="n">scale_color_manual</span><span class="p">(</span><span class="n">values</span> <span class="o">=</span> <span class="n">rep</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="s2">&quot;#276FBF&quot;</span><span class="p">,</span> <span class="s2">&quot;#183059&quot;</span><span class="p">),</span> <span class="n">nCHR</span><span class="p">))</span> <span class="o">+</span>
      <span class="n">scale_size_continuous</span><span class="p">(</span><span class="nb">range</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="mf">0.5</span><span class="p">,</span><span class="mi">3</span><span class="p">))</span> <span class="o">+</span>
      <span class="c1"># Add highlighted points</span>
      <span class="c1"># geom_point(data=subset(sig.dat, is_highlight==&quot;yes&quot;), color=&quot;orange&quot;, alpha=0.75) +</span>
      <span class="n">labs</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="s2">&quot;Chromosome&quot;</span><span class="p">,</span> 
           <span class="n">y</span> <span class="o">=</span> <span class="s2">&quot;-log10(p)&quot;</span><span class="p">,</span>
           <span class="n">title</span> <span class="o">=</span><span class="s1">&#39;Manhattan plot for ${_phenoCol} (${step_name.rsplit(&quot;_&quot;,1)[0]})&#39;</span><span class="p">)</span> <span class="o">+</span> 
      <span class="n">theme_classic</span><span class="p">()</span> <span class="o">+</span>
      <span class="n">theme</span><span class="p">(</span> 
        <span class="n">legend</span><span class="o">.</span><span class="n">position</span> <span class="o">=</span> <span class="s2">&quot;none&quot;</span><span class="p">,</span>
        <span class="n">panel</span><span class="o">.</span><span class="n">border</span> <span class="o">=</span> <span class="n">element_blank</span><span class="p">(),</span>
        <span class="n">panel</span><span class="o">.</span><span class="n">grid</span><span class="o">.</span><span class="n">major</span><span class="o">.</span><span class="n">x</span> <span class="o">=</span> <span class="n">element_blank</span><span class="p">(),</span>
        <span class="n">panel</span><span class="o">.</span><span class="n">grid</span><span class="o">.</span><span class="n">minor</span><span class="o">.</span><span class="n">x</span> <span class="o">=</span> <span class="n">element_blank</span><span class="p">(),</span>
        <span class="n">axis</span><span class="o">.</span><span class="n">text</span><span class="o">.</span><span class="n">x</span> <span class="o">=</span> <span class="n">element_text</span><span class="p">(</span><span class="n">angle</span> <span class="o">=</span> <span class="mi">90</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="mi">8</span><span class="p">,</span> <span class="n">vjust</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">)</span>
      <span class="p">)</span>

    <span class="c1"># To save a plot created with ggplot2 you have to use to print() function</span>

    <span class="n">png</span><span class="p">(</span><span class="s1">&#39;${_output[2]}&#39;</span><span class="p">,</span> <span class="n">width</span> <span class="o">=</span> <span class="mi">6</span><span class="p">,</span> <span class="n">height</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span> <span class="n">unit</span><span class="o">=</span><span class="s1">&#39;in&#39;</span><span class="p">,</span> <span class="n">res</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="k">print</span><span class="p">(</span><span class="n">manhplot</span><span class="p">)</span>
    <span class="n">dev</span><span class="o">.</span><span class="n">off</span><span class="p">()</span>
  
    <span class="c1"># save significant data to a file for further evaluations</span>
    <span class="n">tmp</span> <span class="o">=</span> <span class="n">sig</span><span class="o">.</span><span class="n">dat</span><span class="p">[,</span><span class="n">c</span><span class="p">(</span><span class="s1">&#39;CHR&#39;</span><span class="p">,</span> <span class="s1">&#39;${bp}&#39;</span><span class="p">,</span> <span class="s1">&#39;BPcum&#39;</span><span class="p">,</span> <span class="s1">&#39;${snp}&#39;</span><span class="p">,</span> <span class="s1">&#39;${pval}&#39;</span><span class="p">)]</span>
    <span class="n">colnames</span><span class="p">(</span><span class="n">tmp</span><span class="p">)</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s1">&#39;CHR&#39;</span><span class="p">,</span> <span class="s1">&#39;POS&#39;</span><span class="p">,</span> <span class="s1">&#39;POScum&#39;</span><span class="p">,</span> <span class="s1">&#39;SNP&#39;</span><span class="p">,</span> <span class="s1">&#39;pvalue&#39;</span><span class="p">)</span>
    <span class="n">saveRDS</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">tmp</span><span class="p">,</span> 
                 <span class="n">ylim</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">floor</span><span class="p">(</span><span class="n">log10</span><span class="p">(</span><span class="nb">min</span><span class="p">(</span><span class="n">sig</span><span class="o">.</span><span class="n">dat</span><span class="err">$$</span><span class="p">{</span><span class="n">pval</span><span class="p">}))))</span> <span class="o">+</span> <span class="mi">2</span><span class="p">,</span>
                 <span class="n">axis</span><span class="o">.</span><span class="n">set</span> <span class="o">=</span> <span class="n">axis</span><span class="o">.</span><span class="n">set</span><span class="p">),</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">4</span><span class="p">]:</span><span class="n">r</span><span class="p">})</span>

<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container_lmm</span><span class="p">,</span> <span class="n">expand</span> <span class="o">=</span> <span class="bp">True</span>
  <span class="nb">set</span> <span class="o">-</span><span class="n">e</span>
  <span class="n">echo</span> <span class="o">-</span><span class="n">e</span> <span class="s2">&quot;# QQ plot for {_phenoCol}</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">&gt;&gt;</span> <span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">3</span><span class="p">]}</span>
  <span class="n">echo</span> <span class="o">-</span><span class="n">e</span> <span class="s2">&quot;![]({_output[1]:bn}.png){sep}&quot;</span> <span class="o">&gt;&gt;</span> <span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">3</span><span class="p">]}</span>
  <span class="n">echo</span> <span class="o">-</span><span class="n">e</span> <span class="s2">&quot;# Manhattan plot for {_phenoCol}</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">&gt;&gt;</span> <span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">3</span><span class="p">]}</span>
  <span class="n">echo</span> <span class="o">-</span><span class="n">e</span> <span class="s2">&quot;![]({_output[0]:bn}.png){sep}&quot;</span> <span class="o">&gt;&gt;</span> <span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">3</span><span class="p">]}</span>
  <span class="n">echo</span> <span class="o">-</span><span class="n">e</span> <span class="s2">&quot;# Manhattan plot for {_phenoCol}</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">&gt;&gt;</span> <span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">3</span><span class="p">]}</span>
  <span class="n">echo</span> <span class="o">-</span><span class="n">e</span> <span class="s2">&quot;![]({_output[2]:bn}.png){sep}&quot;</span> <span class="o">&gt;&gt;</span> <span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">3</span><span class="p">]}</span>
  <span class="n">echo</span> <span class="o">-</span><span class="n">e</span> <span class="s2">&quot;# Result files</span><span class="se">\n</span><span class="s2">\`\`\`&quot;</span> <span class="o">&gt;&gt;</span> <span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">3</span><span class="p">]}</span>
  <span class="n">ls</span> <span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">nn</span><span class="p">}</span><span class="o">.*</span> <span class="o">|</span> <span class="n">grep</span> <span class="o">-</span><span class="n">vP</span> <span class="s1">&#39;stderr|stdout&#39;</span><span class="o">&gt;&gt;</span> <span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">3</span><span class="p">]}</span>
  <span class="n">echo</span> <span class="o">-</span><span class="n">e</span> <span class="s2">&quot;\`\`\`&quot;</span> <span class="o">&gt;&gt;</span> <span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">3</span><span class="p">]}</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Create-analysis-report">Create analysis report<a class="anchor-link" href="#Create-analysis-report">&#182;</a></h2><p>To install <code>marp</code>:</p>
<div class="highlight"><pre><span></span>npm install -g @marp-team/marp-cli
</pre></div>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[2]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># Generate analysis report: HTML file, and optionally PPTX file</span>
<span class="p">[</span><span class="n">boltlmm_4</span><span class="p">,</span> <span class="n">fastGWA_4</span><span class="p">,</span> <span class="n">SAIGE_5</span><span class="p">,</span> <span class="n">regenie_5</span><span class="p">,</span> <span class="n">regenie_burden_5</span><span class="p">,</span> <span class="n">GMMAT_3</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">group_by</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">group_with</span><span class="o">=</span><span class="s1">&#39;phenoCol&#39;</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s2">&quot;{_input[&#39;analysis_summary&#39;]:n}.html&quot;</span>
<span class="kn">sh:</span> <span class="n">container</span><span class="o">=</span><span class="n">container_marp</span><span class="p">,</span> <span class="n">expand</span> <span class="o">=</span> <span class="bp">True</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.stdout&#39;</span>
    <span class="n">node</span> <span class="o">/</span><span class="n">opt</span><span class="o">/</span><span class="n">marp</span><span class="o">/.</span><span class="n">cli</span><span class="o">/</span><span class="n">marp</span><span class="o">-</span><span class="n">cli</span><span class="o">.</span><span class="n">js</span> <span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="s1">&#39;analysis_summary&#39;</span><span class="p">]}</span> <span class="o">-</span><span class="n">o</span> <span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">a</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">title</span> <span class="s1">&#39;{_phenoCol} {step_name.rsplit(&quot;_&quot;,1)[0]} analysis&#39;</span> \
        <span class="o">--</span><span class="n">allow</span><span class="o">-</span><span class="n">local</span><span class="o">-</span><span class="n">files</span>
    <span class="n">node</span> <span class="o">/</span><span class="n">opt</span><span class="o">/</span><span class="n">marp</span><span class="o">/.</span><span class="n">cli</span><span class="o">/</span><span class="n">marp</span><span class="o">-</span><span class="n">cli</span><span class="o">.</span><span class="n">js</span> <span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="s1">&#39;analysis_summary&#39;</span><span class="p">]}</span> <span class="o">-</span><span class="n">o</span> <span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">an</span><span class="p">}</span><span class="o">.</span><span class="n">pptx</span> \
        <span class="o">--</span><span class="n">title</span> <span class="s1">&#39;{_phenoCol} {step_name.rsplit(&quot;_&quot;,1)[0]} analysis&#39;</span> \
        <span class="o">--</span><span class="n">allow</span><span class="o">-</span><span class="n">local</span><span class="o">-</span><span class="n">files</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Results">Results<a class="anchor-link" href="#Results">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Take BoltLMM for example, there are some analysis files:</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="o">%</span><span class="n">preview</span> <span class="n">output</span><span class="o">/</span><span class="n">phenotypes_BMI</span><span class="o">.</span><span class="n">boltlmm</span><span class="o">.</span><span class="n">snp_stats</span><span class="o">.</span><span class="n">gz</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="o">%</span><span class="n">preview</span> <span class="n">output</span><span class="o">/</span><span class="n">phenotypes_BMI</span><span class="o">.</span><span class="n">boltlmm</span><span class="o">.</span><span class="n">ref_stats</span><span class="o">.</span><span class="n">gz</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="o">%</span><span class="n">preview</span> <span class="n">output</span><span class="o">/</span><span class="n">phenotypes_BMI</span><span class="o">.</span><span class="n">boltlmm</span><span class="o">.</span><span class="n">snp_counts</span><span class="o">.</span><span class="n">txt</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="o">%</span><span class="n">preview</span> <span class="n">output</span><span class="o">/</span><span class="n">phenotypes_BMI</span><span class="o">.</span><span class="n">boltlmm</span><span class="o">.</span><span class="n">log</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>The result of analysis will be summarized to a <code>PPTX</code> file,</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="n">ls</span> <span class="n">output</span><span class="o">/*.</span><span class="n">boltlmm</span><span class="o">.</span><span class="n">analysis_summary</span><span class="o">.</span><span class="n">pptx</span>
</pre></div>

    </div>
</div>
</div>

</div>
<hr>
&copy 2020-current Gao Wang and others, Department of Neurology at Columbia University
<p><small>Exported from <a href="http://github.com/cumc/bioworkflows/blob/b3b5ab4f99caef55c50787c54c8307c008bd1cc7/GWAS/LMM.ipynb"><code>GWAS/LMM.ipynb</code></a> committed by Gao Wang on Thu May 20 23:53:28 2021 <a href="http://github.com/cumc/bioworkflows/commit/b3b5ab4f99caef55c50787c54c8307c008bd1cc7">revision 62, b3b5ab4</a> <a href="https://stephenslab.github.io/ipynb-website/notes.html#Note-about-commit-ids"><span class="fa fa-question-circle"></span></a></small></p>
</div>
</div>
</body>
</html>
