var tipuesearch = {"pages": [
{"mainTitle": "Welcome to Some computational genomics workflows!", "title": "Welcome to Some computational genomics workflows!", "text": "Welcome to Some computational genomics workflows ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/index.html", "url": "http://cumc.github.io/bioworkflows/index.html#Welcome-to-Some-computational-genomics-workflows!"},
{"mainTitle": "A multivariate EBNM approach for mixture multivariate distribution estimate", "title": "A multivariate EBNM approach for mixture multivariate distribution estimate", "text": "A multivariate EBNM approach for mixture multivariate distribution estimate ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mixture_prior.html", "url": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mixture_prior.html#A-multivariate-EBNM-approach-for-mixture-multivariate-distribution-estimate"},
{"mainTitle": "A multivariate EBNM approach for mixture multivariate distribution estimate", "title": "Overview of approach", "text": "Overview of approach ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mixture_prior.html", "url": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mixture_prior.html#Overview-of-approach"},
{"mainTitle": "A multivariate EBNM approach for mixture multivariate distribution estimate", "title": "Minimal working example", "text": "Minimal working example ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mixture_prior.html", "url": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mixture_prior.html#Minimal-working-example"},
{"mainTitle": "A multivariate EBNM approach for mixture multivariate distribution estimate", "title": "Merge univariate results", "text": "Merge univariate results ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mixture_prior.html", "url": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mixture_prior.html#Merge-univariate-results"},
{"mainTitle": "A multivariate EBNM approach for mixture multivariate distribution estimate", "title": "Select and merge univariate effects", "text": "Select and merge univariate effects ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mixture_prior.html", "url": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mixture_prior.html#Select-and-merge-univariate-effects"},
{"mainTitle": "A multivariate EBNM approach for mixture multivariate distribution estimate", "title": "Perform mixture model fitting", "text": "Perform mixture model fitting ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mixture_prior.html", "url": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mixture_prior.html#Perform-mixture-model-fitting"},
{"mainTitle": "A multivariate EBNM approach for mixture multivariate distribution estimate", "title": "Plot results", "text": "Plot results ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mixture_prior.html", "url": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mixture_prior.html#Plot-results"},
{"mainTitle": "A multivariate EBNM approach for mixture multivariate distribution estimate", "title": "Global parameters", "text": "Global parameters ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mixture_prior.html", "url": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mixture_prior.html#Global-parameters"},
{"mainTitle": "A multivariate EBNM approach for mixture multivariate distribution estimate", "title": "Merge PLINK univariate association summary statistic to RDS format", "text": "Merge PLINK univariate association summary statistic to RDS format ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mixture_prior.html", "url": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mixture_prior.html#Merge-PLINK-univariate-association-summary-statistic-to-RDS-format"},
{"mainTitle": "A multivariate EBNM approach for mixture multivariate distribution estimate", "title": "Get top, random and null effects per analysis unit", "text": "Get top, random and null effects per analysis unit ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mixture_prior.html", "url": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mixture_prior.html#Get-top,-random-and-null-effects-per-analysis-unit"},
{"mainTitle": "A multivariate EBNM approach for mixture multivariate distribution estimate", "title": "Factor analyses", "text": "Factor analyses ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mixture_prior.html", "url": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mixture_prior.html#Factor-analyses"},
{"mainTitle": "A multivariate EBNM approach for mixture multivariate distribution estimate", "title": "Fit mixture model", "text": "Fit mixture model ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mixture_prior.html", "url": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mixture_prior.html#Fit-mixture-model"},
{"mainTitle": "A multivariate EBNM approach for mixture multivariate distribution estimate", "title": "Plot patterns of sharing", "text": "Plot patterns of sharing ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mixture_prior.html", "url": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mixture_prior.html#Plot-patterns-of-sharing"},
{"mainTitle": "Multivariate fine-mapping workflow", "title": "Multivariate fine-mapping workflow", "text": "Multivariate fine-mapping workflow ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mvSuSiE.html", "url": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mvSuSiE.html#Multivariate-fine-mapping-workflow"},
{"mainTitle": "Multivariate fine-mapping workflow", "title": "Input", "text": "Input ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mvSuSiE.html", "url": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mvSuSiE.html#Input"},
{"mainTitle": "Multivariate fine-mapping workflow", "title": "Output", "text": "Output ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mvSuSiE.html", "url": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mvSuSiE.html#Output"},
{"mainTitle": "Multivariate fine-mapping workflow", "title": "Analysis examples", "text": "Analysis examples ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mvSuSiE.html", "url": "http://cumc.github.io/bioworkflows/multivariate-fine-mapping/mvSuSiE.html#Analysis-examples"},
{"mainTitle": "Fine-mapping result post processing", "title": "Fine-mapping result post processing", "text": "Fine-mapping result post processing ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/finemapping_results_wrangler.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/finemapping_results_wrangler.html#Fine-mapping-result-post-processing"},
{"mainTitle": "Fine-mapping result post processing", "title": "Input data", "text": "Input data ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/finemapping_results_wrangler.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/finemapping_results_wrangler.html#Input-data"},
{"mainTitle": "Fine-mapping result post processing", "title": "Output data", "text": "Output data ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/finemapping_results_wrangler.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/finemapping_results_wrangler.html#Output-data"},
{"mainTitle": "Fine-mapping result post processing", "title": "Additional data processing", "text": "Additional data processing ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/finemapping_results_wrangler.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/finemapping_results_wrangler.html#Additional-data-processing"},
{"mainTitle": "Fine-mapping result post processing", "title": "The workflow", "text": "The workflow ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/finemapping_results_wrangler.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/finemapping_results_wrangler.html#The-workflow"},
{"mainTitle": "Fine-mapping result post processing", "title": "Fine-mapping results consolidation", "text": "Fine-mapping results consolidation ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/finemapping_results_wrangler.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/finemapping_results_wrangler.html#Fine-mapping-results-consolidation"},
{"mainTitle": "Fine-mapping result post processing", "title": "Update variant ID via genomic coordinantes", "text": "Update variant ID via genomic coordinantes ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/finemapping_results_wrangler.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/finemapping_results_wrangler.html#Update-variant-ID-via-genomic-coordinantes"},
{"mainTitle": "Fine-mapping result post processing", "title": "Update genomic coordinates via variant ID", "text": "Update genomic coordinates via variant ID ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/finemapping_results_wrangler.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/finemapping_results_wrangler.html#Update-genomic-coordinates-via-variant-ID"},
{"mainTitle": "Fine-mapping result post processing", "title": "Example run", "text": "Example run ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/finemapping_results_wrangler.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/finemapping_results_wrangler.html#Example-run"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "text": "LDpred2 Pipeline for Polygenic Risk Score Prediction ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#LDpred2-Pipeline-for-Polygenic-Risk-Score-Prediction"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Overview", "text": "Overview ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Overview"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Methods", "text": "Methods ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Methods"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "PRS model", "text": "PRS model ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#PRS-model"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "LD reference panel", "text": "LD reference panel ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#LD-reference-panel"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "LDpred prior effect size model", "text": "LDpred prior effect size model ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#LDpred-prior-effect-size-model"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Heritability (parameter) explained by the genotypes", "text": "Heritability parameter explained by the genotypes ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Heritability-(parameter)-explained-by-the-genotypes"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Fraction of causal markers (i.e., the fraction of markers with non-zero effects)", "text": "Fraction of causal markers i.e., the fraction of markers with non-zero effects ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Fraction-of-causal-markers-(i.e.,-the-fraction-of-markers-with-non-zero-effects)"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "LDpred-inf (infinitesimal model)", "text": "LDpred-inf infinitesimal model ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#LDpred-inf-(infinitesimal-model)"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "LDpred-grid/auto (non-infinitesimal model)", "text": "LDpred-grid auto non-infinitesimal model ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#LDpred-grid/auto-(non-infinitesimal-model)"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "LDpred2", "text": "LDpred2 ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#LDpred2"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Other Methods for PRS Prediction", "text": "Other Methods for PRS Prediction ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Other-Methods-for-PRS-Prediction"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Method for Phenotype Prediction", "text": "Method for Phenotype Prediction ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Method-for-Phenotype-Prediction"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Required software", "text": "Required software ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Required-software"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Input", "text": "Input ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Input"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Output", "text": "Output ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Output"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "General workflow", "text": "General workflow ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#General-workflow"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Preliminary quality control on reference panel", "text": "Preliminary quality control on reference panel ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Preliminary-quality-control-on-reference-panel"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Output", "text": "Output ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Output"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Intersect SNPs among summary stats, reference panel and target data", "text": "Intersect SNPs among summary stats, reference panel and target data ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Intersect-SNPs-among-summary-stats,-reference-panel-and-target-data"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Output", "text": "Output ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Output"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "SNP Matching", "text": "SNP Matching ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#SNP-Matching"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Output", "text": "Output ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Output"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Quality Control on Summary Statistics", "text": "Quality Control on Summary Statistics ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Quality-Control-on-Summary-Statistics"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Output", "text": "Output ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Output"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Calculate LD matrix and Fit LDSC Model", "text": "Calculate LD matrix and Fit LDSC Model ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Calculate-LD-matrix-and-Fit-LDSC-Model"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Output", "text": "Output ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Output"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Estimate posterior effect sizes and PRS", "text": "Estimate posterior effect sizes and PRS ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Estimate-posterior-effect-sizes-and-PRS"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Output", "text": "Output ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Output"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Predict phenotype", "text": "Predict phenotype ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Predict-phenotype"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Output", "text": "Output ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Output"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Example analysis", "text": "Example analysis ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Example-analysis"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Command Interface", "text": "Command Interface ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Command-Interface"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Global Parameter Setting", "text": "Global Parameter Setting ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Global-Parameter-Setting"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Workflow", "text": "Workflow ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Workflow"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Preliminary quality control and preprocessing for genotype data", "text": "Preliminary quality control and preprocessing for genotype data ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Preliminary-quality-control-and-preprocessing-for-genotype-data"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Find common SNPs among summary statistics, reference panel and test genotypes.", "text": "Find common SNPs among summary statistics, reference panel and test genotypes. ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Find-common-SNPs-among-summary-statistics,-reference-panel-and-test-genotypes."},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Convert reference genotype to bigsnpr format and get the genetic distance cM information", "text": "Convert reference genotype to bigsnpr format and get the genetic distance cM information ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Convert-reference-genotype-to-bigsnpr-format-and-get-the-genetic-distance-cM-information"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Allele harmonizing", "text": "Allele harmonizing ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Allele-harmonizing"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Summary statistics quality Control", "text": "Summary statistics quality Control ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Summary-statistics-quality-Control"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Calculate LD matrix and perform LD score regression", "text": "Calculate LD matrix and perform LD score regression ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Calculate-LD-matrix-and-perform-LD-score-regression"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Get adjusted betas and PRS", "text": "Get adjusted betas and PRS ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Get-adjusted-betas-and-PRS"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Infinitesimal model", "text": "Infinitesimal model ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Infinitesimal-model"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Grid model", "text": "Grid model ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Grid-model"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Auto model", "text": "Auto model ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Auto-model"},
{"mainTitle": "LDpred2 Pipeline for Polygenic Risk Score Prediction", "title": "Predict phenotype", "text": "Predict phenotype ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/index.html", "url": "http://cumc.github.io/bioworkflows/ldpred/index.html#Predict-phenotype"},
{"mainTitle": "Annotation of exome variants using Annovar", "title": "Annotation of exome variants using Annovar", "text": "Annotation of exome variants using Annovar ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-annotation/annovar.html", "url": "http://cumc.github.io/bioworkflows/variant-annotation/annovar.html#Annotation-of-exome-variants-using-Annovar"},
{"mainTitle": "Annotation of exome variants using Annovar", "title": "Aim", "text": "Aim ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-annotation/annovar.html", "url": "http://cumc.github.io/bioworkflows/variant-annotation/annovar.html#Aim"},
{"mainTitle": "Annotation of exome variants using Annovar", "title": "Description of the pipeline", "text": "Description of the pipeline ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-annotation/annovar.html", "url": "http://cumc.github.io/bioworkflows/variant-annotation/annovar.html#Description-of-the-pipeline"},
{"mainTitle": "Annotation of exome variants using Annovar", "title": "Scenario 1 : you have multiple bim files (e.g. one per chromosome) and you want to merge them into one file for later annotation with annovar", "text": "Scenario 1 you have multiple bim files e.g. one per chromosome and you want to merge them into one file for later annotation with annovar ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-annotation/annovar.html", "url": "http://cumc.github.io/bioworkflows/variant-annotation/annovar.html#Scenario-1-:-you-have-multiple-bim-files-(e.g.-one-per-chromosome)-and-you-want-to-merge-them-into-one-file-for-later-annotation-with-annovar"},
{"mainTitle": "Annotation of exome variants using Annovar", "title": "Scenario 2: you either want to work with common or rare variants.", "text": "Scenario 2 you either want to work with common or rare variants. ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-annotation/annovar.html", "url": "http://cumc.github.io/bioworkflows/variant-annotation/annovar.html#Scenario-2:-you-either-want-to-work-with-common-or-rare-variants."},
{"mainTitle": "Annotation of exome variants using Annovar", "title": "Scenario 3: you already have a specific list of variants you would like to annotate stored in a bim file.", "text": "Scenario 3 you already have a specific list of variants you would like to annotate stored in a bim file. ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-annotation/annovar.html", "url": "http://cumc.github.io/bioworkflows/variant-annotation/annovar.html#Scenario-3:-you-already-have-a-specific-list-of-variants-you-would-like-to-annotate-stored-in-a-bim-file."},
{"mainTitle": "Annotation of exome variants using Annovar", "title": "Command interface", "text": "Command interface ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-annotation/annovar.html", "url": "http://cumc.github.io/bioworkflows/variant-annotation/annovar.html#Command-interface"},
{"mainTitle": "Annotation of exome variants using Annovar", "title": "Illustration with minimal working example", "text": "Illustration with minimal working example ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-annotation/annovar.html", "url": "http://cumc.github.io/bioworkflows/variant-annotation/annovar.html#Illustration-with-minimal-working-example"},
{"mainTitle": "Annotation of exome variants using Annovar", "title": "Format file for plink .bim", "text": "Format file for plink .bim ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-annotation/annovar.html", "url": "http://cumc.github.io/bioworkflows/variant-annotation/annovar.html#Format-file-for-plink-.bim"},
{"mainTitle": "Annotation of exome variants using Annovar", "title": "Step to merge *.bim files from plink formatted data (e.g exome data in the UKBB, genotype array data)", "text": "Step to merge .bim files from plink formatted data e.g exome data in the UKBB, genotype array data ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-annotation/annovar.html", "url": "http://cumc.github.io/bioworkflows/variant-annotation/annovar.html#Step-to-merge-*.bim-files-from-plink-formatted-data-(e.g-exome-data-in-the-UKBB,-genotype-array-data)"},
{"mainTitle": "Annotation of exome variants using Annovar", "title": "Step to create the bim files from mfi (bgen) files to annotate (e.g imputed genotype data UKBB)", "text": "Step to create the bim files from mfi bgen files to annotate e.g imputed genotype data UKBB ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-annotation/annovar.html", "url": "http://cumc.github.io/bioworkflows/variant-annotation/annovar.html#Step-to-create-the-bim-files-from-mfi-(bgen)-files-to-annotate-(e.g-imputed-genotype-data-UKBB)"},
{"mainTitle": "Annotation of exome variants using Annovar", "title": "Annovar details", "text": "Annovar details ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-annotation/annovar.html", "url": "http://cumc.github.io/bioworkflows/variant-annotation/annovar.html#Annovar-details"},
{"mainTitle": "Annotation of exome variants using Annovar", "title": "Important note", "text": "Important note ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-annotation/annovar.html", "url": "http://cumc.github.io/bioworkflows/variant-annotation/annovar.html#Important-note"},
{"mainTitle": "Annotation of exome variants using Annovar", "title": "Format file for annovar input", "text": "Format file for annovar input ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-annotation/annovar.html", "url": "http://cumc.github.io/bioworkflows/variant-annotation/annovar.html#Format-file-for-annovar-input"},
{"mainTitle": "Annotation of exome variants using Annovar", "title": "Generate files for burden_test regenie from the annotated file", "text": "Generate files for burden_test regenie from the annotated file ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-annotation/annovar.html", "url": "http://cumc.github.io/bioworkflows/variant-annotation/annovar.html#Generate-files-for-burden_test-regenie-from-the-annotated-file"},
{"mainTitle": "Polygenic Risk Score Analysis on Height GWAS data", "title": "Polygenic Risk Score Analysis on Height GWAS data", "text": "Polygenic Risk Score Analysis on Height GWAS data ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html", "url": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html#Polygenic-Risk-Score-Analysis-on-Height-GWAS-data"},
{"mainTitle": "Polygenic Risk Score Analysis on Height GWAS data", "title": "Data used", "text": "Data used ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html", "url": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html#Data-used"},
{"mainTitle": "Polygenic Risk Score Analysis on Height GWAS data", "title": "Reference panel", "text": "Reference panel ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html", "url": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html#Reference-panel"},
{"mainTitle": "Polygenic Risk Score Analysis on Height GWAS data", "title": "GWAS summary statistics data", "text": "GWAS summary statistics data ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html", "url": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html#GWAS-summary-statistics-data"},
{"mainTitle": "Polygenic Risk Score Analysis on Height GWAS data", "title": "Target test data", "text": "Target test data ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html", "url": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html#Target-test-data"},
{"mainTitle": "Polygenic Risk Score Analysis on Height GWAS data", "title": "PRS Models applied", "text": "PRS Models applied ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html", "url": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html#PRS-Models-applied"},
{"mainTitle": "Polygenic Risk Score Analysis on Height GWAS data", "title": "Data preparation", "text": "Data preparation ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html", "url": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html#Data-preparation"},
{"mainTitle": "Polygenic Risk Score Analysis on Height GWAS data", "title": "Summary statistics formatting", "text": "Summary statistics formatting ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html", "url": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html#Summary-statistics-formatting"},
{"mainTitle": "Polygenic Risk Score Analysis on Height GWAS data", "title": "Phenotype and covariates formatting", "text": "Phenotype and covariates formatting ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html", "url": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html#Phenotype-and-covariates-formatting"},
{"mainTitle": "Polygenic Risk Score Analysis on Height GWAS data", "title": "Analysis of height GWAS data", "text": "Analysis of height GWAS data ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html", "url": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html#Analysis-of-height-GWAS-data"},
{"mainTitle": "Polygenic Risk Score Analysis on Height GWAS data", "title": "Step 1: QC on reference panel", "text": "Step 1 QC on reference panel ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html", "url": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html#Step-1:-QC-on-reference-panel"},
{"mainTitle": "Polygenic Risk Score Analysis on Height GWAS data", "title": "Step 2: Intersect SNPs among summary stats, reference panel and target data", "text": "Step 2 Intersect SNPs among summary stats, reference panel and target data ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html", "url": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html#Step-2:-Intersect-SNPs-among-summary-stats,-reference-panel-and-target-data"},
{"mainTitle": "Polygenic Risk Score Analysis on Height GWAS data", "title": "Step 3: Harmonize alleles for shared SNPs", "text": "Step 3 Harmonize alleles for shared SNPs ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html", "url": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html#Step-3:-Harmonize-alleles-for-shared-SNPs"},
{"mainTitle": "Polygenic Risk Score Analysis on Height GWAS data", "title": "Step 4: Summary statistics quality control (optional)", "text": "Step 4 Summary statistics quality control optional ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html", "url": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html#Step-4:-Summary-statistics-quality-control-(optional)"},
{"mainTitle": "Polygenic Risk Score Analysis on Height GWAS data", "title": "Step 5: Calculate LD matrix and fit LDSC model", "text": "Step 5 Calculate LD matrix and fit LDSC model ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html", "url": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html#Step-5:-Calculate-LD-matrix-and-fit-LDSC-model"},
{"mainTitle": "Polygenic Risk Score Analysis on Height GWAS data", "title": "Step 6: Estimate posterior effect sizes and PRS", "text": "Step 6 Estimate posterior effect sizes and PRS ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html", "url": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html#Step-6:-Estimate-posterior-effect-sizes-and-PRS"},
{"mainTitle": "Polygenic Risk Score Analysis on Height GWAS data", "title": "Step 7: predict phenotypes", "text": "Step 7 predict phenotypes ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html", "url": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html#Step-7:-predict-phenotypes"},
{"mainTitle": "Polygenic Risk Score Analysis on Height GWAS data", "title": "Results summary", "text": "Results summary ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html", "url": "http://cumc.github.io/bioworkflows/ldpred/ldpred2_example.html#Results-summary"},
{"mainTitle": "SuSiE RSS", "title": "SuSiE RSS", "text": "SuSiE RSS ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/SuSiE_RSS.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/SuSiE_RSS.html#SuSiE-RSS"},
{"mainTitle": "Extracting data for genomic regions of interest", "title": "Extracting data for genomic regions of interest", "text": "Extracting data for genomic regions of interest ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/Region_Extraction.html", "url": "http://cumc.github.io/bioworkflows/GWAS/Region_Extraction.html#Extracting-data-for-genomic-regions-of-interest"},
{"mainTitle": "Extracting data for genomic regions of interest", "title": "Aim", "text": "Aim ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/Region_Extraction.html", "url": "http://cumc.github.io/bioworkflows/GWAS/Region_Extraction.html#Aim"},
{"mainTitle": "Extracting data for genomic regions of interest", "title": "Pre-requisites", "text": "Pre-requisites ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/Region_Extraction.html", "url": "http://cumc.github.io/bioworkflows/GWAS/Region_Extraction.html#Pre-requisites"},
{"mainTitle": "Extracting data for genomic regions of interest", "title": "Input and Output", "text": "Input and Output ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/Region_Extraction.html", "url": "http://cumc.github.io/bioworkflows/GWAS/Region_Extraction.html#Input-and-Output"},
{"mainTitle": "Extracting data for genomic regions of interest", "title": "Input", "text": "Input ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/Region_Extraction.html", "url": "http://cumc.github.io/bioworkflows/GWAS/Region_Extraction.html#Input"},
{"mainTitle": "Extracting data for genomic regions of interest", "title": "Output", "text": "Output ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/Region_Extraction.html", "url": "http://cumc.github.io/bioworkflows/GWAS/Region_Extraction.html#Output"},
{"mainTitle": "Extracting data for genomic regions of interest", "title": "Workflow usage", "text": "Workflow usage ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/Region_Extraction.html", "url": "http://cumc.github.io/bioworkflows/GWAS/Region_Extraction.html#Workflow-usage"},
{"mainTitle": "Extracting data for genomic regions of interest", "title": "For bgen format input", "text": "For bgen format input ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/Region_Extraction.html", "url": "http://cumc.github.io/bioworkflows/GWAS/Region_Extraction.html#For-bgen-format-input"},
{"mainTitle": "Extracting data for genomic regions of interest", "title": "For PLINK format input", "text": "For PLINK format input ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/Region_Extraction.html", "url": "http://cumc.github.io/bioworkflows/GWAS/Region_Extraction.html#For-PLINK-format-input"},
{"mainTitle": "Extracting data for genomic regions of interest", "title": "Workflow codes", "text": "Workflow codes ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/Region_Extraction.html", "url": "http://cumc.github.io/bioworkflows/GWAS/Region_Extraction.html#Workflow-codes"},
{"mainTitle": "Extracting data for genomic regions of interest", "title": "Some utility functions", "text": "Some utility functions ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/Region_Extraction.html", "url": "http://cumc.github.io/bioworkflows/GWAS/Region_Extraction.html#Some-utility-functions"},
{"mainTitle": "Extracting data for genomic regions of interest", "title": "Extract data", "text": "Extract data ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/Region_Extraction.html", "url": "http://cumc.github.io/bioworkflows/GWAS/Region_Extraction.html#Extract-data"},
{"mainTitle": "WGS GVCF samples joint calling, filtering and annotation", "title": "WGS GVCF samples joint calling, filtering and annotation", "text": "WGS GVCF samples joint calling, filtering and annotation ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-calling/gatk_joint_calling.html", "url": "http://cumc.github.io/bioworkflows/variant-calling/gatk_joint_calling.html#WGS-GVCF-samples-joint-calling,-filtering-and-annotation"},
{"mainTitle": "WGS GVCF samples joint calling, filtering and annotation", "title": "Overview", "text": "Overview ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-calling/gatk_joint_calling.html", "url": "http://cumc.github.io/bioworkflows/variant-calling/gatk_joint_calling.html#Overview"},
{"mainTitle": "WGS GVCF samples joint calling, filtering and annotation", "title": "Input data", "text": "Input data ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-calling/gatk_joint_calling.html", "url": "http://cumc.github.io/bioworkflows/variant-calling/gatk_joint_calling.html#Input-data"},
{"mainTitle": "WGS GVCF samples joint calling, filtering and annotation", "title": "Reference data preparation", "text": "Reference data preparation ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-calling/gatk_joint_calling.html", "url": "http://cumc.github.io/bioworkflows/variant-calling/gatk_joint_calling.html#Reference-data-preparation"},
{"mainTitle": "WGS GVCF samples joint calling, filtering and annotation", "title": "Run the workflow", "text": "Run the workflow ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-calling/gatk_joint_calling.html", "url": "http://cumc.github.io/bioworkflows/variant-calling/gatk_joint_calling.html#Run-the-workflow"},
{"mainTitle": "WGS GVCF samples joint calling, filtering and annotation", "title": "Minimal working example", "text": "Minimal working example ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-calling/gatk_joint_calling.html", "url": "http://cumc.github.io/bioworkflows/variant-calling/gatk_joint_calling.html#Minimal-working-example"},
{"mainTitle": "WGS GVCF samples joint calling, filtering and annotation", "title": "Share the workflow with people", "text": "Share the workflow with people ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-calling/gatk_joint_calling.html", "url": "http://cumc.github.io/bioworkflows/variant-calling/gatk_joint_calling.html#Share-the-workflow-with-people"},
{"mainTitle": "WGS GVCF samples joint calling, filtering and annotation", "title": "Software configuration", "text": "Software configuration ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-calling/gatk_joint_calling.html", "url": "http://cumc.github.io/bioworkflows/variant-calling/gatk_joint_calling.html#Software-configuration"},
{"mainTitle": "WGS GVCF samples joint calling, filtering and annotation", "title": "Global parameter settings", "text": "Global parameter settings ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-calling/gatk_joint_calling.html", "url": "http://cumc.github.io/bioworkflows/variant-calling/gatk_joint_calling.html#Global-parameter-settings"},
{"mainTitle": "WGS GVCF samples joint calling, filtering and annotation", "title": "Joint variant calling from GVCF files", "text": "Joint variant calling from GVCF files ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-calling/gatk_joint_calling.html", "url": "http://cumc.github.io/bioworkflows/variant-calling/gatk_joint_calling.html#Joint-variant-calling-from-GVCF-files"},
{"mainTitle": "WGS GVCF samples joint calling, filtering and annotation", "title": "Variant filtering", "text": "Variant filtering ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-calling/gatk_joint_calling.html", "url": "http://cumc.github.io/bioworkflows/variant-calling/gatk_joint_calling.html#Variant-filtering"},
{"mainTitle": "WGS GVCF samples joint calling, filtering and annotation", "title": "Annotation", "text": "Annotation ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-calling/gatk_joint_calling.html", "url": "http://cumc.github.io/bioworkflows/variant-calling/gatk_joint_calling.html#Annotation"},
{"mainTitle": "WGS GVCF samples joint calling, filtering and annotation", "title": "Submit jobs to the cluster", "text": "Submit jobs to the cluster ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-calling/gatk_joint_calling.html", "url": "http://cumc.github.io/bioworkflows/variant-calling/gatk_joint_calling.html#Submit-jobs-to-the-cluster"},
{"mainTitle": "Summary statistics data mungling in preparation for fine-mapping pipelines", "title": "Summary statistics data mungling in preparation for fine-mapping pipelines", "text": "Summary statistics data mungling in preparation for fine-mapping pipelines ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_wrangler.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_wrangler.html#Summary-statistics-data-mungling-in-preparation-for-fine-mapping-pipelines"},
{"mainTitle": "Summary statistics data mungling in preparation for fine-mapping pipelines", "title": "Input", "text": "Input ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_wrangler.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_wrangler.html#Input"},
{"mainTitle": "Summary statistics data mungling in preparation for fine-mapping pipelines", "title": "Pitfalls in data mungling with external reference panel", "text": "Pitfalls in data mungling with external reference panel ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_wrangler.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_wrangler.html#Pitfalls-in-data-mungling-with-external-reference-panel"},
{"mainTitle": "Summary statistics data mungling in preparation for fine-mapping pipelines", "title": "Data extraction steps", "text": "Data extraction steps ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_wrangler.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_wrangler.html#Data-extraction-steps"},
{"mainTitle": "Summary statistics data mungling in preparation for fine-mapping pipelines", "title": "Outputs", "text": "Outputs ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_wrangler.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_wrangler.html#Outputs"},
{"mainTitle": "Summary statistics data mungling in preparation for fine-mapping pipelines", "title": "Software requirement", "text": "Software requirement ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_wrangler.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_wrangler.html#Software-requirement"},
{"mainTitle": "Summary statistics data mungling in preparation for fine-mapping pipelines", "title": "Preparing external reference genotypes", "text": "Preparing external reference genotypes ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_wrangler.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_wrangler.html#Preparing-external-reference-genotypes"},
{"mainTitle": "Summary statistics data mungling in preparation for fine-mapping pipelines", "title": "Obtain summary statistics data-set", "text": "Obtain summary statistics data-set ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_wrangler.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_wrangler.html#Obtain-summary-statistics-data-set"},
{"mainTitle": "Summary statistics data mungling in preparation for fine-mapping pipelines", "title": "Obtain annotation $A_2$", "text": "Obtain annotation A_2 ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_wrangler.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_wrangler.html#Obtain-annotation-$A_2$"},
{"mainTitle": "Summary statistics data mungling in preparation for fine-mapping pipelines", "title": "Run preprocessing pipeline", "text": "Run preprocessing pipeline ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_wrangler.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_wrangler.html#Run-preprocessing-pipeline"},
{"mainTitle": "Summary statistics data mungling in preparation for fine-mapping pipelines", "title": "Results preview", "text": "Results preview ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_wrangler.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_wrangler.html#Results-preview"},
{"mainTitle": "Summary statistics data mungling in preparation for fine-mapping pipelines", "title": "Reference panel matching summary", "text": "Reference panel matching summary ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_wrangler.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_wrangler.html#Reference-panel-matching-summary"},
{"mainTitle": "Summary statistics data mungling in preparation for fine-mapping pipelines", "title": "Summary statistics matrix S3", "text": "Summary statistics matrix S3 ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_wrangler.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_wrangler.html#Summary-statistics-matrix-S3"},
{"mainTitle": "Summary statistics data mungling in preparation for fine-mapping pipelines", "title": "Annotation/prior matrix A2", "text": "Annotation prior matrix A2 ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_wrangler.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_wrangler.html#Annotation/prior-matrix-A2"},
{"mainTitle": "Summary statistics data mungling in preparation for fine-mapping pipelines", "title": ": LD matrix", "text": "LD matrix ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_wrangler.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_wrangler.html#:-LD-matrix"},
{"mainTitle": "Fine-mapping workflow using summary statistics", "title": "Fine-mapping workflow using summary statistics", "text": "Fine-mapping workflow using summary statistics ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html#Fine-mapping-workflow-using-summary-statistics"},
{"mainTitle": "Fine-mapping workflow using summary statistics", "title": "Software preparation", "text": "Software preparation ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html#Software-preparation"},
{"mainTitle": "Fine-mapping workflow using summary statistics", "title": "Input data", "text": "Input data ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html#Input-data"},
{"mainTitle": "Fine-mapping workflow using summary statistics", "title": "Output format", "text": "Output format ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html#Output-format"},
{"mainTitle": "Fine-mapping workflow using summary statistics", "title": "The workflow", "text": "The workflow ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html#The-workflow"},
{"mainTitle": "Fine-mapping workflow using summary statistics", "title": "SuSiE $\hat{b}$ version", "text": "SuSiE \hat b version ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html#SuSiE-$\hat{b}$-version"},
{"mainTitle": "Fine-mapping workflow using summary statistics", "title": "SuSiE $z$ version", "text": "SuSiE z version ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html#SuSiE-$z$-version"},
{"mainTitle": "Fine-mapping workflow using summary statistics", "title": "DAP-G", "text": "DAP-G ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html#DAP-G"},
{"mainTitle": "Fine-mapping workflow using summary statistics", "title": "CAVIAR", "text": "CAVIAR ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html#CAVIAR"},
{"mainTitle": "Fine-mapping workflow using summary statistics", "title": "FINEMAP", "text": "FINEMAP ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html#FINEMAP"},
{"mainTitle": "Fine-mapping workflow using summary statistics", "title": "Make diagnostic plot", "text": "Make diagnostic plot ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html#Make-diagnostic-plot"},
{"mainTitle": "Fine-mapping workflow using summary statistics", "title": "Example usage", "text": "Example usage ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html#Example-usage"},
{"mainTitle": "Fine-mapping workflow using summary statistics", "title": "SuSiE $\hat{b}$", "text": "SuSiE \hat b ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html#SuSiE-$\hat{b}$"},
{"mainTitle": "Fine-mapping workflow using summary statistics", "title": "SuSiE $z$", "text": "SuSiE z ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html#SuSiE-$z$"},
{"mainTitle": "Fine-mapping workflow using summary statistics", "title": "DAP", "text": "DAP ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html#DAP"},
{"mainTitle": "Fine-mapping workflow using summary statistics", "title": "CAVIAR", "text": "CAVIAR ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html#CAVIAR"},
{"mainTitle": "Fine-mapping workflow using summary statistics", "title": "FINEMAP", "text": "FINEMAP ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/summary_statistics_finemapping.html#FINEMAP"},
{"mainTitle": "Identifying loci of interest from GWAS", "title": "Identifying loci of interest from GWAS", "text": "Identifying loci of interest from GWAS ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LD_Clumping.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LD_Clumping.html#Identifying-loci-of-interest-from-GWAS"},
{"mainTitle": "Identifying loci of interest from GWAS", "title": "Input", "text": "Input ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LD_Clumping.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LD_Clumping.html#Input"},
{"mainTitle": "Identifying loci of interest from GWAS", "title": "Output", "text": "Output ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LD_Clumping.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LD_Clumping.html#Output"},
{"mainTitle": "Identifying loci of interest from GWAS", "title": "Command interface", "text": "Command interface ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LD_Clumping.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LD_Clumping.html#Command-interface"},
{"mainTitle": "Identifying loci of interest from GWAS", "title": "Global parameter settings", "text": "Global parameter settings ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LD_Clumping.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LD_Clumping.html#Global-parameter-settings"},
{"mainTitle": "Identifying loci of interest from GWAS", "title": "Illustration with a minimal working example", "text": "Illustration with a minimal working example ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LD_Clumping.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LD_Clumping.html#Illustration-with-a-minimal-working-example"},
{"mainTitle": "Identifying loci of interest from GWAS", "title": "Note", "text": "Note ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LD_Clumping.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LD_Clumping.html#Note"},
{"mainTitle": "Identifying loci of interest from GWAS", "title": "Select a random subset of unrelated samples", "text": "Select a random subset of unrelated samples ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LD_Clumping.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LD_Clumping.html#Select-a-random-subset-of-unrelated-samples"},
{"mainTitle": "Identifying loci of interest from GWAS", "title": "Select Filter the BGEN files", "text": "Select Filter the BGEN files ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LD_Clumping.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LD_Clumping.html#Select-Filter-the-BGEN-files"},
{"mainTitle": "Identifying loci of interest from GWAS", "title": "Make PLINK files with the selected samples", "text": "Make PLINK files with the selected samples ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LD_Clumping.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LD_Clumping.html#Make-PLINK-files-with-the-selected-samples"},
{"mainTitle": "Identifying loci of interest from GWAS", "title": "Filter PLINK files", "text": "Filter PLINK files ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LD_Clumping.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LD_Clumping.html#Filter-PLINK-files"},
{"mainTitle": "Identifying loci of interest from GWAS", "title": "Merge all chroms to one file", "text": "Merge all chroms to one file ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LD_Clumping.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LD_Clumping.html#Merge-all-chroms-to-one-file"},
{"mainTitle": "Identifying loci of interest from GWAS", "title": "Perform LD clumping per chrom", "text": "Perform LD clumping per chrom ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LD_Clumping.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LD_Clumping.html#Perform-LD-clumping-per-chrom"},
{"mainTitle": "SuSiE RSS", "title": "SuSiE RSS", "text": "SuSiE RSS ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/SuSiE_RSS_before_initialization.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/SuSiE_RSS_before_initialization.html#SuSiE-RSS"},
{"mainTitle": "Enrichment analysis workflows", "title": "Enrichment analysis workflows", "text": "Enrichment analysis workflows ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html#Enrichment-analysis-workflows"},
{"mainTitle": "Enrichment analysis workflows", "title": "Copy this GIT depository", "text": "Copy this GIT depository ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html#Copy-this-GIT-depository"},
{"mainTitle": "Enrichment analysis workflows", "title": "To run enrichment analysis (overview)", "text": "To run enrichment analysis overview ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html#To-run-enrichment-analysis-(overview)"},
{"mainTitle": "Enrichment analysis workflows", "title": "Step 1 (start from bed format annotation)", "text": "Step 1 start from bed format annotation ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html#Step-1-(start-from-bed-format-annotation)"},
{"mainTitle": "Enrichment analysis workflows", "title": "Step 2 (start from binary annotation for SNPs)", "text": "Step 2 start from binary annotation for SNPs ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html#Step-2-(start-from-binary-annotation-for-SNPs)"},
{"mainTitle": "Enrichment analysis workflows", "title": "Software intallation", "text": "Software intallation ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html#Software-intallation"},
{"mainTitle": "Enrichment analysis workflows", "title": "SoS", "text": "SoS ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html#SoS"},
{"mainTitle": "Enrichment analysis workflows", "title": "docker", "text": "docker ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html#docker"},
{"mainTitle": "Enrichment analysis workflows", "title": "Reference data preparation", "text": "Reference data preparation ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html#Reference-data-preparation"},
{"mainTitle": "Enrichment analysis workflows", "title": "hg19.fa", "text": "hg19.fa ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html#hg19.fa"},
{"mainTitle": "Enrichment analysis workflows", "title": "Annotation files", "text": "Annotation files ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html#Annotation-files"},
{"mainTitle": "Enrichment analysis workflows", "title": "Annotation files download", "text": "Annotation files download ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html#Annotation-files-download"},
{"mainTitle": "Enrichment analysis workflows", "title": "bed format annotation input", "text": "bed format annotation input ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html#bed-format-annotation-input"},
{"mainTitle": "Enrichment analysis workflows", "title": "Binary annotations of interest", "text": "Binary annotations of interest ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html#Binary-annotations-of-interest"},
{"mainTitle": "Enrichment analysis workflows", "title": "torus workflow data format", "text": "torus workflow data format ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html#torus-workflow-data-format"},
{"mainTitle": "Enrichment analysis workflows", "title": "GWAS z-scores specification", "text": "GWAS z-scores specification ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html#GWAS-z-scores-specification"},
{"mainTitle": "Enrichment analysis workflows", "title": "LD block files", "text": "LD block files ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html#LD-block-files"},
{"mainTitle": "Enrichment analysis workflows", "title": "SNP map", "text": "SNP map ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html#SNP-map"},
{"mainTitle": "Enrichment analysis workflows", "title": "Minimal working example for torus workflow", "text": "Minimal working example for torus workflow ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html#Minimal-working-example-for-torus-workflow"},
{"mainTitle": "Enrichment analysis workflows", "title": "Workflow", "text": "Workflow ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html#Workflow"},
{"mainTitle": "Enrichment analysis workflows", "title": "Utility steps", "text": "Utility steps ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html#Utility-steps"},
{"mainTitle": "Enrichment analysis workflows", "title": "Convert variants from z-score file to bed format", "text": "Convert variants from z-score file to bed format ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html#Convert-variants-from-z-score-file-to-bed-format"},
{"mainTitle": "Enrichment analysis workflows", "title": "Merge annotations for joint torus analysis", "text": "Merge annotations for joint torus analysis ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html#Merge-annotations-for-joint-torus-analysis"},
{"mainTitle": "Enrichment analysis workflows", "title": "Prepare torus format binary annotations", "text": "Prepare torus format binary annotations ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html#Prepare-torus-format-binary-annotations"},
{"mainTitle": "Enrichment analysis workflows", "title": "Enrichment analysis via torus", "text": "Enrichment analysis via torus ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html#Enrichment-analysis-via-torus"},
{"mainTitle": "Enrichment analysis workflows", "title": "Apply pre-trained deepsea model to variants", "text": "Apply pre-trained deepsea model to variants ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html#Apply-pre-trained-deepsea-model-to-variants"},
{"mainTitle": "Enrichment analysis workflows", "title": "Enrichment analysis via GREGOR", "text": "Enrichment analysis via GREGOR ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html#Enrichment-analysis-via-GREGOR"},
{"mainTitle": "Enrichment analysis workflows", "title": "To run GREGOR", "text": "To run GREGOR ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/gwas_enrichment.html#To-run-GREGOR"},
{"mainTitle": "GWAS data QC workflow", "title": "GWAS data QC workflow", "text": "GWAS data QC workflow ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/GWAS_QC.html", "url": "http://cumc.github.io/bioworkflows/GWAS/GWAS_QC.html#GWAS-data-QC-workflow"},
{"mainTitle": "GWAS data QC workflow", "title": "Overview", "text": "Overview ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/GWAS_QC.html", "url": "http://cumc.github.io/bioworkflows/GWAS/GWAS_QC.html#Overview"},
{"mainTitle": "GWAS data QC workflow", "title": "Run this workflow", "text": "Run this workflow ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/GWAS_QC.html", "url": "http://cumc.github.io/bioworkflows/GWAS/GWAS_QC.html#Run-this-workflow"},
{"mainTitle": "GWAS data QC workflow", "title": "Minimal working example", "text": "Minimal working example ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/GWAS_QC.html", "url": "http://cumc.github.io/bioworkflows/GWAS/GWAS_QC.html#Minimal-working-example"},
{"mainTitle": "GWAS data QC workflow", "title": "First scenario: estimate kinship", "text": "First scenario estimate kinship ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/GWAS_QC.html", "url": "http://cumc.github.io/bioworkflows/GWAS/GWAS_QC.html#First-scenario:-estimate-kinship"},
{"mainTitle": "GWAS data QC workflow", "title": "First scenario: do qc", "text": "First scenario do qc ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/GWAS_QC.html", "url": "http://cumc.github.io/bioworkflows/GWAS/GWAS_QC.html#First-scenario:-do-qc"},
{"mainTitle": "GWAS data QC workflow", "title": "Estimate kinship in the sample", "text": "Estimate kinship in the sample ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/GWAS_QC.html", "url": "http://cumc.github.io/bioworkflows/GWAS/GWAS_QC.html#Estimate-kinship-in-the-sample"},
{"mainTitle": "GWAS data QC workflow", "title": "Genotype and sample QC", "text": "Genotype and sample QC ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/GWAS_QC.html", "url": "http://cumc.github.io/bioworkflows/GWAS/GWAS_QC.html#Genotype-and-sample-QC"},
{"mainTitle": "WGS Mapping to Variant Calls - Version 1.0", "title": "WGS Mapping to Variant Calls - Version 1.0", "text": "WGS Mapping to Variant Calls - Version 1.0 ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-calling/samtools_variant_calling.html", "url": "http://cumc.github.io/bioworkflows/variant-calling/samtools_variant_calling.html#WGS-Mapping-to-Variant-Calls---Version-1.0"},
{"mainTitle": "WGS Mapping to Variant Calls - Version 1.0", "title": "Overview", "text": "Overview ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-calling/samtools_variant_calling.html", "url": "http://cumc.github.io/bioworkflows/variant-calling/samtools_variant_calling.html#Overview"},
{"mainTitle": "WGS Mapping to Variant Calls - Version 1.0", "title": "Input data", "text": "Input data ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-calling/samtools_variant_calling.html", "url": "http://cumc.github.io/bioworkflows/variant-calling/samtools_variant_calling.html#Input-data"},
{"mainTitle": "WGS Mapping to Variant Calls - Version 1.0", "title": "Reference data preparation", "text": "Reference data preparation ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-calling/samtools_variant_calling.html", "url": "http://cumc.github.io/bioworkflows/variant-calling/samtools_variant_calling.html#Reference-data-preparation"},
{"mainTitle": "WGS Mapping to Variant Calls - Version 1.0", "title": "Run the workflow", "text": "Run the workflow ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-calling/samtools_variant_calling.html", "url": "http://cumc.github.io/bioworkflows/variant-calling/samtools_variant_calling.html#Run-the-workflow"},
{"mainTitle": "WGS Mapping to Variant Calls - Version 1.0", "title": "Do we keep all loci even if they are not variants?", "text": "Do we keep all loci even if they are not variants ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-calling/samtools_variant_calling.html", "url": "http://cumc.github.io/bioworkflows/variant-calling/samtools_variant_calling.html#Do-we-keep-all-loci-even-if-they-are-not-variants?"},
{"mainTitle": "WGS Mapping to Variant Calls - Version 1.0", "title": "Software configuration", "text": "Software configuration ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-calling/samtools_variant_calling.html", "url": "http://cumc.github.io/bioworkflows/variant-calling/samtools_variant_calling.html#Software-configuration"},
{"mainTitle": "WGS Mapping to Variant Calls - Version 1.0", "title": "Global parameter settings", "text": "Global parameter settings ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-calling/samtools_variant_calling.html", "url": "http://cumc.github.io/bioworkflows/variant-calling/samtools_variant_calling.html#Global-parameter-settings"},
{"mainTitle": "WGS Mapping to Variant Calls - Version 1.0", "title": "Reference preparation", "text": "Reference preparation ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-calling/samtools_variant_calling.html", "url": "http://cumc.github.io/bioworkflows/variant-calling/samtools_variant_calling.html#Reference-preparation"},
{"mainTitle": "WGS Mapping to Variant Calls - Version 1.0", "title": "Reference genome", "text": "Reference genome ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-calling/samtools_variant_calling.html", "url": "http://cumc.github.io/bioworkflows/variant-calling/samtools_variant_calling.html#Reference-genome"},
{"mainTitle": "WGS Mapping to Variant Calls - Version 1.0", "title": "Known sites", "text": "Known sites ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-calling/samtools_variant_calling.html", "url": "http://cumc.github.io/bioworkflows/variant-calling/samtools_variant_calling.html#Known-sites"},
{"mainTitle": "WGS Mapping to Variant Calls - Version 1.0", "title": "Mapping", "text": "Mapping ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-calling/samtools_variant_calling.html", "url": "http://cumc.github.io/bioworkflows/variant-calling/samtools_variant_calling.html#Mapping"},
{"mainTitle": "WGS Mapping to Variant Calls - Version 1.0", "title": "Improvement", "text": "Improvement ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-calling/samtools_variant_calling.html", "url": "http://cumc.github.io/bioworkflows/variant-calling/samtools_variant_calling.html#Improvement"},
{"mainTitle": "WGS Mapping to Variant Calls - Version 1.0", "title": "Variant Calling", "text": "Variant Calling ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-calling/samtools_variant_calling.html", "url": "http://cumc.github.io/bioworkflows/variant-calling/samtools_variant_calling.html#Variant-Calling"},
{"mainTitle": "WGS Mapping to Variant Calls - Version 1.0", "title": "References", "text": "References ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/variant-calling/samtools_variant_calling.html", "url": "http://cumc.github.io/bioworkflows/variant-calling/samtools_variant_calling.html#References"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "LMM/GLMM analyses for UK Biobank data", "text": "LMM GLMM analyses for UK Biobank data ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#LMM/GLMM-analyses-for-UK-Biobank-data"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "Aim", "text": "Aim ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#Aim"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "Input data", "text": "Input data ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#Input-data"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "Software specific inputs", "text": "Software specific inputs ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#Software-specific-inputs"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "BoltLMM additional input", "text": "BoltLMM additional input ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#BoltLMM-additional-input"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "Output", "text": "Output ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#Output"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "References", "text": "References ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#References"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "Command interface", "text": "Command interface ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#Command-interface"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "Global parameter setting", "text": "Global parameter setting ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#Global-parameter-setting"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "Illustration with minimal working examples", "text": "Illustration with minimal working examples ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#Illustration-with-minimal-working-examples"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "BOLT-LMM example command", "text": "BOLT-LMM example command ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#BOLT-LMM-example-command"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "fastGWA example command", "text": "fastGWA example command ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#fastGWA-example-command"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "REGENIE example command", "text": "REGENIE example command ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#REGENIE-example-command"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "REGENIE burden example command", "text": "REGENIE burden example command ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#REGENIE-burden-example-command"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "SAIGE example command", "text": "SAIGE example command ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#SAIGE-example-command"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "Input using bed format", "text": "Input using bed format ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#Input-using-bed-format"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "GMMAT example command", "text": "GMMAT example command ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#GMMAT-example-command"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "SMMAT example command", "text": "SMMAT example command ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#SMMAT-example-command"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "Run workflow on a cluster", "text": "Run workflow on a cluster ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#Run-workflow-on-a-cluster"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "BoltLMM workflow implementation", "text": "BoltLMM workflow implementation ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#BoltLMM-workflow-implementation"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "fastGWA workflow implementation", "text": "fastGWA workflow implementation ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#fastGWA-workflow-implementation"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "Step 1: Creation of the GRM", "text": "Step 1 Creation of the GRM ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#Step-1:-Creation-of-the-GRM"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "Step 2: Combine all the GRM parts into one file", "text": "Step 2 Combine all the GRM parts into one file ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#Step-2:-Combine-all-the-GRM-parts-into-one-file"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "Step 3: Make a sparse GRM to be used in the association analyses", "text": "Step 3 Make a sparse GRM to be used in the association analyses ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#Step-3:-Make-a-sparse-GRM-to-be-used-in-the-association-analyses"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "Step 4: Run the single variant association analysis using FastGWA", "text": "Step 4 Run the single variant association analysis using FastGWA ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#Step-4:-Run-the-single-variant-association-analysis-using-FastGWA"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "REGENIE workflow implementation", "text": "REGENIE workflow implementation ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#REGENIE-workflow-implementation"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "Step 1: fitting the null", "text": "Step 1 fitting the null ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#Step-1:-fitting-the-null"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "Step 2: association analysis", "text": "Step 2 association analysis ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#Step-2:-association-analysis"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "Regenie burden test", "text": "Regenie burden test ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#Regenie-burden-test"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "SAIGE workflow implementation", "text": "SAIGE workflow implementation ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#SAIGE-workflow-implementation"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "Step 1: fitting the null", "text": "Step 1 fitting the null ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#Step-1:-fitting-the-null"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "Step 2: perform single variant association test", "text": "Step 2 perform single variant association test ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#Step-2:-perform-single-variant-association-test"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "GMMAT workflow implementation", "text": "GMMAT workflow implementation ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#GMMAT-workflow-implementation"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "Step 1: Creation of the GRM", "text": "Step 1 Creation of the GRM ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#Step-1:-Creation-of-the-GRM"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "Step 2: Fitting the null", "text": "Step 2 Fitting the null ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#Step-2:-Fitting-the-null"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "Step 3: Perform single variant score test for common variants", "text": "Step 3 Perform single variant score test for common variants ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#Step-3:-Perform-single-variant-score-test-for-common-variants"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "SMMAT workflow implementation", "text": "SMMAT workflow implementation ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#SMMAT-workflow-implementation"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "Step 1: Creation of the GRM", "text": "Step 1 Creation of the GRM ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#Step-1:-Creation-of-the-GRM"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "Step 2: Fitting the null", "text": "Step 2 Fitting the null ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#Step-2:-Fitting-the-null"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "Step 3: Perform variant set test for rare variants", "text": "Step 3 Perform variant set test for rare variants ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#Step-3:-Perform-variant-set-test-for-rare-variants"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "Merge results", "text": "Merge results ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#Merge-results"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "Manhattan and QQ plots", "text": "Manhattan and QQ plots ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#Manhattan-and-QQ-plots"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "Create analysis report", "text": "Create analysis report ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#Create-analysis-report"},
{"mainTitle": "LMM/GLMM analyses for UK Biobank data", "title": "Results", "text": "Results ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/LMM.html", "url": "http://cumc.github.io/bioworkflows/GWAS/LMM.html#Results"},
{"mainTitle": "Principal component analysis", "title": "Principal component analysis", "text": "Principal component analysis ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/PCA.html", "url": "http://cumc.github.io/bioworkflows/GWAS/PCA.html#Principal-component-analysis"},
{"mainTitle": "Principal component analysis", "title": "Overview", "text": "Overview ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/PCA.html", "url": "http://cumc.github.io/bioworkflows/GWAS/PCA.html#Overview"},
{"mainTitle": "Principal component analysis", "title": "Input", "text": "Input ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/PCA.html", "url": "http://cumc.github.io/bioworkflows/GWAS/PCA.html#Input"},
{"mainTitle": "Principal component analysis", "title": "Output", "text": "Output ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/PCA.html", "url": "http://cumc.github.io/bioworkflows/GWAS/PCA.html#Output"},
{"mainTitle": "Principal component analysis", "title": "General workflow", "text": "General workflow ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/PCA.html", "url": "http://cumc.github.io/bioworkflows/GWAS/PCA.html#General-workflow"},
{"mainTitle": "Principal component analysis", "title": "Method", "text": "Method ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/PCA.html", "url": "http://cumc.github.io/bioworkflows/GWAS/PCA.html#Method"},
{"mainTitle": "Principal component analysis", "title": "How to run this workflow", "text": "How to run this workflow ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/PCA.html", "url": "http://cumc.github.io/bioworkflows/GWAS/PCA.html#How-to-run-this-workflow"},
{"mainTitle": "Principal component analysis", "title": "Step 1: Estimate kinship in the sample", "text": "Step 1 Estimate kinship in the sample ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/PCA.html", "url": "http://cumc.github.io/bioworkflows/GWAS/PCA.html#Step-1:-Estimate-kinship-in-the-sample"},
{"mainTitle": "Principal component analysis", "title": "Step 2 and 3: Sample selection and QC the genetic data", "text": "Step 2 and 3 Sample selection and QC the genetic data ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/PCA.html", "url": "http://cumc.github.io/bioworkflows/GWAS/PCA.html#Step-2-and-3:-Sample-selection-and-QC-the-genetic-data"},
{"mainTitle": "Principal component analysis", "title": "Step 4: PCA analysis for all samples", "text": "Step 4 PCA analysis for all samples ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/PCA.html", "url": "http://cumc.github.io/bioworkflows/GWAS/PCA.html#Step-4:-PCA-analysis-for-all-samples"},
{"mainTitle": "Principal component analysis", "title": "Step 5: Projection of related individuals and outlier detection", "text": "Step 5 Projection of related individuals and outlier detection ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/PCA.html", "url": "http://cumc.github.io/bioworkflows/GWAS/PCA.html#Step-5:-Projection-of-related-individuals-and-outlier-detection"},
{"mainTitle": "Principal component analysis", "title": "Step 6: Split data by population", "text": "Step 6 Split data by population ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/PCA.html", "url": "http://cumc.github.io/bioworkflows/GWAS/PCA.html#Step-6:-Split-data-by-population"},
{"mainTitle": "Principal component analysis", "title": "Step 7: For each population perform QC", "text": "Step 7 For each population perform QC ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/PCA.html", "url": "http://cumc.github.io/bioworkflows/GWAS/PCA.html#Step-7:-For-each-population-perform-QC"},
{"mainTitle": "Principal component analysis", "title": "Step 8: For each population do PCA", "text": "Step 8 For each population do PCA ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/PCA.html", "url": "http://cumc.github.io/bioworkflows/GWAS/PCA.html#Step-8:-For-each-population-do-PCA"},
{"mainTitle": "Principal component analysis", "title": "Step 9: For each population projection of related individuals and outlier detection", "text": "Step 9 For each population projection of related individuals and outlier detection ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/PCA.html", "url": "http://cumc.github.io/bioworkflows/GWAS/PCA.html#Step-9:-For-each-population-projection-of-related-individuals-and-outlier-detection"},
{"mainTitle": "Principal component analysis", "title": "Step 10: For each population remove outliers", "text": "Step 10 For each population remove outliers ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/PCA.html", "url": "http://cumc.github.io/bioworkflows/GWAS/PCA.html#Step-10:-For-each-population-remove-outliers"},
{"mainTitle": "Principal component analysis", "title": "Minimal working example", "text": "Minimal working example ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/PCA.html", "url": "http://cumc.github.io/bioworkflows/GWAS/PCA.html#Minimal-working-example"},
{"mainTitle": "Principal component analysis", "title": "Data extraction for common variants", "text": "Data extraction for common variants ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/PCA.html", "url": "http://cumc.github.io/bioworkflows/GWAS/PCA.html#Data-extraction-for-common-variants"},
{"mainTitle": "Principal component analysis", "title": "Data extraction for rare variants", "text": "Data extraction for rare variants ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/PCA.html", "url": "http://cumc.github.io/bioworkflows/GWAS/PCA.html#Data-extraction-for-rare-variants"},
{"mainTitle": "Principal component analysis", "title": "Command interface", "text": "Command interface ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/PCA.html", "url": "http://cumc.github.io/bioworkflows/GWAS/PCA.html#Command-interface"},
{"mainTitle": "Principal component analysis", "title": "PCA analysis", "text": "PCA analysis ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/PCA.html", "url": "http://cumc.github.io/bioworkflows/GWAS/PCA.html#PCA-analysis"},
{"mainTitle": "Principal component analysis", "title": "Project related individuals back", "text": "Project related individuals back ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/PCA.html", "url": "http://cumc.github.io/bioworkflows/GWAS/PCA.html#Project-related-individuals-back"},
{"mainTitle": "Principal component analysis", "title": "Plot PCA results", "text": "Plot PCA results ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/PCA.html", "url": "http://cumc.github.io/bioworkflows/GWAS/PCA.html#Plot-PCA-results"},
{"mainTitle": "Principal component analysis", "title": "Detect outliers", "text": "Detect outliers ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/PCA.html", "url": "http://cumc.github.io/bioworkflows/GWAS/PCA.html#Detect-outliers"},
{"mainTitle": "Hudson plot", "title": "Hudson plot", "text": "Hudson plot ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/Hudson_Plot.html", "url": "http://cumc.github.io/bioworkflows/GWAS/Hudson_Plot.html#Hudson-plot"},
{"mainTitle": "Hudson plot", "title": "Aim", "text": "Aim ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/Hudson_Plot.html", "url": "http://cumc.github.io/bioworkflows/GWAS/Hudson_Plot.html#Aim"},
{"mainTitle": "Hudson plot", "title": "Command interface", "text": "Command interface ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/Hudson_Plot.html", "url": "http://cumc.github.io/bioworkflows/GWAS/Hudson_Plot.html#Command-interface"},
{"mainTitle": "Hudson plot", "title": "Illustration with minimal working examples", "text": "Illustration with minimal working examples ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/Hudson_Plot.html", "url": "http://cumc.github.io/bioworkflows/GWAS/Hudson_Plot.html#Illustration-with-minimal-working-examples"},
{"mainTitle": "Hudson plot", "title": "Hudson plot example command", "text": "Hudson plot example command ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/Hudson_Plot.html", "url": "http://cumc.github.io/bioworkflows/GWAS/Hudson_Plot.html#Hudson-plot-example-command"},
{"mainTitle": "Hudson plot", "title": "Visualization of the plot created (e.g real tinnitus data from UK Biobank)", "text": "Visualization of the plot created e.g real tinnitus data from UK Biobank ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/GWAS/Hudson_Plot.html", "url": "http://cumc.github.io/bioworkflows/GWAS/Hudson_Plot.html#Visualization-of-the-plot-created-(e.g-real-tinnitus-data-from-UK-Biobank)"},
{"mainTitle": "finemapping_to_ldsc.sos
         
      ", "title": "finemapping_to_ldsc.sos
         
      ", "text": "finemapping_to_ldsc.sos ", "tags": "", "mainUrl": "http://cumc.github.io/bioworkflows/fine-mapping/finemapping_to_ldsc.pipeline.html", "url": "http://cumc.github.io/bioworkflows/fine-mapping/finemapping_to_ldsc.pipeline.html"}
]};
